// advertising/sticky-node.js
import _ from 'lodash';
import getViewportWidth from 'helpers/viewport/get-width';

let stickyParams = {}, scrollEvent;

let getPosition = function(element) {
  let e = element;
  let xPosition = 0;
  let yPosition = 0;

  while (e) {
    xPosition += (e.offsetLeft - e.scrollLeft + e.clientLeft);
    yPosition += (e.offsetTop - e.scrollTop + e.clientTop);
    e = e.offsetParent;
  }

  return { x: xPosition, y: yPosition };
};

let stickyScrollEvent = function() {
  let scrollY = window.scrollY || window.pageYOffset;
  scrollBehaviour(scrollY);
};

let toggleStickyScrolling = function() {
  let breakpoint = stickyParams.breakpoint || 0;
  scrollEvent = _.throttle(stickyScrollEvent, 10);
  if (getViewportWidth() > breakpoint) {
    window.addEventListener('scroll', scrollEvent);
    stickyScrollEvent();
  } else {
    window.removeEventListener('scroll', scrollEvent);
  }
};

let fixedPosition = function(y) {
  stickyParams.element.style.position = y ? 'fixed' :  'static';
  stickyParams.element.style.top = y ? y + 'px' : 'auto';
};

let scrollBehaviour = function() {

  var yPosition = getPosition(stickyParams.parent).y;
  var stopY = getPosition(stickyParams.container).y + stickyParams.container.clientHeight - stickyParams.parent.offsetHeight;
  var y = 0;

  if (stopY - stickyParams.offset <= 0) {
    y = stopY;
  } else if (yPosition - stickyParams.offset <= 0) {
    y = stickyParams.offset;
  }

  fixedPosition(y);
};

export default {

  init: function(params) {
    let element = document.querySelector(params.element);
    let parent = document.querySelector(params.parent);
    if (element && parent) {
      // form stick params
      stickyParams = {
        element: element,
        parent: parent,
        container: document.querySelector(params.container) || parent.parentNode,
        offset: params.offset,
        breakpoint: params.breakpoint
      };
      // set events and init
      let debounced = _.debounce(toggleStickyScrolling, 200);
      window.addEventListener('resize', debounced, false);
      toggleStickyScrolling();
    } else {
      console.warn('Please define both element and parent for sticky node'); // eslint-disable-line no-console
    }
  },

};
