// advertising/sticky-node.jsx
var _ = require('lodash');
var getViewportWidth = require('helpers/get-viewport-width');

var stickyParams = {}, scrollEvent, stickyNodesCount = 0;

var stickyScrollEvent = function() {
  var scrollY = window.scrollY || window.pageYOffset;
  scrollBehaviour(scrollY);
};

var documentHeight = function() {
  return Math.max(
      window.innerHeight,
      document.body.offsetHeight,
      document.documentElement.clientHeight
  );
};

var toggleStickyScrolling = function() {
  var breakpoint = stickyParams.breakpoint || 0;
  scrollEvent = _.throttle(stickyScrollEvent, 10);
  if (getViewportWidth() > breakpoint) {
    // calculate dimensions
    var elHeightDifference = stickyParams.element.offsetHeight - (stickyParams.stopper ? stickyParams.stopper.offsetHeight : 0);
    // store these for later
    stickyParams.startStick = stickyParams.element.offsetTop - stickyParams.offset;
    stickyParams.stopStick = stickyParams.stopper
      ? stickyParams.stopper.offsetTop + (elHeightDifference - stickyParams.offset)
      : documentHeight() - stickyParams.offset - Math.min(window.innerHeight, stickyParams.element.offsetHeight);
    // add event and trigger
    window.addEventListener('scroll', scrollEvent);
    stickyScrollEvent();
  } else {
    window.removeEventListener('scroll', scrollEvent);
  }
};

var fixedPosition = function(y) {
  stickyParams.wrapper.style.position = y !== false ? 'fixed' :  'static';
  stickyParams.wrapper.style.top = y !== false ? y + 'px' : 'auto';
};

var scrollBehaviour = function(current) {
  var stickPosition = false;

  if (current >= stickyParams.startStick && current < stickyParams.stopStick) {
    stickPosition = stickyParams.offset;
  }

  if (current >= stickyParams.stopStick) {
    stickPosition = stickyParams.offset - (current - stickyParams.stopStick);
  }

  fixedPosition(stickPosition);
};

var createStickyWrapper = function(el, i) {
  var org_html = el.innerHTML;
  var new_html = '<div id="stickyNode--' + i + '">' + org_html + '</div>';
  el.innerHTML = new_html;
};

var StickyNodes = {

  init: function(params) {
    var element = document.querySelector(params.element);
    if (element) {
      // wrap the element so we can always get offsetTop value from the parent
      createStickyWrapper(element, stickyNodesCount);
      // form stick params
      stickyParams = {
        element: element,
        wrapper: document.getElementById('stickyNode--' + stickyNodesCount),
        stopper: document.querySelector(params.stopper),
        offset: params.offset,
        breakpoint: params.breakpoint
      };
      // set events and init
      var debounced = _.debounce(toggleStickyScrolling, 200);
      window.addEventListener('resize', debounced, false);
      toggleStickyScrolling();
      stickyNodesCount++;
    }
  },

};

module.exports = StickyNodes;
