// helpers/teaser/get-from-article.jsx
import _ from 'lodash';

import {
  getArticleTitle,
  getArticleUrl,
  getArticleImage,
  getArticleMarker,
  getArticleSchema,
  getArticleTheme,
  getTeaserByline,
  getArticleSummary,
  getArticleCatchword
} from 'helpers/article';

import getTeaserVariation from './get-variation';

const switchedVariationsForLabelType = {
  basic: {
    meinung: 'quote'
  }
};

export default function(data) {

  if (!data) {
    return false;
  }

  const { article, variation } = data;

  // test for changed variation of certain labeltypes
  let labelType = article.labeltype;
  let find = variation + '.' + labelType;
  let switchedVariation = _.get(switchedVariationsForLabelType, find);

  // set variation
  let teaserVariation = switchedVariation || variation;

  return {
    title: getArticleTitle(article),
    link: getArticleUrl(article),
    marker: getArticleMarker(article),
    schema: getArticleSchema(article),
    // props which change based on teaser variation
    catchword: getArticleCatchword(article, teaserVariation),
    variation: getTeaserVariation(article, teaserVariation),
    theme: getArticleTheme(article, teaserVariation),
    image: getArticleImage(article, teaserVariation),
    byline: getTeaserByline(article, teaserVariation),
    summary: getArticleSummary(article, teaserVariation)
  };
}
