// helpers/teaser/get-variation-config.jsx
import _ from 'lodash';

const config = {
  'basic': {
    image: {
      size: 'small',
      ratio: '16x9'
    },
    summary: false,
    byline: {
      dateFormat: 'long'
    },
  },
  'large': {
    image: {
      size: 'small',
      ratio: '16x9'
    },
    summary: false,
    byline: {
      dateFormat: false
    },
  },
  'inlined-left': {
    image: {
      size: 'tiny',
      ratio: '1x1'
    },
    summary: false,
    byline: {
      dateFormat: 'short'
    }
  },
  'inlined-right': {
    image: {
      size: 'tiny',
      ratio: '1x1'
    },
    summary: false,
    byline: {
      dateFormat: 'short'
    }
  },
  'mega': {
    image: {
      size: 'full',
      ratio: '9x4'
    },
    summary: false,
    byline: {
      dateFormat: false
    }
  },
  'quote': {
    image: {
      size: 'tiny',
      ratio: '1x1'
    },
    summary: false,
    byline: {
      dateFormat: false
    }
  },
  'light': {
    image: false,
    summary: false,
    byline: {
      dateFormat: 'long'
    }
  },
  'light-boxed': {
    image: false,
    summary: false,
    byline: {
      dateFormat: 'long'
    }
  },
  'plain': {
    image: false,
    summary: false,
    byline: false
  },
  'plain-inverted': {
    image: false,
    summary: false,
    byline: false
  },
  // summary and listed have a switch-view teaser variation
  // so we keep image, summary and catchword object structure the same
  'summary': {
    image: {
      size: 'small',
      ratio: '16x9'
    },
    summary: {
      showing: true
    },
    catchword: {
      showing: true
    },
    byline: {
      dateFormat: 'long',
      icon: false
    }
  },
  'listed': {
    image: {
      size: false,
      ratio: false
    },
    summary: {
      showing: false
    },
    catchword: {
      showing: false
    },
    byline: {
      dateFormat: 'relative',
      icon: 'bullet'
    }
  },
};

export default function(variation) {
  return _.get(config, variation);
}
