// helpers/teaser/get-from-article.jsx
import _ from 'lodash';

import {
  getArticleTitle,
  getArticleUrl,
  getArticleImage,
  getArticleMarker,
  getArticleSchema,
  getArticleTheme,
  getArticleByline,
  getArticleSummary,
  getArticleCatchword
} from 'helpers/article';

import getResponsiveConfig from './get-responsive-config';

export default function(data) {

  if (!data) {
    return false;
  }

  const article = data.article;
  const _variation = data.variation;

  let responsiveConfig = getResponsiveConfig(_variation);
  // always serve the small version of responsive teasers first, enhance in the browser
  let variation = _.get(responsiveConfig, 'small') || _variation;

  return {
    variation: variation,
    title: getArticleTitle(article),
    link: getArticleUrl(article),
    marker: getArticleMarker(article),
    schema: getArticleSchema(article),
    catchword: getArticleCatchword(article),
    // props which change based on teaser variation
    responsive: responsiveConfig,
    theme: getArticleTheme(article, variation),
    image: getArticleImage(article, variation),
    byline: getArticleByline(article, variation),
    summary: getArticleSummary(article, variation)
  };
}
