// helpers/teaser/get-from-article.jsx
import {
  getArticleTitle,
  getArticleUrl,
  getArticleImage,
  getArticleMarker,
  getArticleSchema,
  getArticleTheme,
  getArticleByline,
  getArticleSummary,
  getArticleCatchword
} from 'helpers/article';

import getResponsiveConfig from './get-responsive-config';
import getVariationExceptions from './get-variation-exceptions';
import getVariationForView from './get-variation-for-view';

export default function(data) {

  if (!data) {
    return false;
  }

  const article = data.article;
  const variation = data.variation;

  // always serve the small version of responsive teasers first and enhance in the browser
  let variationForView = getVariationForView(variation, 'small');

  return {
    title: getArticleTitle(article),
    link: getArticleUrl(article),
    marker: getArticleMarker(article),
    schema: getArticleSchema(article),
    catchword: getArticleCatchword(article),
    // props which change based on teaser variation / responsiveness
    variation: getVariationExceptions(variationForView),
    responsive: getResponsiveConfig(variation),
    theme: getArticleTheme(article, variationForView),
    image: getArticleImage(article, variationForView),
    byline: getArticleByline(article, variationForView),
    summary: getArticleSummary(article, variationForView)
  };
}
