// components/primary-nav/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Icon from 'components/icon';

const classes = new bem({
  name: 'primary-nav',
  prefix: ''
});

export default React.createClass({

  displayName: 'PrimaryNav',

  propTypes: {
    items: React.PropTypes.array.isRequired,
  },

  showAlert() {
    /* eslint-disable no-alert */
    window.alert('Leider wurde dieses Feature noch nicht implementiert.');
  },

  render() {
    const { items } = this.props;
    return (
      <nav {...classes()} role='navigation'>
        <ul {...classes('list')}>
          <li {...classes('item')}>
            <a {...classes('link', 'login')} href="#" onClick={this.showAlert}>
              <Icon id="user"/> Login
            </a>
          </li>
          {items.map((item, index) =>
            <li {...classes('item')} key={index}>
              <a {...classes('link', null, [item.active ? 'is-active' : null])} href={item.href}>
                {item.label}
              </a>
            </li>
          )}
        </ul>
      </nav>
    );
  }

});
