// components/primary-nav/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Icon from 'components/icon';
import isLoggedIn from 'helpers/dashboard/is-logged-in';

const classes = new bem({
  name: 'primary-nav',
  prefix: ''
});

export default React.createClass({

  displayName: 'PrimaryNav',

  propTypes: {
    items: React.PropTypes.array.isRequired,
  },

  componentDidMount: function() {
    this.setState({
      browser: true,
      isLoggedIn: isLoggedIn()
    });
  },

  getInitialState: () => ({
    browser: false,
    isLoggedIn: false
  }),

  renderLoginLink() {
    let link = '/anmelden';
    let label = this.state.isLoggedIn ? 'Profil' : 'Anmelden';

    if (this.state.isLoggedIn) {
      link = '/dashboard/profil';
    } else if (this.state.browser) {
      link += '?redirectTo=' + encodeURIComponent(document.location.href);
    }

    return (
      <a {...classes('link', 'login')} href={ link }>
        <Icon id="user"/> { label }
      </a>
    );
  },

  render() {
    const { items } = this.props;
    return (
      <nav {...classes()} role='navigation'>
        <ul {...classes('list')}>
          <li {...classes('item')}>
            { this.renderLoginLink() }
          </li>
          {items.map((item, index) =>
            <li {...classes('item')} key={index}>
              <a {...classes('link', null, [item.active ? 'is-active' : null])} href={item.href}>
                {item.label}
              </a>
            </li>
          )}
        </ul>
      </nav>
    );
  }

});
