// components/gallery/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import GalleryDataMap from 'higher-order/gallery-data-map';
import GalleryMixin from 'mixins/gallery';
import ResponsiveImage from 'components/responsive-image';
import Caption from 'components/caption';

import tapEventPlugin from 'react-tap-event-plugin';
tapEventPlugin();

const classes = new bem({
  name: 'gallery',
  prefix: ''
});

const Gallery = React.createClass({

  mixins: [GalleryMixin],

  propTypes: {
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    images: React.PropTypes.array.isRequired
  },

  getInitialState() {
    return {
      currentIndex: 0,
      fullscreen: false
    };
  },

  render() {
    return (
      <div {...classes()}>
        <figure itemScope itemType='http://schema.org/ImageObject' {...classes('image')}>
          <ResponsiveImage
            ratio='16x9'
            size='max'
            title={this.imageData('title')}
            alt={this.imageData('description')}
            src={this.imageData('url')} />
          <div {...classes('slide-number')}>
            {this.state.currentIndex + 1}/{this.props.images.length}
          </div>
          <Caption
            title={this.imageData('title')}
            description={this.imageData('description')}
            source={this.imageData('source')} />
        </figure>
        <div {...classes('controls')}>
          <button type='button' onTouchTap={this.navigateSlider.bind(this, -1)}>Zur&uuml;ck</button>
          <button type='button' onTouchTap={this.navigateSlider.bind(this, +1)}>Weiter</button>
        </div>
      </div>
    );
  }

});

export default GalleryDataMap(Gallery);
