// components/weather-forecast/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';
import { getAssetPath } from 'helpers';

const classes = new BEM('weather-forecast');

export default React.createClass({

  displayName: 'WeatherForecast',

  propTypes: {
    label: React.PropTypes.string,
    id: React.PropTypes.string.isRequired,
    modifier: React.PropTypes.string,
    url: React.PropTypes.string,

    minTemperature: React.PropTypes.string.isRequired,
    maxTemperature: React.PropTypes.string.isRequired
  },

  getDefaultProps() {
    return {
      modifier: 'default'
    };
  },

  render() {

    const { label, id, modifier, url, minTemperature, maxTemperature } = this.props;

    let weatherImgUrl = label ? 'blue/' : 'color/';
    let href = getAssetPath('images/weather/' + weatherImgUrl + id + '.svg');

    const modifiers = {
      img: id && !label,
      [modifier]: !!modifier
    };

    if (label) {
      return (
        <a href={url} {...classes(null, modifiers)}>
          <span {...classes('label')}>{label}</span>
          <div {...classes('icon')}>
            <img {...classes('img')} src={href} />
          </div>
          <div {...classes('temperature')}>
            <div {...classes('max')}>{maxTemperature}째</div>
            <div {...classes('min')}>{minTemperature}째</div>
          </div>
        </a>
      );
    }
    return (
      <div {...classes(null, modifiers)}>
        <div {...classes('icon')}>
          <img {...classes('img')} src={href} />
        </div>
        <div {...classes('temperature')}>
          <div {...classes('max')}>{maxTemperature}째</div>
          <div {...classes('min')}>{minTemperature}째</div>
        </div>
      </div>
    );
  }
});
