// dashboard/sidebar/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Icon from 'components/icon';
import { Link } from 'react-router';
// import { get } from 'lodash/object';

const classes = new bem('dashboard');
const menuItems = [
  {
    path: '/dashboard/profil',
    label: 'Profil',
    icon: 'user'
  },
  {
    path: '/dashboard/beitrage',
    label: 'Beitrage',
    icon: 'megaphone'
  },
  {
    path: '/dashboard/vereine',
    label: 'Vereine',
    icon: 'crest'
  },
  {
    path: 'http://www.aargauerzeitung.ch/hilfe',
    label: 'Hilfe',
    icon: 'help'
  }
];

export default React.createClass({

  displayName: 'DashboardSidebar',

  propTypes: {
    currentPath: React.PropTypes.string
  },

  renderItem(options) {
    var isActive = options.path === this.props.currentPath;
    var contents;

    if (options.path.indexOf('http') === 0) {
      contents = (
        <a href={options.path} >
          <Icon id={options.icon}/>
          { options.label }
        </a>
      );
    } else {
      contents = (
        <Link to={options.path}>
          <Icon id={options.icon}/>
          { options.label }
        </Link>
      );
    }

    return (
      <li key={options.label} {...classes('sidebar-item', { active: isActive })}>
        { contents }
      </li>
    );
  },

  render() {
    return (
      <nav {...classes('sidebar')} >
        <ul>
          { menuItems.map(this.renderItem) }
        </ul>
      </nav>
    );
  }
});
