// dashboard/sidebar/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Icon from 'components/icon';
import { Link } from 'react-router';
// import { get } from 'lodash/object';

const classes = new bem('dashboard');

export default React.createClass({

  displayName: 'DashboardSidebar',

  propTypes: {
    currentPath: React.PropTypes.string
  },

  renderItem(options) {
    var isActive = options.path === this.props.currentPath;
    return (
      <li {...classes('sidebar-item', { active: isActive })}>
        <Link to={options.path}>
          <Icon id={options.icon}/>
          { options.label }
        </Link>
      </li>
    );
  },

  render() {
    return (
      <nav {...classes('sidebar')} >
        <ul>
          { this.renderItem({
              path: '/dashboard/profil',
              label: 'Profil',
              icon: 'user'
          })}
          { this.renderItem({
              path: '/dashboard/beitrage',
              label: 'Beitrage',
              icon: 'megaphone'
          })}
          { this.renderItem({
              path: '/dashboard/vereine',
              label: 'Vereine',
              icon: 'crest'
          })}
          { this.renderItem({
              path: 'http://www.aargauerzeitung.ch/hilfe',
              label: 'Hilfe',
              icon: 'help'
          })}
        </ul>
      </nav>
    );
  }
});
