// components/play-time/index.jsx
import React from 'react';
import moment from 'moment';

import Icon from 'components/icon';

export default React.createClass({

  displayName: 'PlayTime',

  propTypes: {
    duration: React.PropTypes.number.isRequired
  },

  getFormattedTime(duration) {
    var time;
    var format;

    format = (duration < 3600) ? 'mm:ss' : 'HH:mm:ss';

    time = moment.utc(duration * 1000).format(format);
    time = (duration ? time : '');

    return time;
  },

  getIsoTime(duration) {
    return moment.duration(duration, 'seconds').toISOString();
  },

  render() {
    var playTime = this.getFormattedTime(this.props.duration);
    var isoTime = this.getIsoTime(this.props.duration);
    var metaDuration;

    if (this.props.duration && this.props.duration !== 0) {
      metaDuration = (
        <meta itemProp="duration" content={isoTime} />
      );
    }

    return (
      <div className="play-time">
        <Icon id="play" />
        {playTime}
        {metaDuration}
      </div>
    );
  }
});
