// components/newsletter-for-region/index.jsx
import { get } from 'lodash';
import React from 'react';
import bem from 'react-bem-helper';
import NewsletterForm from 'components/newsletter-form';
import TextInput from 'inputs/text';
import Button from 'components/button';
import regionData from 'config/newsletters.json';

const classes = new bem('newsletter-for-region');

export default React.createClass({

  displayName: 'NewsletterForRegion',

  propTypes: {
    regionPath: React.PropTypes.array.isRequired
  },

  getInitialState() {
    return {
      region: get(regionData, this.props.regionPath)
    };
  },

  render() {
    if (!this.state.region) {
      return null;
    }

    const { group, id, townName } = this.state.region;
    const fieldName = `group[${group}][${id}]`;

    return (
      <aside {...classes()}>
        <div {...classes('text')}>
          <h3 {...classes('header')}>Newsletter</h3>
          <p>Wissen was in {townName || 'der Region'} läuft? <br/>Hier anmelden für den täglichen Newsletter.</p>
        </div>

        <NewsletterForm {...classes('form')}>
          <input name={fieldName} key={fieldName} type="hidden" value={id} />
          <TextInput
            {...classes('input')}
            label=""
            name="EMAIL"
            type="email"
            placeholder="E-Mail Adresse"
          />
          <Button attributes={{ type: 'submit' }} label="Anmelden" modifier="primary"/>
        </NewsletterForm>
      </aside>
    );
  }
});
