// higher-order/get-enrichment.jsx
import React from 'react';
import _ from 'lodash';

// List all enrichments
import EnrichmentText from 'components/enrichment-text';
import EnrichmentImage from 'components/enrichment-image';
import EnrichmentDocument from 'components/enrichment-document';
import EnrichmentVideo from 'components/enrichment-video';
import EnrichmentAudio from 'components/enrichment-audio';
import EnrichmentGallery from 'components/enrichment-gallery';
import EnrichmentSnippet from 'components/enrichment-snippet';

export default React.createClass({

  displayName: 'GetEnrichment',

  propTypes: {
    enrichment: React.PropTypes.object.isRequired
  },

  getEnrichmentComponent(kind) {
    const kinds = {
      'text_stoerer': EnrichmentText,
      'image_stoerer': EnrichmentImage,
      'document_stoerer': EnrichmentDocument,
      'video_stoerer': EnrichmentVideo,
      'audio_stoerer': EnrichmentAudio,
      'gallery_stoerer': EnrichmentGallery,
      'snippet_stoerer': EnrichmentSnippet,
      // This is meant for the main asset which is not an interferer in itself
      // and has its own `content_type`. Still, we want it to behave the same
      // way as any other störer.
      'asset_image': EnrichmentImage,
      'asset_image_gallery': EnrichmentGallery,
      'asset_video': EnrichmentVideo
    };

    return kinds[kind];
  },

  isGalleryEnrichment(type) {
    return type === 'gallery_stoerer' || type === 'asset_image_gallery';
  },

  galleryWithSingleImage(enrichment) {
    const type = enrichment.stoerer_type || enrichment.asset.content_type;
    let images = _.get(enrichment, 'asset.images');
    if (this.isGalleryEnrichment(type) && images.length === 1) return true;
  },

  render() {
    let { enrichment } = this.props;
    let type = enrichment.stoerer_type || enrichment.asset.content_type;

    // test for single images in gallery and switch to image stoerer
    if (this.galleryWithSingleImage(enrichment)) {
      type = 'image_stoerer';
      enrichment = enrichment.asset.images[0];
    }

    let Enrichment = this.getEnrichmentComponent(type);

    if (Enrichment) {
      return <Enrichment {...enrichment} />;
    }

  }

});
