// higher-order/kaltura-api-proxy.jsx
import React from 'react';
import axios from 'axios';

export default React.createClass({

  displayName: 'KalturaAPIProxy',

  propTypes: {
    format: React.PropTypes.number.isRequired,
    mediaTypeEqual: React.PropTypes.number.isRequired,
    pageSize: React.PropTypes.number.isRequired,
    component: React.PropTypes.react
  },

  getDefaultProps() {
    return {
      // response in json format
      format: 1,
      // =1 for video
      mediaTypeEqual: 1,
      pageSize: 25,
      videos: [
        {
          name: 'Dummy Video'
        }
      ]
    };
  },

  getInitialState() {
    return {
      //// baseEntry for all media objects like Kaltura videos, YouTube videos, audio etc.
      service: 'baseEntry',
      action: 'list',
      filter: {
        freeText: null,
        //@TODO: add constants and function
        orderBy: '-createdAt',
        categoriesIdsMatchAnd: null
      },
      pager: {
        pageIndex: 1
      },
      videos: this.props.videos
    };
  },

  getURL() {
    var protocol = 'http://';
    var server = 'www.aargauerzeitung.ch';
    var path = '/__kaltura_api_proxy__';
    var queryString = '?action='+this.state.action
                      + '&filter:categoriesIdsMatchAnd'+this.state.filter.categoriesIdsMatchAnd
                      // @TODO: put in separate function getFreeTextQuery
                      + '&filter:freeText='+this.state.filter.freeText+'*'
                      + '&filter:mediaTypeEqual='+this.props.mediaTypeEqual
                      + '&filter:orderBy='+this.state.filter.orderBy
                      + '&format='+this.props.format
                      + '&pager:pageIndex='+this.state.pager.pageIndex
                      + '&pager:pageSize='+this.props.pageSize
                      + '&service='+this.state.service;

    return protocol+server+path+queryString;
  },

  getVideos() {
    var url = this.getURL();

    axios.get(url)
    .then(function (response) {
      this.setState({
        videos: response.data
      });
      // console.log('response', response);
    })
    .catch(function () {
      // console.log('Error loading KalturaAPIProxy data', response);
    });
  },

  render() {
    var Component = this.props.component;
    return Component;
  }
});
