// pages/dossier/index.jsx
import React from 'react';

import Container from 'components/container';
import DossierHeader from 'components/dossier-header';
import Grid from 'components/grid';
import Teaser from 'components/teaser';
import { getTeaserFromArticle } from 'helpers/teaser';

export default React.createClass({

  displayName: 'DossierPage',

  propTypes: {
    dossier: React.PropTypes.object.isRequired,
    articles: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mapTeaserItems(articles) {
    return articles.map((article, index) => {
      var articleTeaser = getTeaserFromArticle({
        article: article,
        variation: index === 0 ? 'large' : index < 6 ? 'basic' : 'basic-listed',
        fallback: 'light'
      });
      return <Teaser key={index} {...articleTeaser} />;
    });
  },

  render() {
    return (
      <Container>
        <DossierHeader {...this.props.dossier} />
        <Grid modifier='three-up-featured' items={this.mapTeaserItems(this.props.articles)} />
      </Container>
    );
  }

});
