// components/video-library-list/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import VideoTeaser from 'components/video-teaser';
import VideoLibraryViewer from 'components/video-library-viewer';
import Expander from 'components/expander';
import Grid from 'components/grid';

import Button from 'components/button';

const classes = new bem({
  name: 'video-library-list',
  prefix: ''
});

export default React.createClass({

  displayName: 'VideoLibraryList',

  propTypes: {
    accountID: React.PropTypes.string.isRequired,
    playerID: React.PropTypes.string.isRequired,
    videos: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    selectedVideo: React.PropTypes.string
  },

  mapTeaserItems(items) {
    return items.map((item, index) => {
      var selectedVideo = item.id;
      var listItem, listItemViewer, expander;
      var expanded;
      var key = 'video' + index;

      // if video is selected render viewer in expander component.
      // otherwise just render the teaser in collapsed state.
      if (selectedVideo && selectedVideo===this.props.selectedVideo) {
        listItemViewer = <VideoLibraryViewer
                          {...this.props}
                          videoID={item.id}
                          name={item.name}
                          description={item.description}
                          thumbnailUrl={item.thumbnailUrl}/>;
        expander = <Expander
                          item={listItemViewer}/>;
      }
      // @TODO:@WithRichard:expanded: refactor to bundle expander's state in a component or mixin?
      expanded = this.props.selectedVideo===item.id ? true : false;
      listItem = <VideoTeaser
                    expanded={expanded}
                    name={item.name}
                    thumbnailUrl={item.thumbnailUrl}
                    videoID={item.id}
                    viewer={expander}
                    key={key}
                    selectVideo={this.props.selectVideo}/>;

      return listItem;
    });
  },

  render() {
    var videoTeasers = this.props.videos;
    var errorMessage = 'Leider konnten keine Videos zu Ihrem Suchbegriff gefunden werden';
    var items;
    var content;


    if (videoTeasers.length) {
      items = this.mapTeaserItems(videoTeasers);
      content = <Grid modifier='video-library' items={items} />;

    } else {
      content = (
        <div {...classes('fallback')}>
          <Alert text={errorMessage} modifier='centered' />
        </div>
      );
    }

    return (
      <div {...classes('container')}>
        {content}
        <Button label="Mehr anzeigen" _target="#"/>
      </div>
    );
  }

});
