// components/video-library-list/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import VideoTeaser from 'components/video-teaser';
import VideoLibraryViewer from 'components/video-library-viewer';
import Expandable from 'components/expandable';
import Expander from 'components/expander';

import Button from 'components/button';

const classes = new bem({
  name: 'video-library-list',
  prefix: ''
});

export default React.createClass({

  displayName: 'VideoLibraryList',

  propTypes: {
    accountID: React.PropTypes.string.isRequired,
    playerID: React.PropTypes.string.isRequired,
    videos: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    selectedVideo: React.PropTypes.string
  },

  mapTeaserItems(items) {
    return items.map((item, index) => {
      var selectedVideo = item.id;
      var listItem, listItemViewer, expander, expanded;
      var key = 'video' + index;

      // if video is selected render viewer in expander component.
      // otherwise just render the teaser in collapsed state.
      if (selectedVideo && selectedVideo===this.props.selectedVideo) {
        listItemViewer = (
          <VideoLibraryViewer
            {...this.props}
            videoID={item.id}
            name={item.name}
            description={item.description}
            thumbnailUrl={item.thumbnailUrl}
          />
        );
        expander = <Expander item={listItemViewer}/>;
      }
      expanded = this.props.selectedVideo===item.id ? true : false;
      listItem = (
        <Expandable expanded={expanded} key={key}>
          <VideoTeaser
            name={item.name}
            thumbnailUrl={item.thumbnailUrl}
            videoID={item.id}
            onToggleVideo={this.props.onToggleVideo}
          >
            {expander}
          </VideoTeaser>
        </Expandable>
      );

      return listItem;
    });
  },

  render() {
    var videoTeasers = this.props.videos;
    var errorMessage = 'Leider konnten keine Videos zu Ihrem Suchbegriff gefunden werden';
    var content;


    if (videoTeasers.length) {
      content = this.mapTeaserItems(videoTeasers);
    } else {
      content = (
        <div {...classes('fallback')}>
          <Alert text={errorMessage} modifier='centered' />
        </div>
      );
    }

    return (
      <div {...classes('container')}>
        {content}
        <div {...classes('showmore')}>
          <Button label="Mehr anzeigen" _target="#"/>
        </div>
      </div>
    );
  }

});
