// components/enrichment-audio/wrapper.jsx
import React from 'react';

import getAssetMeta from 'helpers/get-asset-meta';

export default Component =>  React.createClass({

  displayName: 'EnrichmentAudioWrapper',

  propTypes: {
    asset: React.PropTypes.object.isRequired,
    title: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    placement: React.PropTypes.string,
  },

  createAudioMeta(props) {
    return {
      thumbnailUrl: props.still_image ? props.still_image.image_url || props.still_image.asset.image_url : null
    };
  },

  mappedData(props) {
    return {
      kalturaID: props.asset.kaltura_id,
      accountID: process.env.KALTURA_ACCOUNT_ID,
      playerID: process.env.KALTURA_PLAYER_ID,
      playerWithoutAdsID: process.env.KALTURA_PLAYER_NOADS_ID,
      trackingURL: process.env.KALTURA_TRACKING_URL,
      meta: this.createVideoMeta(props.asset),
      caption: getAssetMeta(props),
      placement: props.placement
    };
  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
