// inputs/text/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import InputWrapper from 'inputs/wrapper';

const classes = new bem('input');

export default React.createClass({
  displayName: 'TextInput',

  propTypes: {
    label: React.PropTypes.string,
    className: React.PropTypes.string,
    name: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    hint: React.PropTypes.string,
    onChange: React.PropTypes.func,
    placeholder: React.PropTypes.string,
    id: React.PropTypes.string,
    type: React.PropTypes.string,
    value: React.PropTypes.string,
  },

  getInitialState() {
    return {
      value: this.props.value
    }
  },

  getDefaultProps: () => ({
    type: 'text'
  }),

  onValueChange: function() {
    var value = this.refs.input.getDOMNode().value;
    this.setState({ value: value })

    if (this.props.onChange) {
      this.props.onChange(this.props.name, value);
    }
  },

  render() {
    var id = this.props.id || this.props.name;

    return (
      <InputWrapper {...this.props} id={id}>
        <input {...this.props} {...classes()} id={id} ref='input' onChange={this.onValueChange} value={this.state.value} />
      </InputWrapper>
    )
  }
});
