// components/related-videos/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getScript } from 'helpers';
import Slider from 'mixins/slider';

const classes = new bem({
  name: 'related-videos',
  prefix: ''
});

var RelatedVideos = React.createClass({

  displayName: 'RelatedVideos',

  mixins: [Slider],

  getInitialState() {
    return {
      mobile: false,
      loading: true
    };
  },

  statics: {
    veeseoPromise: null,
    loadVeeseo: function () {
      // TODO: make this dynamic from domain config
      var domain = 'aargauerzeitung';
      var url = 'http://rce.veeseo.com/widgets/'+ domain +'/widget.js';

      RelatedVideos.veeseoPromise = getScript(url);
      return RelatedVideos.veeseoPromise;
    }
  },

  componentDidMount() {
    this.loadVeeseo();
  },

  loadVeeseo() {
    RelatedVideos.loadVeeseo().then( () => {
      this.initSlider();
    }).catch( () => {
      this.setState({
        loadError: true
      });
    });
  },

  initSlider() {
    this.renderFlickity({
      startIndex: 1,
      element: '.recommendation ul',
      delay: 2500
    });
  },

  render() {

    var extraClasses = {
      'is-slider-loading': this.state.loading,
      'is-slider-resizing': this.state.resizing
    };

    return (
      <section ref='related' {...classes(null, null, extraClasses)}>
        <h2 {...classes('title')}>Verwandte Videos</h2>
        <div className="veeseoRA2VW2"></div>
      </section>
    );
  }
});

export default RelatedVideos;
