// components/related-videos/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getScript } from 'helpers';
import Slider from 'mixins/slider';

const classes = new bem({
  name: 'related-videos',
  prefix: ''
});

var RelatedVideos = React.createClass({

  displayName: 'RelatedVideos',

  mixins: [Slider],

  getInitialState() {
    return {
      mobile: false,
      loading: true
    };
  },

  statics: {
    loadedVeeseo: false,
    loadVeeseo: function (callback) {
      if (!RelatedVideos.loadedVeeseo) {
        var nestedCallback = function () {
          RelatedVideos.loadedVeeseo = false;
          callback();
        };
        var domain = 'aargauerzeitung';

        getScript(
          'http://rce.veeseo.com/widgets/'+ domain +'/widget.js',
          nestedCallback
        );
      } else {
        callback();
      }
    }
  },

  componentDidMount() {
    this.loadVeeseo();
  },

  loadVeeseo() {
    RelatedVideos.loadVeeseo(() => {
      this.renderFlickity({
        startIndex: 1,
        element: '.recommendation ul',
        delay: 2500
      });
    });
  },

  render() {

    var extraClasses = {
      'is-slider-loading': this.state.loading,
      'is-slider-resizing': this.state.resizing
    };

    return (
      <section ref='related' {...classes(null, null, extraClasses)}>
        <h2 {...classes('title')}>Verwandte Videos</h2>
        <div className="veeseoRA2VW2"></div>
      </section>
    );
  }
});

export default RelatedVideos;
