// components/related-videos/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getScript } from 'helpers';
import Slider from 'mixins/slider';
import Section from 'components/section';
import Container from 'components/container';
import SectionHeader from 'components/section-header';

const classes = new bem({
  name: 'related-videos',
  prefix: ''
});

export default React.createClass({

  displayName: 'RelatedVideos',

  mixins: [Slider],

  getInitialState() {
    return {
      mobile: false,
      loading: true
    };
  },

  componentDidMount() {
    this.loadVeeseo();
  },

  loadVeeseo() {
    let widgetUrl = '//rce.veeseo.com/widgets/aargauerzeitung/widget.js';

    getScript(widgetUrl).then(() => {
      this.renderFlickity({
        startIndex: 1,
        element: '.recommendation ul',
        delay: 1000
      });
    });
  },

  render() {

    var extraClasses = {
      'is-slider-loading': this.state.loading,
      'is-slider-resizing': this.state.resizing
    };

    return (
      <div ref='related' {...classes(null, null, extraClasses)}>
        <Section modifier='dark'>
          <Container>
            <SectionHeader title='Verwandte Videos' modifier='inverted' />
            <div className="veeseoRA2VW2"></div>
          </Container>
        </Section>
      </div>
    );
  }
});
