// components/article-content/wrapper/index.jsx
import React from 'react';

import GetEnrichment from 'helpers/get-enrichment';
import RawContent from 'components/raw-content';
import Aside from 'components/aside';
import ArticleContent from 'components/article-content'

export default React.createClass({

  propTypes: {
    assets: React.PropTypes.array.isRequired,
    lead: React.PropTypes.string.isRequired,
    text: React.PropTypes.string.isRequired,
    stoerer: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    keywords: React.PropTypes.arrayOf(React.PropTypes.string).isRequired
  },

  mapMainAsset() {
    if (this.props.assets.length) {
      return <GetEnrichment enrichment={this.props.assets[0]} />;
    }
  },

  mapContentAndEnrichments() {
    var chunks = [];
    var token = '<p>${{stoerer}}</p>';
    var text = this.props.text
      // Remove junk from CMS
      .replace(/<p>(\s|&nbsp;)*<\/p>/g, '');
    var index = text.indexOf(token);
    var enrichmentIndex = 0, rawIndex = 0;

    // Look for token
    while (index > -1) {
      // If there is token, render raw content before token
      if (index > 0) {
        chunks.push(<RawContent key={'raw' + rawIndex} content={text.slice(0, index)} />);
        rawIndex += 1;
      }

      // Then render enrichment
      var enrichment = this.props.stoerer[enrichmentIndex];
      if (enrichment) {
        var Enrichment = <GetEnrichment key={'enrichment_' + enrichmentIndex} enrichment={enrichment} />;
        var wrapped = enrichment.placement && ['left', 'right'].indexOf(enrichment.placement) > -1;
        chunks.push(wrapped ? <Aside key={'enrichmentAside_' + enrichmentIndex}>{Enrichment}</Aside> : Enrichment);
      }

      // Update the remaining text
      text = text.slice(index + token.length);
      // Look for next token
      index = text.indexOf(token);
      enrichmentIndex += 1;
    }

    // Make sure we don't leave any content if there are more inteferers than content chunks
    if (text.length > 0) {
      chunks.push(<RawContent key='rawLast' content={text} />);
    }

    return chunks;
  },

  mappedData() {
    return {
      asset: this.mapMainAsset(),
      intro: this.props.lead,
      body: this.mapContentAndEnrichments(),
      tags: this.props.keywords
    };
  },

  render() {
    return (
      <ArticleContent {...this.mappedData()} />
    );
  }

});
