// components/article-detail/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import ArticleHeader from 'components/article-header';
import MainAsset from 'components/main-asset';
import ArticleBody from 'components/article-body';
import Container from 'components/container';

const classes = new bem({
  name: 'article-content',
  prefix: ''
});

export default React.createClass({

  displayName: 'ArticleDetail',

  propTypes: {
    header: React.PropTypes.object.isRequired,
    mainAsset: React.PropTypes.object,
    enrichments: React.PropTypes.arrayOf(React.PropTypes.object),
    curated: React.PropTypes.arrayOf(React.PropTypes.object),
    intro: React.PropTypes.string,
    text: React.PropTypes.string.isRequired,
    tags: React.PropTypes.array
  },

  render() {
    return (
      <div {...classes()}>
        <Container>
          <ArticleHeader {...this.props.header} />
          {this.props.mainAsset ? <MainAsset asset={this.props.mainAsset} /> : ''}
          <ArticleBody {...this.props.body} />
        </Container>
      </div>
    );
  }

});
