// components/article-detail/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import ArticleHeader from 'components/article-header';
import MainAsset from 'components/main-asset';
import Intro from 'components/intro';
import ArticleBody from 'components/article-body';
import Tags from 'components/tags';
import Comments from 'components/comments';

const classes = new bem({
  name: 'article-content',
  prefix: ''
});

export default React.createClass({

  displayName: 'ArticleDetail',

  propTypes: {
    mainAsset: React.PropTypes.object,
    enrichments: React.PropTypes.arrayOf(React.PropTypes.object),
    curated: React.PropTypes.arrayOf(React.PropTypes.object),
    intro: React.PropTypes.string,
    text: React.PropTypes.string.isRequired,
    tags: React.PropTypes.array
  },

  render() {
    return (
      <div {...classes()}>
        <ArticleHeader {...this.props} />
        {this.props.mainAsset ? <MainAsset asset={this.props.mainAsset} /> : ''}
        {this.props.intro ? <Intro text={this.props.intro} /> : ''}
        <ArticleBody text={this.props.text} enrichments={this.props.enrichments} curated={this.props.curated} />
        {this.props.tags ? <Tags items={this.props.tags} /> : ''}
        <Comments />
      </div>
    );
  }

});
