// components/article-detail/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import ArticleBody from 'components/article-body';
import Tags from 'components/tags';
import Comments from 'components/comments';
import SocialButtons from 'components/social-buttons';

const classes = new bem({
  name: 'article-detail',
  prefix: ''
});

export default React.createClass({

  displayName: 'ArticleDetail',

  propTypes: {
    body: React.PropTypes.object.isRequired,
    tags: React.PropTypes.arrayOf(React.PropTypes.string),
    adKeywords: React.PropTypes.arrayOf(React.PropTypes.string),
    socialButtons: React.PropTypes.object.isRequired,
    commentable: React.PropTypes.bool.isRequired
  },

  render() {

    var { body, tags, adKeywords, socialButtons, commentable } = this.props;

    return (
      <div {...classes()} data-ad-keywords={tags.concat(adKeywords)}>
        <ArticleBody {...body} />
        {tags
          ? <Tags items={tags} />
        :null}
        <SocialButtons {...socialButtons} orientation='horizontal' />
        {commentable
          ? <Comments />
        :null}
      </div>

    );

  }

});
