// components/article-detail/slot.js
var _ = require('lodash');

module.exports = function(state) {

  // helper function
  var isOfType = function(_asset, contentType) {
    return _asset.asset && _asset.asset.content_type === contentType;
  };

  var getMainAsset = function(assets) {

    // check if there is an image asset
    var imageAsset = _.find(assets, function(_asset) {
      return isOfType(_asset, 'asset_image');
    });

    if (imageAsset) {
      return imageAsset;
    }

    // No image asset, but maybe there is an image_gallery. If so, use the
    // first image from the gallery.
    var galleryAsset = _.find(assets, function(_asset) {
      return isOfType(_asset, 'asset_image_gallery');
    });

    if (galleryAsset) {
      // get the first image
      var asset = galleryAsset.asset &&
                  galleryAsset.asset.images &&
                  galleryAsset.asset.images[0];

      return asset;
    }

    return null;
  };

  var parsedData = function(article) {
    var source = article.source.title || (article.source.source ? article.source.source.title : null);

    return {
      header: {
        catchword: article.context_label,
        title: article.title,
        author: article.author || source,
        source: article.author && article.author !== source ? source : null,
        published: article.dc.effective,
        modified: article.dc.modified
      },
      mainAsset: getMainAsset(article.assets),
      body: {
        author: article.author || source,
        enrichments: article.stoerer,
        curated: article.curated,
        intro: article.lead,
        text: article.text,
        tags: article.keywords,
        title: article.title
      }
    };
  };

  return parsedData(state.article);

};
