// components/article-detail/wrapper.jsx
import React from 'react';
import _ from 'lodash';

export default Component => React.createClass({

  displayName: 'ArticleDetailWrapper',

  propTypes: {
    article: React.PropTypes.object.isRequired
  },

  getMainAsset(assets) {
    var isOfType = function(_asset, contentType) {  // helper function
      return _asset.asset && _asset.asset.content_type === contentType;
    };

    // check if there is an image asset
    var imageAsset = _.find(assets, function(_asset) {
      return isOfType(_asset, 'asset_image');
    });

    if (imageAsset) {
      return imageAsset;
    }

    // No image asset, but maybe there is an image_gallery. If so, use the
    // first image from the gallery.
    var galleryAsset = _.find(assets, function(_asset) {
      return isOfType(_asset, 'asset_image_gallery');
    });

    if (galleryAsset) {
      // get the first image
      var asset = galleryAsset.asset &&
                  galleryAsset.asset.images &&
                  galleryAsset.asset.images[0];

      return asset;
    }

    return null;
  },

  mappedData(article) {
    var source = article.source.title || (article.source.source ? article.source.source.title : null);

    return {
      header: {
        catchword: article.context_label,
        title: article.title,
        author: article.author || source,
        source: article.author && article.author !== source ? source : null,
        published: article.dc.effective,
        modified: article.dc.modified
      },
      mainAsset: this.getMainAsset(article.assets),
      enrichments: article.stoerer,
      curated: article.curated,
      intro: article.lead,
      text: article.text,
      tags: article.keywords
    };
  },

  render() {
    return (
      <Component {...this.mappedData(this.props.article)} />
    );
  }

});
