// components/meta-nav/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Button from 'components/button';
import SearchInput from 'components/search-input';

const classes = new bem({
  name: 'meta-nav',
  prefix: ''
});

export default React.createClass({

  displayName: 'MetaNav',

  componentDidMount: function() {
    this.setState({ browser: true });
  },

  getInitialState: () => ({
    browser: false
  }),

  render() {
    let loginURL = '/anmelden';

    if (this.state.browser) {
      loginURL += '?redirectTo=' + encodeURIComponent(document.location.href);
    }

    return (
      <div {...classes()}>
        <div {...classes('container')}>
          <ul {...classes('nav')}>
            <li {...classes('nav-item')}><a {...classes('item')} href="/login">Abo</a></li>
            <li {...classes('nav-item')}><a {...classes('item')} href="/login">E-Paper</a></li>
            <li {...classes('nav-item')}><a {...classes('item')} href="/login">Werbung</a></li>
            <li {...classes('nav-item')}><a {...classes('item')} href="/login">Todesanzeigen</a></li>
          </ul>
          <SearchInput />
          <Button icon='user' url={loginURL} label="Anmelden" modifier="meta" attributes={{'title': 'Anmelden'}} />
        </div>
      </div>
    );
  }

});
