// clientside/recommendations-top.jsx
export default function(params) {

  var options = params || {
    breakpoint: 1150,
    selector: '.recommendations-top',
    buttonClass: 'recommendations-top__button',
    buttonText: 'Mehr Artikel',
    collapsedClass: 'recommendations-top--collapsed',
    expandedClass: 'recommendations-top--expanded'
  };

  var active = false;
  var panelEl = document.querySelector(options.selector), buttonEl;

  var getWindowWidth = function() {
    return Math.max(document.documentElement.clientWidth, window.innerWidth);
  };

  var activateToggling = function() {
    if (!active) {
      injectButton();
      panelEl.classList.add(options.collapsedClass);
      active = true;
    }
  };

  var destroyToggling = function() {
    if (buttonEl) {
      removeButton();
    }
    if (active) {
      panelEl.classList.remove(options.expandedClass);
      panelEl.classList.remove(options.collapsedClass);
      active = false;
    }
  };

  var injectButton = function() {
    var toggleButton = document.createElement('button');
    toggleButton.type = 'button';
    toggleButton.innerHTML = options.buttonText;
    toggleButton.classList.add(options.buttonClass);
    buttonEl = toggleButton;
    panelEl.appendChild(toggleButton);
    toggleButton.addEventListener('click', openArticles);
  };

  var removeButton = function() {
    panelEl.removeChild(buttonEl);
    buttonEl = null;
  };

  var openArticles = function() {
    panelEl.classList.remove(options.collapsedClass);
    removeButton(panelEl.querySelector('.' + options.buttonClass));
  };

  var initLogic = function() {
    if (getWindowWidth() <= options.breakpoint) {
      activateToggling(panelEl);
    } else {
      destroyToggling(panelEl);
    }
  };

  var init = function() {
    if (panelEl) {
      initLogic();
      window.addEventListener('resize', initLogic);
    }
  };

  return Object.freeze({
    init: init
  });
}
