// mixins/auto-play-video.jsx
import jQuery from 'jQuery';
import objectAssign from 'react/lib/Object.assign';

export default {

  isMobile: function() {
    if ( /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent) ) {
      return true;
    }
    return false;
  },
  defineIsOnScreen: function() {
    // define jQuery plugin used for checking if an DOM node is on screen
    jQuery.fn.isOnScreen = function() {
      var win = jQuery(window);

      var viewport = {
        top: win.scrollTop(),
        left: win.scrollLeft()
      };
      viewport.right = viewport.left + win.width();
      viewport.bottom = viewport.top + win.height();

      var bounds = this.offset();
      bounds.right = bounds.left + this.outerWidth();
      bounds.bottom = bounds.top + this.outerHeight();

      return (!(viewport.right < bounds.left || viewport.left > bounds.right || viewport.bottom < bounds.top || viewport.top > bounds.bottom));

    };
  },

  // state
  getInitialState: function() {
    return {
      onScreen: false,
      inReplacementMode: true,
      inThirdQuartile: false,
      playerEventsToWatch: ['doPlay', 'doStop', 'doPause', 'doSeek', 'changeVolume', 'firstPlay', 'playerPlayEnd', 'thirdQuartile', 'playerPlayed']
    };
  },

  setupPlayerReplacement: function() {

    // define on screen handling function jQuery.fn.isOnScreen
    this.defineIsOnScreen();

    // setup players with ads
    this.setupPlayerWithAds();

    // start playing video in non mobile browsers if autoPlay feature is activated. The video has to be started even before the user scrolls or interacts with the page.
    if (this.props.autoPlay && !this.isMobile()  && this.onScreenChangeHandlingOfPlayerWithoutAds) {
      this.onScreenChangeHandlingOfPlayerWithoutAds();
    }

    // watch onScreen changes (player with ads goes/leaves visible screen area f.ex. when scrolling down a page)
    this.setupOnScreenChangeHandlingOfPlayerWithoutAds();
  },

  setupOnScreenChangeHandlingOfPlayerWithoutAds: function() {
    var $ = jQuery;
    if (typeof(window)!=='undefined') {
      // listen if video goes on screen to automatically play it if autoPlay feature is activated and the browser is not mobile.
      if (this.props.autoPlay && !this.isMobile()) {
        $(window).on('DOMContentLoaded load resize scroll', this.onScreenChangeHandlingOfPlayerWithoutAds);
      }
    }
  },

  onScreenChangeHandlingOfPlayerWithoutAds: function() {
    // handle player: start auto-playing when visible
    var onScreenVisibilityChanged = function() {
      var kdp = document.getElementById(this.props.kalturaID);
      if (!this.state.onScreen) {
        // console.log(this.props.kalturaID.toUpperCase(),' - onScreenVisibilityChanged - ',':	became visible - trigger doPlay');
        kdp.sendNotification('doPlay');
      } else {
        // console.log(this.props.kalturaID.toUpperCase(),' - onScreenVisibilityChanged - ',':	was hidden');
      }
    };

    var onScreenCurrently = jQuery('#'+this.props.kalturaID).isOnScreen();
    if (onScreenCurrently!==this.state.onScreen) {
      onScreenVisibilityChanged.apply(this);
    }
    this.setState({
      onScreen: onScreenCurrently
    });
  },

  setupPlayerWithAds: function() {
    var self = this;

    var setupPlayerWithAds = function(i) {

      // listen for events to setup player with ads
      var j = self.state.playerEventsToWatch[i];
      var kdp = document.getElementById(self.props.kalturaID);
      var triggerPlayerReplacement = function() {

        if (self.state.inReplacementMode) {
          kWidget.destroy(self.props.kalturaID);
          kWidget.embed(self.state.playerWithAdsConfig);
        }

        // track NetMetrix on player replacement (new player starts auto-playing)
        // console.log('@TODO: NetMetrix tracking on player replacement');
        jQuery.get(self.props.trackingURL);

      };

      // trigger player replacement on pause and seek events
      kdp.kBind(j, function() {
        if (j === 'doPause' || j === 'doSeek') {
          triggerPlayerReplacement();
        }
      });


      // Temporarily deactivate player replacement by setting state (replacement mode & inThirdQuartile)
      //
      // Problem: Player replacement is triggered by events doPause (clicking on view area or pause button) and doSeek (click on the progess bar).
      // Once the playhead reaches the end unfortunately the first event that is triggered is doPause.
      // This means the doPause event triggers player replacement without user interaction which is unacceptable.
      //
      // Solution (hacky): Remove player replacement on doPause event at the latest time possible which is on the thirdQuartile event (when fourth quartile is reached).
      // Additionally, setup player replacement on doPlay event once the playhead reaches the end and playerPlayEnd event is fired. This makes sure once the video finishes playing clicking on the view area or the play button does trigger player replacement.
      //
      // Limitations: Temporarily removing player replacement on doPause makes it impossible for the user to restart the video unmuted by clicking on the view are or the pause button (both trigger doPause events) in the fourth quartile.
      if (j==='thirdQuartile') {

        // deactivate replacement mode on doPause and activate it on doSeek for the last quartile
        kdp.kBind('thirdQuartile', function() {
          // console.log(self.props.kalturaID.toUpperCase()+':	'+'No Ads: thirdQuartile');

          // Switch to non replacement mode. Make sure component ist NOT updated on changing the state. We will handle player replacements outside of react.
          this.setState({
            inReplacementMode: false, // reliable - use instead of unbinding doPause event which is NOT reliable: kdp.kUnbind('doPause');
            inThirdQuartile: true
          });
        });

        // activate player replacement on doPlay when end of video is reached
        kdp.kBind('playerPlayEnd', function() {
          // console.log(self.props.kalturaID.toUpperCase()+':	', 'PLAYERPLAYEND: ', 'setup triggering of replay in new player on doPlay event.');

          kdp.kBind('doPlay', function() {
            // console.log(self.props.kalturaID.toUpperCase()+':	'+'doPlay fired after thirdQuartile:');

            // Switch to replacement mode. Make sure component ist NOT updated on changing the state. We will trigger replacement manually below.
            this.setState({
              inReplacementMode: true, // reliable - use instead of unbinding doPause event which is NOT reliable: kdp.kUnbind('doPause');
              inThirdQuartile: false
            });
          });
        });
      }
    };

    // setup players with ads
    jQuery.each(this.state.playerEventsToWatch, setupPlayerWithAds);
  },

  // react lifecycle
  shouldComponentUpdate: function() {
    return false;
  }

};
