// mixins/responsive-image.jsx
/*eslint-disable no-console */
'use strict';

const CryptoJS = require('crypto-js');
const path = require('path');
const isRetina = require('is-retina');

const imageSalt = 'FyfgR5fadpR7S9w';
const screenFactor = 320;
const fallback = 'medium';
const retinaMax = 1280;

var viewportWidth = 0;

// Requested sizes i.e. <ResponsiveImage size="small" />
const sizes = {
  micro: 80,
  tiny: 160,
  small: 320,
  medium: 640,
  large: 960,
  huge: 1280,
  full: 1600,
  wide: 1920
};

// Screen size -> size mapping
const screens = {
  // 0 -> 320
  tiny: {
    micro: 'micro',
    tiny: 'tiny', small: 'tiny', medium: 'tiny',
    large: 'small', huge: 'small', full: 'small'
  },
  // 320 -> 640
  small: {
    micro: 'micro',
    tiny: 'tiny',
    small: 'small', medium: 'small',
    large: 'medium', huge: 'medium', full: 'medium'
  },
  // 640 -> 960
  medium: {
    micro: 'micro',
    tiny: 'tiny',
    small: 'small',
    medium: 'medium',
    large: 'large', huge: 'large', full: 'large'
  },
  // 960 -> 1280
  large: {
    micro: 'micro',
    tiny: 'tiny',
    small: 'small',
    medium: 'medium',
    large: 'large',
    huge: 'huge', full: 'huge'
  },
  // 1280 -> 1600
  huge: {
    micro: 'micro',
    tiny: 'tiny',
    small: 'small',
    medium: 'medium',
    large: 'huge',
    huge: 'full', full: 'full'
  },
  // 1600+
  wide: {
    micro: 'micro',
    tiny: 'tiny',
    small: 'small',
    medium: 'medium',
    large: 'huge',
    huge: 'wide', full: 'wide'
  }
};


const ResponsiveImageMixin = {

  urlEncode(str) {
    return str.replace(/\=$/, '').replace(/\+/g, '-').replace(/\//g, '_').replace(/\=+$/, '');
  },

  createImageSignature(width, height) {
    var size = '/'.concat(width, 'x', height);
    var signature = CryptoJS.HmacSHA1(size, imageSalt);
    var encodedSignature = this.urlEncode(CryptoJS.enc.Base64.stringify(signature));
    return path.join(encodedSignature, size);
  },

  createImageSrc(params) {
    return params.src + '/' + this.createImageSignature(params.width, params.height);
  },

  getScreenSize() {
    var screenSizes = Object.keys(screens);
    // No JS / server.js
    if ( ! this.state.support) {
      return fallback;
    }
    // Responsive Image
    viewportWidth = viewportWidth || Math.max(document.documentElement.clientWidth, window.innerWidth) || screenFactor;
    var current = Math.min(Math.ceil(viewportWidth/screenFactor) - 1, screenSizes.length - 1);
    return screenSizes[current];
  },

  sizeOrFallback(size) {
    if ( ! sizes[size]) {
      console.warn('ResponsiveImage size "' + size + '" not found, using fallback');
      // Return default
      return fallback;
    }
    return size;
  },

  getImageVariation(size) {
    return screens[this.getScreenSize()][this.sizeOrFallback(size)];
  },

  imageDimensionsForSize(size, ratio) {
    var imageRatio = ratio || '3x2';
    var imageWidth = sizes[size];
    var factors = imageRatio.split('x');
    return {
      width: imageWidth,
      height: Math.ceil((factors[1] / factors[0]) * imageWidth)
    };
  },

  responsiveImageSrc(params) {
    var dimensions = this.imageDimensionsForSize(this.getImageVariation(params.size), params.ratio);
    // Double output size on retina (but not the largest images)
    if (isRetina() && this.state.support && dimensions.width <= retinaMax) {
      dimensions.width = dimensions.width * 2;
      dimensions.height = dimensions.height * 2;
    }
    return this.createImageSrc({
      src: params.src,
      width: dimensions.width,
      height: dimensions.height
    });
  },

  getImg() {
    return this.refs.img.getDOMNode();
  },

  hasLoaded() {
    this.getImg().removeEventListener('load', this.hasLoaded);
    this.setState({ loading: false });
  },

  lazyLoadSrc() {
    var src = this.getImg().getAttribute('data-src');
    this.setState({ lazySrc: src });
  },

  testLoaded() {
    if (this.state.loading && this.state.support) {
      this.getImg().addEventListener('load', this.hasLoaded);
    }
  }

};

module.exports = ResponsiveImageMixin;
