// mixins/tracking.jsx
/* eslint-disable no-console */
import _ from 'lodash';
import Tracker from 'tracking/tracker';

const PAGEVIEW = 1;
const EVENT = 2;

export default {

  PAGEVIEW: PAGEVIEW,
  EVENT: EVENT,

  /**
   * Track certain events
   * @param data Object  Configuration object for the call.
   *                        Should contain: event, id, index
   */
  track(type, data) {
    let path,
        defaultData,
        extendedData;

    path = window.location.pathname;

    // remove slashes at start and end
    path = path.replace(/^[\/]*/, '').replace(/[\/]*$/, '');

    defaultData = {
      domain: 'aznetz',
      path: {
        product: 'live',
        sitename: this.props.site,
        view: this.constructor.displayName.toLowerCase(),
        path: path
      },
      screenWidth: screen.width,
      screenHeight: screen.height
    };
    extendedData = _.extend(defaultData, {path: _.extend(defaultData.path, data)});

    // do the actual tracking
    if (type === this.EVENT) {
      Tracker.trackEvent(extendedData);
    }

    if (type === this.PAGEVIEW) {
      Tracker.trackPageView(extendedData);
    }

  }

};
