// mixins/tracking.jsx
/* eslint-disable no-console */
import _ from 'lodash';
import Tracker from 'tracking/tracker';

export default {

  /**
   * Track certain events
   * @param options Object  Configuration object for the call.
   *                        Should contain: event, id, index
   */
  track(options) {
    let path,
        defaultOptions,
        extendedOptions;

    path = window.location.pathname;

    // remove slashes at start and end
    path = path.replace(/^[\/]*/, '').replace(/[\/]*$/, '');

    defaultOptions = {
      domain: 'aznetz',
      path: {
        product: 'live',
        sitename: this.props.site,
        view: this.constructor.displayName.toLowerCase(),
        path: path
      },
      screenWidth: screen.width,
      screenHeight: screen.height
    };
    extendedOptions = _.extend(defaultOptions, {path: _.extend(defaultOptions.path, options)});

    let tracker = new Tracker(extendedOptions);

    // track Net Metrix
    let tag = tracker.track();

    console.info('Net Metrix Tag (currently inactive: ', tag);
  }

};
