// mixins/tracking.jsx
/* eslint-disable no-console */
import _ from 'lodash';
import Tracker from 'tracking/tracker';

const PAGEVIEW = 1;
const EVENT = 2;

/**
 * Mixin for tracking.
 *
 * @module mixins/tracking
 */
module.exports = {

  /**
   * Constant for specifying the type "pageview".
   *
   * @type  {integer}
   */
  PAGEVIEW: PAGEVIEW,

  /**
   * Constant for specifying the type "event".
   *
   * @type  {integer}
   */
  EVENT: EVENT,

  /**
   * Track certain events
   *
   * @param type {string}   Type of tracking, either `PAGEVIEW` or `EVENT`.
   * @param data {Object}   Configuration object for the call.
   *                      Required fields: event, id, paging
   *
   * @example
   * // import mixin
   * import Tracking from 'mixins/tracking';
   *
   * const VideoLibrary = React.createClass({
   *
   *   mixins: [Tracking],
   *
   *   componentDidMount() {
   *     this.track(Tracking.EVENT, {
   *       event: 'selectVideo',
   *       id: '1_sdfsdfs'
   *     });
   *   }
   *
   *   // Further component logic goes here
   *
   * });
   */
  track: function(type, data) {
    let defaultData,
        processedData;

    defaultData = {
      domain: 'aznetz',
      path: {
        product: 'live',
        sitename: this.props.site,
        view: this.constructor.displayName.toLowerCase(),
        path: window.location.pathname
      },
      referer: document.referer,
      screenWidth: screen.width,
      screenHeight: screen.height
    };
    processedData = _.extend(defaultData, {path: _.extend(defaultData.path, data)});

    // do the actual tracking
    if (type === this.EVENT) {
      Tracker.trackEvent(processedData);
    }

    if (type === this.PAGEVIEW) {
      Tracker.trackPageView(processedData);
    }

  }

};
