// components/header-nav/index.jsx
import React from 'react';

export default React.createClass({

  displayName: 'HeaderNav',

  propTypes: {
    isClosed: React.PropTypes.bool,
    id: React.PropTypes.number
  },

  render() {
    const { isClosed, id } = this.props;

    const domainVariations = [
      {
        href: 'http://aargauerzeitung.ch',
        label: 'Aargau'
      },
      {
        href: 'http://bzbasel.ch',
        label: 'Basel'
      },
      {
        href: 'http://basellandschaftlichezeitung.ch',
        label: 'Basel'
      },
      {
        href: 'oltnertagblatt.ch',
        label: 'Solothurn'
      },
      {
        href: 'http://grenchnertagblatt.ch',
        label: 'Solothurn'
      },
      {
        href: 'http://solothurnerzeitung.ch ',
        label: 'Solothurn'
      },
      {
        href: 'http://limmattalerzeitung.ch',
        label: 'Limmattal'
      },
      {
        href: 'http://badenertagblatt.ch ',
        label: 'Aargau'
      }
    ];

    var navItems = [
      {
        href: '/',
        label: 'Start'
      },
      {
        href: '/',
        label: 'Aargau'
      },
      {
        href: '/',
        label: 'Schweiz'
      },
      {
        href: '/',
        label: 'Ausland'
      },
      {
        href: '/',
        label: 'Wirtschaft'
      },
      {
        href: '/',
        label: 'Sport'
      },
      {
        href: '/',
        label: 'Panorama'
      },
      {
        href: '/',
        label: 'Kultur'
      },
      {
        href: '/',
        label: 'Blaulicht'
      }
    ];

    if (id && domainVariations.length <= id) {
      navItems.splice(1, 0,
        {
          href: domainVariations[id].href,
          label: domainVariations[id].label
        }
      );
    }

    return (
      <div className="header-nav">
        <nav role='navigation'>
          <ul>
            <li>
              <a
                href="/login"
                className='header-nav__link header-nav__link--login'>
                Login
              </a>
            </li>
            {navItems.map((navItem, index) =>
              <li key={index}>
                <a
                  href={navItem.href}
                  className='header-nav__link'>
                  {navItem.label}
                </a>
              </li>
            )}
          </ul>
        </nav>
      </div>
    );
  }

});
