// components/video-teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'video-teaser',
  prefix: ''
});

export default React.createClass({

  displayName: 'VideoTeaser',

  propTypes: {
    thumbnailUrl: React.PropTypes.string.isRequired,
    name: React.PropTypes.string.isRequired,
    viewer: React.PropTypes.object.isRequired,
    videoID: React.PropTypes.string.isRequired,
    expanded: React.PropTypes.bool.isRequired
  },

  handleClick() {
    var videoID = this.props.videoID;
    // callback to handle state in a parent component
    this.props.onToggleVideo(videoID);
  },

  render() {
    var viewer;

    if (this.props.viewer) {
      viewer = this.props.viewer;
    }

    // @TODO:@WithRichard:expanded: refactor to bundle expander's state in a component or mixin?
    var classNames = classes();
    if (this.props.expanded) {
      classNames.className = classNames.className + " is-expanded";
    }

    return (
      <div {...classNames}>
        <figure onClick={this.handleClick} {...classes('container')} >
          <img {...classes('image')} src={this.props.thumbnailUrl}/>
          <figcaption>
            <p {...classes('title')}>{this.props.name}</p>
          </figcaption>
        </figure>
        {viewer}
      </div>
    );
  }

});
