// components/video-teaser/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';

import ResponsiveVideoImage from 'components/responsive-video-image';
import PlayTime from 'components/play-time';

const classes = new BEM({
  name: 'video-teaser',
  prefix: ''
});

export default React.createClass({

  displayName: 'VideoTeaser',

  propTypes: {
    thumbnailUrl: React.PropTypes.string.isRequired,
    name: React.PropTypes.string.isRequired,
    video: React.PropTypes.object.isRequired,
    maxLength: React.PropTypes.number.isRequired
  },

  handleClick() {
    var videoID = this.props.video.id;
    // callback to handle state in a parent component
    this.props.onToggleVideo(videoID);
  },

  trimTitle(string, length) {
    if (string.length <= length) {
      return string;
    }
    var trim = string.substr(0, length);
    return trim.replace(/ [^ ]*$/m, ' ...');
  },

  // pass on properties to children by cloning react child element
  renderChildren() {
    return React.Children.map(this.props.children, (child) => {
      return React.cloneElement(child, {
        onSetExpanderHeight: this.props.onSetExpanderHeight
      });
    });
  },

  render() {
    var accountID = String(this.props.video.partnerId);

    return (
      <div {...classes()}>
        <figure
          {...classes('item')}
          onClick={this.handleClick}
        >
          <div {...classes('gradient')}>
            <ResponsiveVideoImage
              src={this.props.thumbnailUrl}
              title={this.props.video.name}
              alt={this.props.video.name}
              accountID={accountID}
              videoID={this.props.video.id}
              ssl={false}
            />
            <PlayTime duration={this.props.video.duration} />
          </div>
          <figcaption {...classes('meta')}>
            <p {...classes('title')}>
              {this.trimTitle(this.props.name, this.props.maxLength)}
            </p>
          </figcaption>
        </figure>
        {this.renderChildren()}
      </div>
    );
  }
});
