// components/video-library/api.js
/* eslint-disable no-console */
var loadDomain = require('../../../../server/api/loadDomain');
var KalturaAPI = require('helpers/kaltura-api.jsx');
var videoLibraryMockData = require('components/video-library/.data.json');
// var _ = require('lodash');

module.exports = function(req, res) {
  console.info('============api.js=============');

  // load domain data
  loadDomain(req, res);


  // @TEMP:@TODO: get mock data first and overwrite later on
  req.api._data.videos = videoLibraryMockData;

  // tell the api that on more async request goes in line
  req.api._startApiCall();

  var kaltura = new KalturaAPI({
    ssl: false,
    domain: 'www.aargauerzeitung.ch',
    path: '/__kaltura_api_proxy__',
    service: 'baseEntry',
    action: 'list',
    format: 1,
    orderByKey: 'createdAt',
    orderDescending: true,
    pageIndex: 1,
    pageSize: 12,
    filters: {
      mediaTypeEqual: 1,
      freeText: '',
      categoriesIdsMatchAnd: 23690341
    }
  });

  // load videos using promises
  kaltura.getData().then(function(response) {
    req.api._data.videos.initialVideos = response.data.objects;

    // tell the api that one request has been successfully completed
    req.api._endApiCall();
  }).catch(function() {
    console.error('KalturaAPI Error: Could not load videos from Kaltura');
  });


  // signal that we need to wait for the API to finish the request
  return true;
};
