// components/video-library/api.js
/* eslint-disable no-console */
var KalturaAPI = require('helpers/video/kaltura-api.jsx');
var extractSource = require('helpers/video/extract-source.jsx');

var loadVideos = function(req) {
  var kaltura = new KalturaAPI({
    ssl: false,
    domain: req.api._data.domain.data.properties.domain_name,
    path: '/__kaltura_api_proxy__',
    service: 'baseEntry',
    action: 'list',
    format: 1,
    orderByKey: 'createdAt',
    orderDescending: true,
    pageIndex: 1,
    pageSize: 12,
    filters: {
      mediaTypeEqual: 1,
      freeText: '',
      categoriesIdsMatchAnd: req.api._data.domain.data.properties.kaltura_mediathek_category_id
    }
  });

  req.api.startExternalApiCall('videos');

  kaltura.getData().then(function(response) {
    var videos = {};
    videos.initialVideos = response.data.objects;
    videos.initialTotalCount = response.data.totalCount;
    if (req.params.variation && req.params.variation !== 'all') {
      videos.initialSelectedVideo = req.params.variation;
    }
    // tell the api that one request has been successfully completed
    req.api.endExternalApiCall('videos', videos);
  }).catch(function() {
    console.error('KalturaAPI Error: Could not load videos from Kaltura: ', kaltura.getURL(), ' in ', __filename);

    // tell the api that one request has been successfully completed
    req.api.endExternalApiCall();
  });
};

var loadSource = function(req) {

  // load metadata and extract source
  if (req.params.variation && req.params.variation !== 'all') {

    // tell the api that one more async request goes in line
    req.api.startExternalApiCall('initialVideoSource');

    var kaltura = new KalturaAPI({
      ssl: false,
      domain: req.api._data.domain.data.properties.domain_name,
      path: '/__kaltura_api_proxy__',
      service: 'metadata_metadata',
      action: 'list',
      format: 1,
      filters: {
        objectIdEqual: req.params.variation
      }
    });

    kaltura.getData().then(function(response) {
      // tell the api that the request has been completed
      req.api.endExternalApiCall('initialVideoSource',
                                 extractSource(response));
    }).catch(function(error) {
      console.error('KalturaAPI Error: Could not load source from Kaltura: ', kaltura.getURL(), ' in ', __filename);

      // tell the api that one request has beend completed
      req.api.endExternalApiCall();
    });
  }
};

module.exports = function(req) {
  loadVideos(req);
  loadSource(req);
  // signal that we need to wait for the API to finish the request
  return true;
};
