// components/video-library/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Container from 'components/container';
import VideoLibraryHeader from 'components/video-library-header';
import VideoLibraryList from 'components/video-library-list';
import { getScript } from 'helpers';
import KalturaAPIProxy from 'higher-order/kaltura-api-proxy';

var libraryIncluded = false;

const classes = new bem({
  name: 'video-library',
  prefix: ''
});

const VideoLibrary = React.createClass({

  displayName: 'VideoLibrary',

  statics: {
    loadKalturaClient: function() {
      if (libraryIncluded) {
        return;
      }

      // @TODO: maybe use promise based script loader later on
      getScript('/client/scripts/vendor/KalturaClient.min.js', () => {
        libraryIncluded = true;
      });
    },
    // loading statuses
    IS_LOADED: 0,
    IS_LOADING: 1,
    LOADING_FAILED: 2
  },

  propTypes: {
    initialVideos: React.PropTypes.arrayOf(React.PropTypes.object),
    categoryId: React.PropTypes.string.isRequired,
    pageSize: React.PropTypes.number.isRequired
  },

  getInitialState() {
    var initialVideos = this.props.initialVideos || [];
    return {
      searchText: '',
      selectedVideo: null,
      videos: initialVideos,
      loadingStatus: VideoLibrary.IS_LOADED,
      skipGetVideos: true,
      pageIndex: 1
    };
  },

  // alter states
  changeSearchText() {
    this.setState({searchText: event.target.value});
  },

  selectVideo(videoID) {
    this.setState({selectedVideo: videoID});
  },

  deselectVideo() {
    this.setState({selectedVideo: null});
  },

  toggleVideo(videoID) {
    if (this.state.selectedVideo === videoID) {
      this.deselectVideo();
    } else {
      this.selectVideo(videoID);
    }
  },

  // removes unwanted elements at the end of the array and adds new items to array at the end
  addVideoPage(videos) {
    var updatedVideos = this.state.videos;
    var start = (this.state.pageIndex-1) * this.props.pageSize;
    updatedVideos = updatedVideos.slice(0, start);
    updatedVideos = updatedVideos.concat(videos);

    return updatedVideos;
  },

  setVideos(videos) {
    var updatedVideos = this.addVideoPage(videos);
    this.setState({
      videos: updatedVideos,
      loadingStatus: VideoLibrary.IS_LOADED,
      skipGetVideos: true
    });
  },

  search() {
    this.setState({
      skipGetVideos: false,
      pageIndex: 1
    });
  },

  componentDidMount() {
    VideoLibrary.loadKalturaClient();
  },

  increasePageIndex() {
    var nextIndex = this.state.pageIndex + 1;
    this.setState({
      pageIndex: nextIndex,
      skipGetVideos: false
    });
  },

  render() {
    var list, header, proxy;

    list = (
      <VideoLibraryList
        videos={this.state.videos}
        selectedVideo={this.state.selectedVideo}
        playerID={this.props.playerID}
        accountID={this.props.accountID}
        onToggleVideo={this.toggleVideo}
        onNextPage={this.increasePageIndex}
        loadingStatus={this.state.loadingStatus}
      />
    );
    header = (
      <VideoLibraryHeader
        searchText={this.state.searchText}
        changeSearchText={this.changeSearchText}
        onSearch={this.search}
      />
    );
    proxy = (
      <KalturaAPIProxy
        format={1}
        pageSize={25}
        mediaTypeEqual={1}
        component={list}
        onSetVideos={this.setVideos}
        // Do not load videos using KalturaAPIProxy when initial videos are set on mounting of the embedding component
        skipGetVideos={this.state.skipGetVideos}
        searchText={this.state.searchText}
        pageIndex={this.state.pageIndex}
        videosPerPage={this.props.pageSize}
      />
    );

    return (
      <div {...classes()}>
        <Container>
          {header}

          {proxy}
        </Container>
      </div>
    );
  }

});

export default VideoLibrary;
