// components/video-library/index.jsx
/* global KalturaConfiguration: true, KalturaBaseEntryFilter: true, KalturaClient: true */
import React from 'react';

import Container from 'components/container';
import VideoLibraryHeader from 'components/video-library-header';
import VideoLibraryList from 'components/video-library-list';

export default React.createClass({

  displayName: 'VideoLibrary',

  propTypes: {
    initialVideos: React.PropTypes.arrayOf(React.PropTypes.object)
  },

  getInitialState() {
    var initialVideos = this.props.initialVideos || [];
    return {
      searchText: '',
      selectedVideo: "1_yrqmhyoi",
      kaltura: {
        categoryId: '21939571',
        ks: 'djJ8MTcxOTIyMXxUD5myU_CM4vbQHy--gzSyyWqJ8-vaY_zS3WlG54U4VlLu9ckhw0B0kMVmq5EQycOYvPh-gjT4MPylan6bke5igQAn_PJ8CWBYmH9n_5f7CVGfFTqNJye5l3I6128OfWU='
      },
      videos: initialVideos
    };
  },

  // alter states
  changeSearchText() {
    this.setState({searchText: event.target.value});
  },

  selectVideo(videoID) {
    // clicking an already expanded teaser makes him collapse
    var selectedVideo = this.state.selectedVideo===videoID ? null : videoID;
    this.setState({selectedVideo: selectedVideo});
  },

  getKalturaSession() {
    /*var _self = this;
    var setKS = function(data) {
      _self.setState({
          kaltura: {
            ks: data[0]
          }
      });
    };

    jQuery.getJSON('http://yoleidoo.com/az/kaltura-code-samples/az/player%20embedding/getKS.php', function( data ) {
      setKS(data);  //implement this on your production server, make sure you use a user with restricted rights
                    //this link is for development purposed only. The KS generated grants admin rights.
    });*/
  },

  getKalturaVideosByNameAndCategory() {
    var _self = this;
    var getKalturaVideos = function () {
      // get videos using Kaltura Client Library API
      var partnerId =1719221;
      var config = new KalturaConfiguration(partnerId);
      config.serviceUrl = "http://www.kaltura.com/";
      var client = new KalturaClient(config);
      client.ks = _self.state.kaltura.ks;
      var filter = new KalturaBaseEntryFilter();
      filter.categoriesIdsMatchOr = _self.state.kaltura.categoryId;//21939571: ZüriNews
      //filter.categoryAncestorIdIn = _self.state.kaltura.categoryId;//21939571: ZüriNews
      filter.freeText = _self.state.searchText+'*';
      filter.orderBy = '-createdAt';//KalturaBaseEntryOrderBy.CREATED_AT_ASC;
      //var filter.typeEqual = KalturaEntryType.MEDIA_CLIP;
      var pager = null;

      // load media entries
      var mediaDataLoaded = function (success, results) {
        if (!success) {
          //console.log('ERROR: '+results);
        }

        if (results.code && results.message) {
          //console.log(results.objects);
          //console.log('ERROR: '+results.message);
        }

        // set loaded data to component's state
        if (_self.isMounted()) {
          _self.setState({
            videos: results.objects
          });
        }
      };
      client.baseEntry.listAction(mediaDataLoaded, filter, pager);

    };
    getKalturaVideos();
  },

  search() {
    if (this.state.kaltura.ks) {
      this.setState({
        videos: null
      });//remove videos to render "loading spinner"
      this.getKalturaVideosByNameAndCategory();
    }
  },

  componentWillMount() {
    if (!this.state.kaltura.ks) {
      this.getKalturaSession();
    }
  },

  componentWillUpdate() {
    //console.log('componentWillUpdate: nextState: ', nextState);
  },

  render() {
    return (
      <Container>
        <script src="/client/scripts/vendor/KalturaClient.min.js"></script>
        <VideoLibraryHeader
          searchText={this.state.searchText}
          changeSearchText={this.changeSearchText}
        /><a onClick={this.search}>suchen</a>

        <VideoLibraryList
          videos={this.state.videos}
          selectedVideo={this.state.selectedVideo}
          playerID={this.props.playerID}
          accountID={this.props.accountID}
          selectVideo={this.selectVideo}/>
      </Container>
    );
  }

});
