// components/video-library/slot.js
/* eslint-disable no-console */
var _ = require('lodash');

var convertAPIToPropsSchema = function(state, data) {
  var mappedData;

  // map settings to props component requires
  mappedData = _.assign({}, data, {
    site: data._settings.domain.data.id,

    // @TODO: needs to be constructed from site (aaz), env (dev or prod)
    // dev: aaz-azdev.lovelysystems.com, live: www.aargauerzeitung.ch
    videoAPIHost: 'aaz-azdev.lovelysystems.com',

    // @TODO: needs to be in domain configuration -> deprecated: process.env.KALTURA_ACCOUNT_ID,
    accountID: '1789881',
    playerID: data._settings.domain.data.properties.kaltura_mediathek_video_player_id,
    categoryId: data._settings.domain.data.properties.kaltura_mediathek_category_id,
    pageSize: 12,
    initialTotalCount: 1000
  });
  delete mappedData._settings;

  return mappedData;
};

var parsedData = function(state) {
  var data = _.assign({}, state.api._data.videos, {_settings: {domain: state.api._data.domain}});

  // map data from lovely api
  data = convertAPIToPropsSchema(state, data);

  return data;
};

module.exports = {
  iso: true,
  data: parsedData
};
