// styleguide/tabs/index.jsx
'use strict';

const React = require('react');
const Tab = require('styleguide/tab');
const bem = require('react-bem-helper');

const classes = new bem({
  name: 'tabs',
  prefix: 'styleguide-'
});

const Tabs = React.createClass({

  propTypes: {
    sections: React.PropTypes.oneOfType([
      React.PropTypes.array,
      React.PropTypes.object
    ]).isRequired
  },

  getInitialState() {
    return {
      visibleIndex: 0
    };
  },

  showSection(index) {
    this.setState({ visibleIndex: index });
  },

  render() {

    var sections = this.props.sections;
    var visibleIndex =this.state.visibleIndex;

    var activeClass = function(index) {
      return {
        'is-active': this.state.visibleIndex === index
      };
    };

    return (
      <section {...classes()}>
        <nav {...classes('nav')}>
          <ul {...classes('menu')}>
            {sections.map((section, index) =>
              <li key={index} {...classes('menu__item', null, activeClass.call(this, index))}>
                <button {...classes('button')} onClick={this.showSection.bind(this, index)}>
                  {section.title}
                </button>
              </li>
            )}
          </ul>
        </nav>
        <Tab>
          {sections[visibleIndex].content}
        </Tab>
      </section>
    );

  }

});

module.exports = Tabs;
