// styleguide/tabs/index.jsx
'use strict';

const React = require('react');
const Tab = require('styleguide/tab');
const bem = require('react-bem-helper');

const classes = new bem({
  name: 'tabs',
  prefix: 'styleguide-'
});

const Tabs = React.createClass({

  // propTypes: {
  //   sections: React.PropTypes.oneOfType([
  //     React.PropTypes.array,
  //     React.PropTypes.object
  //   ])
  // },

  getInitialState() {
    return {
      visibleIndex: 0
    };
  },

  showSection(index) {
    this.setState({ visibleIndex: index });
  },

  render() {

    var sections = this.props.sections;
    var visibleIndex = this.state.visibleIndex;

    var activeClass = function(index) {
      return {
        'is-active': this.state.visibleIndex === index
      };
    };

    return (
      <section {...classes()}>
        <ul {...classes('menu')} role='tablist'>
          {sections.map((section, index) =>
            <li key={index} {...classes('item', null, activeClass.call(this, index))} role='presentation'>
              <button {...classes('button')} onClick={this.showSection.bind(this, index)} role="tab">
                {section.title}
              </button>
            </li>
          )}
        </ul>
        <Tab>
          {sections[visibleIndex].content}
        </Tab>
      </section>
    );

  }

});

module.exports = Tabs;
