// styleguide/tabs/index.jsx
import React from 'react';
import Tab from 'styleguide/tab';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'tabs',
  prefix: 'styleguide-'
});

export default React.createClass({

  displayName: 'Tabs',

  propTypes: {
    sections: React.PropTypes.array
  },

  getInitialState() {
    return {
      visibleIndex: 0
    };
  },

  showSection(index) {
    this.setState({ visibleIndex: index });
  },

  render() {
    var sections = this.props.sections;
    var visibleIndex = this.state.visibleIndex;

    var activeClass = function(index) {
      return {
        'is-active': this.state.visibleIndex === index
      };
    };

    return (
      <section {...classes()}>
        <ul {...classes('menu')} role='tablist'>
          {sections.map((section, index) =>
            <li key={index} {...classes('item', null, activeClass.call(this, index))} role='presentation'>
              <button {...classes('button')} onClick={this.showSection.bind(this, index)} role="tab">
                {section.title}
              </button>
            </li>
          )}
        </ul>
        <Tab>
          {sections[visibleIndex].content}
        </Tab>
      </section>
    );
  }

});
