// components/social-buttons/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Button from 'components/button';

const classes = new bem({
  name: 'social-buttons',
  prefix: ''
});

export default React.createClass({

  url: null,

  // Breakpoint under which we consider the device a mobile
  breakpoint: 640,

  propTypes: {
    title: React.PropTypes.string,
    description: React.PropTypes.string
  },

  getViewportWidth() {
    return Math.max(document.documentElement.clientWidth, window.innerWidth);
  },

  isMobile() {
    return this.getViewportWidth() < this.breakpoint
      && (
        ('ontouchstart' in window)
        || (navigator.msMaxTouchPoints > 0)
        || (navigator.maxTouchPoints > 0)
      );
  },

  componentDidMount() {
    this.url = window.location.href;
    this.setState({
      support: true,
      mobile: this.isMobile()
    });
  },

  getInitialState() {
    return {
      support: false
    };
  },

  print() {
    window.print();
  },

  getTwitterTarget() {
    return 'http://www.twitter.com/share'
      + '?url=' + this.url
      + '&text=' + this.props.title;
  },

  getFacebookTarget() {
    return 'http://www.facebook.com/sharer/sharer.php'
      + '?u=' + this.url;
  },

  getWhatsappTarget() {
    return 'whatsapp://send'
      + '?text=' + this.url;
  },

  getMailTarget() {
    var target = 'mailto:?subject=__SUBJECT__&body=__BODY__'
      .replace('__SUBJECT__', this.props.title)
      .replace('__BODY__', this.props.description);

    if (this.url) {
      target += ' ' + this.url;
    }

    return target;
  },

  render() {

    var attributes = {
      target: '_blank'
    };

    var whatsappAttributes = {
      target: '_blank',
      'data-action': 'share/whatsapp/share'
    };

    return (
      <div {...classes()}>
        {this.url
          ? <Button icon='facebook' _target={this.getFacebookTarget()} attributes={attributes} />
          : ''}

        {this.url
          ? <Button icon='twitter' _target={this.getTwitterTarget()} attributes={attributes} />
          : ''}

        {this.props.title && this.props.description
          ? <Button icon='mail' _target={this.getMailTarget()} />
          : ''}

        {this.state.support
          ? <Button icon='print' _target={this.print} />
          : ''}

        {this.url && this.state.mobile
          ? <Button icon='whatsapp' _target={this.getWhatsappTarget()} attributes={whatsappAttributes} />
          : ''}
      </div>
    );

  }

});
