// components/article-body/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';
import Intro from 'components/intro';
import Tags from 'components/tags';
import Comments from 'components/comments';
import SocialButtons from 'components/social-buttons';

const classes = new bem({
  name: 'article-body',
  prefix: ''
});

const ArticleBody = React.createClass({

  displayName: 'ArticleBody',

  propTypes: {
    content: React.PropTypes.array.isRequired,
    intro: React.PropTypes.string,
    tags: React.PropTypes.arrayOf(React.PropTypes.string),
    title: React.PropTypes.string
  },

  render() {
    return (
      <div itemProp="articleBody" {...classes()}>
        <SocialButtons title={this.props.title} description={this.props.intro} orientation='vertical' modifier='article' />
        {this.props.intro ? <Intro text={this.props.intro} /> : ''}
        {this.props.content}
        {this.props.tags ? <Tags items={this.props.tags} /> : ''}
        <Comments />
      </div>
    );
  }

});

export default Wrapper(ArticleBody);
