// components/article-body/index.jsx
import React from 'react';
import _ from 'lodash';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';
import RawContent from 'components/raw-content';
import GetEnrichment from 'higher-order/get-enrichment';
import EnrichmentWrap from 'components/enrichment';

const classes = new bem({
  name: 'article-body',
  prefix: ''
});

const ArticleBody = React.createClass({

  displayName: 'ArticleBody',

  propTypes: {
    content: React.PropTypes.array.isRequired
  },

  disallowedEnrichments: ['survey_stoerer', 'map_stoerer'],

  getComponent(item, index) {
    const type = item.type;
    const data = item.data;
    const floatedEnrichments = ['left', 'right'];

    if (type === 'content_block') {
      return <RawContent key={index} content={data} />;
    }

    if (!_.includes(this.disallowedEnrichments, type)) {
      let Enrichment = <GetEnrichment key={index} enrichment={data} />;
      let enrichmentPositon = _.includes(floatedEnrichments, data.placement) ? true : false;
      return (
        <EnrichmentWrap
          key={index}
          type={type}
          aside={enrichmentPositon}>
          {Enrichment}
        </EnrichmentWrap>
      );
    }
  },

  mapContentToComponents(content) {
    let components;
    if (content) {
      components = content.map((item, index) => {
        return this.getComponent(item, index);
      });
    }
    return components;
  },

  render() {
    return (
      <div itemProp="articleBody" {...classes()}>
        {this.mapContentToComponents(this.props.content)}
      </div>
    );
  }

});

export default Wrapper(ArticleBody);
