// components/article-body/wrapper.jsx
import React from 'react';

import CuratedRecommendations from 'components/curated-recommendations';

export default Component => React.createClass({

  displayName: 'ArticleBodyWrapper',

  propTypes: {
    author: React.PropTypes.string,
    text: React.PropTypes.string.isRequired,
    enrichments: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    related: React.PropTypes.arrayOf(React.PropTypes.object)
  },

  markers: {
    enrichment: '${{stoerer}}',
    related: '${{empfehlung}}'
  },

  injectRelated(props) {
    let curatedRecommendations = React.createElement(CuratedRecommendations, {
      author: props.author,
      teasers: props.related
    });
    let curatedRecommendationsString = React.renderToString(curatedRecommendations);
    return props.text.replace(this.markers.related, curatedRecommendationsString);
  },

  mappedData(props) {
    return {
      content: this.injectRelated(props)
    };
  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
