// components/expandable/index.jsx
import React from 'react/addons';

export default React.createClass({

  displayName: 'Expandable',

  propTypes: {
    expandableHeight: React.PropTypes.number.isRequired
  },

  componentDidUpdate() {
    this.resetHeight();
  },

  resetHeight() {
    var expandable = React.findDOMNode(this.refs.expandable);
    if (!this.props.expanded) {
      // reset to start value
      expandable.style.height = '';
    }
  },

  setExpanderHeight(height) {
    var expandable = React.findDOMNode(this.refs.expandable);
    if (this.props.expanded) {
      expandable.style.height = this.props.expandableHeight + height + 'px';
    }
  },

  // pass on properties to children by cloning react child element
  renderChildren() {
    return React.Children.map(this.props.children, (child) => {
      return React.addons.cloneWithProps(child, {
        onSetExpanderHeight: this.setExpanderHeight
      });
    });
  },

  render() {
    var isExpanded = this.props.expanded ? ' is-expanded' : '';
    var className = 'expandable' + isExpanded;

    return (
      <div className={className} ref="expandable" {...this.props}>
        {this.renderChildren()}
      </div>
    );
  }

});
