// styleguide/viewer/index.jsx
'use strict';

const React = require('react');
const ReactToJsx = require('react-to-jsx');
const Highlight = require('react-highlight');
const bem = require('react-bem-helper');

const Tabs = require('styleguide/tabs');
const IFrame = require('styleguide/iframe');
const Loading = require('styleguide/loading');

const classes = new bem({
  name: 'viewer',
  prefix: 'styleguide-'
});

const Viewer = React.createClass({

  propTypes: {
    component: React.PropTypes.object.isRequired,
    variation: React.PropTypes.object.isRequired,
    children: React.PropTypes.element.isRequired
  },

  removeTabs(markup) {
    var tab = RegExp("\\t", "g");
    return markup.replace(tab, '  ');
  },

  render() {

    var sections = [
      {
        title: 'Preview',
        content: <IFrame {...this.props} />
      },
      {
        title: 'JSX',
        content: <Loading><Highlight className='html'>{this.removeTabs(ReactToJsx(this.props.children))}</Highlight></Loading>
      },
      {
        title: 'HTML',
        content: <Loading><Highlight className='html'>{React.renderToStaticMarkup(this.props.children)}</Highlight></Loading>
      },
    ];

    return (
      <div {...classes()}>
        <Tabs sections={sections} />
      </div>
    );

  }

});

module.exports = Viewer;
