// components/teaser-byline/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import DateFormatter from 'components/date-formatter';
import Icon from 'components/icon';

const classes = new bem({
  name: 'teaser-byline',
  prefix: ''
});

export default React.createClass({

  displayName: 'TeaserByline',

  propTypes: {
    text: React.PropTypes.string,
    dateFormat: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.bool
    ]),
    modifiers: React.PropTypes.array,
    suffix: React.PropTypes.string,
    icon: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.bool
    ])
  },

  render() {
    const { text, modifiers, dateFormat, suffix, icon } = this.props;

    if (!dateFormat && !text) {
      return null;
    }

    return (
      <p {...classes(null, modifiers)}>
        {icon
          ? <Icon {...classes('icon')} id={icon} />
        :null}
        {dateFormat
          ? <DateFormatter publicationDate={text} format={dateFormat} />
          : text
        }
        {suffix
          ? <span {...classes('suffix')}>{suffix}</span>
        :null}
      </p>
    );
  }
});
