// components/ressort-header/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'ressort-header',
  prefix: ''
});

export default React.createClass({

  displayName: 'RessortHeader',

  propTypes: {
    hasNav: React.PropTypes.bool.isRequired,
    title: React.PropTypes.string,
    urlpart: React.PropTypes.string,
    subnav: React.PropTypes.arrayOf(React.PropTypes.object)
  },

  subRessortSlug(subressort) {
    return '/' + this.props.urlpart + '/' + subressort;
  },

  render() {

    if (!this.props.hasNav) {
      return (<div/>);
    }

    var { title, subnav } = this.props;

    return (
      <div {...classes()}>
        <h1 {...classes('title')}>{title}</h1>
        <ul {...classes('subnav')}>
          {subnav.map((item, index) => {
            return (
              <li {...classes('subnav-item')} key={index}>
                <a href={this.subRessortSlug(item.urlpart)} key={index}>
                  {item.title}
                </a>
              </li>
            );
          })}
        </ul>
      </div>
    );
  }

});
