// components/ressort-header/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { capitalizeAll } from 'helpers';

const classes = new bem({
  name: 'ressort-header',
  prefix: ''
});

export default React.createClass({

  displayName: 'RessortHeader',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    ressort: React.PropTypes.object,
    subressortSlug: React.PropTypes.string,
    navigation: React.PropTypes.array.isRequired
  },

  subressortSlug(subressort) {
    return '/' + this.props.ressort.urlpart + '/' + subressort;
  },

  render() {
    const { title, navigation, subressortSlug } = this.props;
    return (
      <div {...classes()}>
        <h1 {...classes('title')}>{capitalizeAll(title)}</h1>
        {navigation.length
          ? <ul {...classes('subnav')}>
              {navigation.map((item, index) => {
                return (
                  <li {...classes('subnav-item')} key={index}>
                    <a {...classes('subnav-link', null, [item.slug === subressortSlug ? 'is-current' : null])}
                      href={this.subressortSlug(item.slug)}
                      key={index}>
                      {capitalizeAll(item.title)}
                    </a>
                  </li>
                );
              })}
            </ul>
        :null}
      </div>
    );
  }

});
