// components/ressort-header/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { capitalizeAll } from 'helpers';

const classes = new bem({
  name: 'ressort-header',
  prefix: ''
});

export default React.createClass({

  displayName: 'RessortHeader',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    urlpart: React.PropTypes.string,
    subnav: React.PropTypes.array.isRequired
  },

  subRessortSlug(subressort) {
    return '/' + this.props.urlpart + '/' + subressort;
  },

  render() {
    const { title, subnav } = this.props;
    return (
      <div {...classes()}>
        <h1 {...classes('title')}>{capitalizeAll(title)}</h1>
        {subnav.length
          ? <ul {...classes('subnav')}>
              {subnav.map((item, index) => {
                return (
                  <li {...classes('subnav-item')} key={index}>
                    <a href={this.subRessortSlug(item.urlpart)} key={index}>
                      {capitalizeAll(item.title)}
                    </a>
                  </li>
                );
              })}
            </ul>
        :null}
      </div>
    );
  }

});
