// tracking/net-metrix-tracker.jsx
/* eslint-disable no-console */
import _ from 'lodash';
import deepForEach from 'helpers/deep-foreach';
import QS from 'qs';
import config from 'tracking/config.json';

export default class NetMetrixTracker {

  static removeEmpty(data) {
    let clone = _.cloneDeep(data);

    return deepForEach(clone, function(e, key, obj) {
      if (e === null || e === '' || typeof(e) === 'undefined' ) {
        delete obj[key];
      }
    });
  }

  static getOrderedPath(data) {
    let path = [
      data.path.product,
      data.path.sitename,
      data.path.view,
      data.path.path,
      data.path.event,
      data.path.id,
      data.path.index
    ];

    return _.remove(path, _.isNotUndefined);
  }

  static sanitizePath(path) {
    return path.replace(/^[\/]*/, '').replace(/[\/]*$/, '');
  }

  static buildPath(data) {
    let path = '';

    if (data.length && data.length > 0) {
      path = data.join('/');
    }

    return path;
  }

  static buildQS(data) {
    let params = {
      r: (data.referer ? `${encodeURIComponent(data.referer)}&` : ``),
      d: (new Date()).getTime(),
      x: `${data.screenWidth}x${data.screenHeight}`
    };

    return QS.stringify(params);
  }

  static buildTag(data, path) {
    let queryString = this.buildQS(data);

    return `https://${data.domain}.wemfbox.ch/cgi-bin/ivw/CP/${path}?${queryString}`;
  }

  static getTag(data) {
    let path,
        cleaned;

    // remove empty properties from data.path
    cleaned = this.removeEmpty(data);

    // remove slahes from  data.path.path at start and end
    cleaned.path.path = this.sanitizePath(data.path.path);

    // order path properties
    cleaned.path = this.getOrderedPath(cleaned);

    //  get path as string
    path = this.buildPath(cleaned.path);

    return this.buildTag(cleaned, path);
  }

  static requestTag(tag) {
    let image = new Image();

    image.src = tag;
  }

  static log(tag) {
    console.info(`Detected Net Metrix tag not fired. Current host "${window.location.host}" is not configured as live host: `, tag);
  }

  static isLiveHost() {
    let liveHosts = [];
    let isLive = false;

    // throw error if live hosts are not specified
    if (typeof this.config === 'undefined' || !this.config || !this.config.netMetrix.liveHosts) {
      console.error('You must specify hosts enabled for tracking under netMetrix.liveHosts[] in tracking/config.json in order to track Net Metrix Audit.');
    } else {
      liveHosts = this.config.netMetrix.liveHosts;
    }

    // check if we are on a live host
    _.forEach(liveHosts, function(e) {
      if (window && window.location.host.search(e) > -1) {
        isLive = true;
      }
    });

    return isLive;
  }

  static track(data) {
    let tag = this.getTag(data);

    // Only request tag if we are live. Otherwise log to console
    if (this.isLiveHost()) {
      this.requestTag(tag);
    } else {
      this.log(tag);
    }
  }

  static trackEvent(data) {
    this.track(data);
  }

  /**
   * @description Tracks Net Metrix or logs tracking to `console` for non-live hosts.
   *
   * Example:
   * ```
   * {
   *   domain: 'aznetz',
   *   path: {
   *     product: 'live',
   *     sitename: 'styleguide',
   *     view: 'page',
   *     path: window.location.pathname,
   *     event: 'pageview'
   *   }
   * }
   * ```
   *
   *
   * @param  {Object}  data  Configuration data.
   *
   * @namespace data
   * @property  {string}    domain          - The identifier for the Audit. For "Nordwestschweiz Netz" this is `'aznetz'`.
   * @property  {object}    path            - The path information object.
   * @property  {string}    path.product    - The product. Either `live`, `'epaper'` or `'service'`.
   * @property  {string}    path.sitename   - The newsportal. One of: `'aaz'`, `'bat'`, `'bzb'`, `'blz'`, `'gtb'`, `'liz'`, `'ot'`, `'soz'`.
   * @property  {string}    path.view       - The view. `page` for `path.event=='pageview'` or `'[component name]'` for event tracking.
   * @property  {string}    path.event      - The user generated event. `'pageview'` for page view tracking or `'[name of event]'`.
   * @property  {string}    [path.id]       - The ID of the component or the page.
   * @property  {string}    [path.index]    - The index of the request. Example: For the image gallery this is the index of the image requested.
   * @property  {string}    [referer]       - The referer URL.
   * @property  {integer}   [screenWidth]   - The width of the devices' screen in pixels.
   * @property  {integer}   [screenHeight]  - The height of the devices' screen in pixels.
   *
   *
   * @return  {Void}
   */
  static trackPageView(data) {
    this.track(data);
  }

}

NetMetrixTracker.config = config;
