// components/article-byline/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import DateFormatter from 'components/date-formatter';
import Schema from 'components/schema';

const classes = new bem({
  name: 'article-byline',
  prefix: ''
});

export default React.createClass({

  displayName: 'ArticleByline',

  propTypes: {
    author: React.PropTypes.string,
    source: React.PropTypes.string,
    date: React.PropTypes.object,
    schema: React.PropTypes.object
  },

  render() {
    const { source, author, date, schema } = this.props;

    return (
      <div {...classes()}>
        {author || source
          ? (
              <p itemProp='author' {...classes('author')}>
                {author ? 'von ' + author : null}
                {source && author ? ' — ' : null}
                {source ? source : null}
              </p>
            )
        :null}
        <p {...classes('date')}>
          <Schema items={schema} />
          <DateFormatter {...date} />
        </p>
      </div>
    );
  }

});
