// inputs/wrapper/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem('input');

export default React.createClass({
  displayName: 'TextInput',

  propTypes: {
    label: React.PropTypes.string,
    id: React.PropTypes.string.isRequired,
    className: React.PropTypes.string,
    description: React.PropTypes.string,
    hint: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.element
    ])
  },

  render() {
    var className = [this.props.className, classes('wrapper').className].join(' ');

    return (
      <div className={className}>
        { this.props.label
          ? <label {...classes('label')} htmlFor={this.props.id}>{this.props.label}</label>
        : null }
        { this.props.description
          ? <p {...classes('description')}>{this.props.description}</p>
        : null }
        { this.props.children }
        { this.props.hint
          ? <div {...classes('hint')}>{this.props.hint}</div>
        : null }
      </div>
    )
  }
});
