// inputs/text-area/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem('input');

export default React.createClass({
  displayName: 'TextArea',

  propTypes: {
    label: React.PropTypes.string,
    className: React.PropTypes.string,
    name: React.PropTypes.string.isRequired,
    onChange: React.PropTypes.function,
    placeholder: React.PropTypes.string,
    id: React.PropTypes.string,
    value: React.PropTypes.string,
    rows: React.PropTypes.number
  },

  getDefaultProps() {
    return {
      rows: 4
    };
  },

  getInitialState() {
    return {
      value: this.props.value
    }
  },

  onValueChange: function() {
    this.setState({ value: this.refs.input.getDOMNode().value })

    if (this.props.onChange) {
      this.props.onChange(this.props.name, this.props.name);
    }
  },

  render() {
    var { label, name, placeholder, rows } = this.props;
    var id = this.props.id || this.props.name;

    return (
      <div className={this.props.className}>
        { label ? <label {...classes('label')} htmlFor={id}>{label}</label> : null }
        <textarea {...classes()} id={id} ref='input' rows={rows} name={name} placeholder={placeholder} onChange={this.onValueChange} value={this.state.value} />
      </div>
    )
  }
});
