// inputs/text-area/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import InputWrapper from 'inputs/wrapper';

const classes = new bem('input');

export default React.createClass({
  displayName: 'TextArea',

  propTypes: {
    label: React.PropTypes.string,
    className: React.PropTypes.string,
    name: React.PropTypes.string.isRequired,
    onChange: React.PropTypes.func,
    placeholder: React.PropTypes.string,
    id: React.PropTypes.string,
    value: React.PropTypes.string,
    rows: React.PropTypes.number
  },

  getDefaultProps() {
    return {
      rows: 4
    };
  },

  componentWillReceiveProps: function(nextProps) {
    if (this.state.value !== nextProps.value) {
      this.setState({ value: nextProps.value })
    }
  },

  getInitialState() {
    return {
      value: this.props.value
    }
  },

  onValueChange: function() {
    var value = this.refs.input.getDOMNode().value;
    this.setState({ value: value })

    if (this.props.onChange) {
      this.props.onChange(this.props.name, value);
    }
  },

  render() {
    var id = this.props.id || this.props.name;

    return (
      <InputWrapper {...this.props} id={id}>
        <textarea {...this.props} {...classes()} id={id} ref='input' onChange={this.onValueChange} value={this.state.value} />
      </InputWrapper>
    )
  }
});
