// components/weather-detail/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';

import Section from 'components/section';
import SectionHeader from 'components/section-header';
import Container from 'components/container';
import WeatherForecast from 'components/weather-forecast';

const classes = new BEM({
  name: 'weather-detail',
  prefix: ''
});

export default React.createClass({

  displayName: 'WeatherDetail',

  propTypes: {
    weatherIcon: React.PropTypes.string.isRequired,
    minTemperature: React.PropTypes.number.isRequired,
    maxTemperature: React.PropTypes.number.isRequired,
    weatherText: React.PropTypes.string.isRequired
  },

  render() {

    return (
      <Section modifier='default'>
        <Container>
          <SectionHeader title='Gemeinde ändern:' />
          <div {...classes()}>
            <div  {...classes('data')}>
              <WeatherForecast
                weatherIcon={this.props.weatherIcon}
                minTemperature={this.props.minTemperature}
                maxTemperature={this.props.maxTemperature}
              />
            </div>
            <div {...classes('text')}>{this.props.weatherText}</div>
          </div>
        </Container>
      </Section>
    );
  }
});
