// components/weather-detail/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';
import Dropdown from 'react-dropdown';

import Section from 'components/section';
import SectionHeader from 'components/section-header';
import Container from 'components/container';
import WeatherForecast from 'components/weather-forecast';

const classes = new BEM({
  name: 'weather-detail',
  prefix: ''
});

export default React.createClass({

  displayName: 'WeatherDetail',

  propTypes: {
    // weatherIcon: React.PropTypes.string.isRequired,
    // minTemperature: React.PropTypes.number.isRequired,
    // maxTemperature: React.PropTypes.number.isRequired,
    // weatherText: React.PropTypes.string.isRequired
  },

  getInitialState() {
    return this.state = {
      parish: this.props.parishes[0].name
    };
  },

  select(event) {
    this.setState({parish: event.target.value});
  },

  getInitialState() {
    return this.state = {
      selected: { value: 'aarau', label: 'Aarau' }
    };
  },

  _onSelect(option) {
    this.setState({selected: option});
  },

  render() {
    const options = [
      { value: 'aarau', label: 'Aarau' },
      { value: 'brugg', label: 'Brugg' },
      { value: 'frick', label: 'Frick' }
    ];

    let defaultOption = this.state.selected;

    let options = this.props.parishes.map((item, index) => {
      return <option value={item.name} key={index}>{item.name}</option>;
    });

    return (

      <Section modifier='default'>
        <Container>
          <SectionHeader title='Gemeinde ändern:' />
          <select
            {...classes('dropdown')}
            onChange={this.select}
            value={this.state.parish}
          >
            {options}
          </select>
          <div {...classes()}>
            <div  {...classes('data')}>
              <WeatherForecast
                weatherIcon={this.props.aarau.weatherIcon}
                minTemperature={this.props.aarau.minTemperature}
                maxTemperature={this.props.aarau.maxTemperature}
              />
            </div>
            <div {...classes('text')}>{this.props.aarau.weatherText}</div>
          </div>
        </Container>
      </Section>
    );
  }
});
