// styleguide/sidebar/index.jsx
import React from 'react';
import _ from 'lodash';

import SidebarSection from 'styleguide/sidebar-section';
import Logo from 'styleguide/logo';

export default React.createClass({

  displayName: 'Sidebar',

  propTypes: {
    components: React.PropTypes.array.isRequired,
    active: React.PropTypes.string
  },

  uniqueSections(components) {
    var uniqComponents = _.uniq(components, 'category');
    return uniqComponents.map(function(component) {
      return component.category;
    });
  },

  render() {
    var components = this.props.components;
    var active = this.props.active;
    var sections = this.uniqueSections(components);

    var navigation = (
      <nav role="navigation" className="sidebar-navigation">
        {sections.map((section, index) => {
          var sectionComponents = _.filter(components, chr => {
            return chr.category === section;
          });

          return (
            <SidebarSection title={section} components={sectionComponents} active={active} key={index} />
          );
        })}
      </nav>
    );

    return (
      <header className='styleguide-sidebar'>
        <Logo />
        {navigation}
      </header>
    );
  }
});
