// styleguide/sidebar/index.jsx
'use strict';

const React = require('react');
const Menu = require('styleguide/menu');
const Logo = require('styleguide/logo');

var Sidebar = React.createClass({

  propTypes: {
    components: React.PropTypes.array,
    active: React.PropTypes.object,
    switchComponent: React.PropTypes.func
  },

  mapComponentMenuItems(components) {
    var props = this.props;
    return components.map(function(item) {
      var listItem = {};
      listItem.title = item.title;
      listItem.link = '/styleguide/component/' + item.slug;
      listItem.slug = item.slug;
      if (props.active) {
        listItem.active = item.slug === props.active.slug;
      }
      return listItem;
    });
  },

  render() {

    return (
      <header className='styleguide-sidebar'>
        <Logo />
        <nav role="navigation">
          <Menu
            items={this.mapComponentMenuItems(this.props.components)}
            switchComponent={this.props.switchComponent}
          />
        </nav>
      </header>
    );
  }

});

module.exports = Sidebar;
