// components/tags/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import path from 'path';
import { capitalizeAll } from 'helpers/string';

const classes = new bem({
  name: 'tags',
  prefix: ''
});

export default React.createClass({

  displayName: 'Themen',

  propTypes: {
    items: React.PropTypes.arrayOf(React.PropTypes.string).isRequired
  },

  render() {

    var { items } = this.props;

    if (!items.length) {
      return null;
    }

    return (
      <div {...classes()}>
        <p {...classes('title')}>
          {items.length === 1
            ? 'Verwandtes Thema:'
            : 'Verwandte Themen:'
          }
        </p>
        <ul {...classes('list')}>
        {items.map(function(keyword, index) {
          return (
            <li key={index} {...classes('item')}>
              <a href={path.join('/thema', keyword)} {...classes('link')}>
                {capitalizeAll(keyword)}
              </a>
            </li>
          );
        })}
        </ul>
      </div>
    );
  }

});
