// components/enrichment-gallery/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Swipeable from 'react-swipeable';
import tapEventPlugin from 'react-tap-event-plugin';

import Wrapper from './wrapper';
import GalleryMixin from 'mixins/gallery';
import ResponsiveImage from 'components/responsive-image';
import Caption from 'components/caption';
import Icon from 'components/icon';

tapEventPlugin();

const classes = new bem({
  name: 'gallery',
  prefix: ''
});

const Gallery = React.createClass({

  mixins: [GalleryMixin],

  propTypes: {
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    images: React.PropTypes.array.isRequired
  },

  getInitialState() {
    return {
      currentIndex: 0,
      fullscreen: false
    };
  },

  handlePrev() {
    this.navigateSlider(-1);
  },

  handleNext() {
    this.navigateSlider(+1);
  },

  render() {
    return (
      <Swipeable onSwipedLeft={this.handlePrev} onSwipedRight={this.handleNext}>
        <div {...classes()}>
          <figure itemScope itemType='http://schema.org/ImageObject' {...classes('image')}>
            <div {...classes('inner')}>
              <ResponsiveImage
                ratio='16x9'
                size='max'
                title={this.imageData('title')}
                alt={this.imageData('description')}
                src={this.imageData('url')} />

              <button type='button' onTouchTap={this.handlePrev} {...classes('button', 'prev')}>
                <Icon id='prev' />
              </button>
              <button type='button' onTouchTap={this.handleNext} {...classes('button', 'next')}>
                <Icon id='next' />
              </button>
            </div>
            <div {...classes('slide-number')}>
              {this.state.currentIndex + 1} / {this.props.images.length}
            </div>
            <Caption
              title={this.imageData('title')}
              description={this.imageData('description')}
              source={this.imageData('source')} />
          </figure>
        </div>
      </Swipeable>
    );
  }

});

export default Wrapper(Gallery);
