// components/article-comments/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Button from 'components/button';

const classes = new bem({
  name: 'article-comments',
  prefix: ''
});

const DISQUS_SHORTNAME = 'aargauerzeitung';

export default React.createClass({

  getInitialState() {
    return {
      browser: false
    };
  },

  componentDidMount() {
    this.setState({ browser: true });
  },

  loadDisqus() {
    if (this.state.browser) {
      var node = this.refs.comments.getDOMNode();
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = '//' + DISQUS_SHORTNAME + '.disqus.com/embed.js';
      node.appendChild(dsq);
      this.setState({ loaded: true });
    }
  },

  render() {

    var display = this.state.loaded
      ? <div id="disqus_thread"></div>
      : <Button type='button' onClick={this.loadDisqus.bind(null, this)} {...classes('btn')} label='Load comments (auf Deutsch)' />;

    return (
      <section ref='comments' {...classes()}>
        {display}
      </section>
    );
  }

});
