// components/enrichment-document/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Wrapper from './wrapper';
import Icon from 'components/icon';
import filesize from 'filesize';

const classes = new bem({
  name: 'enrichment-document',
  prefix: ''
});

const EnrichmentDocument = React.createClass({

  displayName: 'EnrichmentDocument',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    path: React.PropTypes.string.isRequired,
    size: React.PropTypes.number,
    fileType: React.PropTypes.string
  },

  render() {

    var { title, path, size, fileType } = this.props;

    var filesizeString = size && size > 0
      ? filesize(size, {round: 0})
      : null;

    var fileTypeString = fileType ? fileType : null;

    return (
      <div {...classes()}>
        <div {...classes('inner')}>
          <a href={path} {...classes('link')} download>
            <Icon id="download"/>
            <span {...classes('label')}>Download</span>
            {filesizeString ?
              <span {...classes('filesize')}>{filesizeString}</span>
            : null}
          </a>
          <p {...classes('title')}>
            {title}
            {fileTypeString ?
              <span {...classes('filetype')}>{fileTypeString}</span>
            : null}
          </p>
        </div>
      </div>
    );
  }

});

export default Wrapper(EnrichmentDocument);
