// components/enrichment-document/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Wrapper from './wrapper';
import Icon from 'components/icon';
import filesize from 'filesize';

const classes = new bem({
  name: 'enrichment-document',
  prefix: ''
});

const EnrichmentDocument = React.createClass({

  displayName: 'EnrichmentDocument',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    path: React.PropTypes.string.isRequired,
    size: React.PropTypes.number
  },

  render() {

    var filesizeString = '';

    if (this.props.size && this.props.size > 0)  {
      filesizeString = filesize(this.props.size, {round: 0});
    }

    return (
      <div {...classes()}>
        <div {...classes('inner')}>
          <a href={this.props.path} {...classes('link')} download>
            <Icon id="download"/>
            <span {...classes('label')}>Download</span>
            <span {...classes('filesize')}>({filesizeString})</span>
          </a>
          <p {...classes('title')}>{this.props.title}</p>
        </div>
      </div>
    );
  }

});

export default Wrapper(EnrichmentDocument);
