// components/enrichment-document/wrapper.jsx
import React from 'react';
import { getFileExtension } from 'helpers/mime-types';

export default Component =>  React.createClass({

  displayName: 'EnrichmentDocumentWrapper',

  propTypes: {
    asset: React.PropTypes.object,
    title: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    placement: React.PropTypes.string,
    size: React.PropTypes.number,
    fileType: React.PropTypes.string
  },

  mappedData(props) {

    var asset = props.asset;
    var assetInfo = asset && asset.info ? asset.info : null;
    var fileType = props.fileType ? props.fileType : null;

    if (assetInfo && assetInfo.content_type) {
      fileType = getFileExtension(assetInfo.content_type);
    }

    return {
      title: props.title || asset.title,
      path: props.path || asset.download_url,
      size: props.size || assetInfo ? assetInfo.content_length : null,
      fileType: fileType
    };

  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
