// helpers/embed-wufoo.jsx
import _ from 'lodash';

export default function(params) {

  var defaults = {
    async: true,
    autoResize: true,
    embedJs: '/scripts/embed/form.js',
    header: 'hide',
    height: '0',
    host: 'wufoo.com',
  };

  var options = _.assign({}, defaults, params.options);

  var cb = function() {
    params.callback.apply();
  };

  var embedWuFoo = function(embedOpts, state) {
    var form;
    var rs = state ? state.readyState : ''; if (rs) if (rs !== 'complete') if (rs !== 'loaded') return;
    try {
      form = new window.WufooForm();
      form.initialize(embedOpts);
      form.display();
      var wrap = document.getElementById('wufoo-' + embedOpts.formHash);
      var iframe = wrap.querySelector('.wufoo-form-container');
      iframe.onload = cb;
    } catch (e) {
      // error
    }
  };

  var createWuFooScript = function(createOpts) {
    var s = document.createElement('script');
    s.src = '//' + createOpts.host + createOpts.embedJs;
    s.onload = s.onreadystatechange = function() {
      return embedWuFoo(createOpts, this);
    };
    var scr = document.getElementsByTagName('script')[0], par = scr.parentNode;
    par.insertBefore(s, scr);
    return scr;
  };

  var injectWuFoo = function(injectOpts) {
    if (typeof window.WufooForm !== 'function') {
      createWuFooScript(injectOpts);
    } else {
      embedWuFoo(injectOpts);
    }
  };

  return injectWuFoo(options);
}
