// helpers/get-teaser-variation-config.jsx
import _ from 'lodash';

const config = {
  'basic': {
    image: {
      size: 'small',
      ratio: '16x9'
    },
    summary: false,
    byline: {
      dateFormat: 'long'
    },
  },
  'summary': {
    image: {
      size: 'small',
      ratio: '16x9'
    },
    summary: true,
    byline: {
      dateFormat: 'long'
    }
  },
  'large': {
    image: {
      size: 'small',
      ratio: '16x9'
    },
    summary: false,
    byline: {
      dateFormat: false
    },
  },
  'inlined-left': {
    image: {
      size: 'tiny',
      ratio: '1x1'
    },
    summary: false,
    byline: {
      dateFormat: 'short'
    }
  },
  'inlined-right': {
    image: {
      size: 'tiny',
      ratio: '1x1'
    },
    summary: false,
    byline: {
      dateFormat: 'short'
    }
  },
  'mega': {
    image: {
      size: 'full',
      ratio: '9x4'
    },
    summary: false,
    byline: {
      dateFormat: false
    }
  },
  'quote': {
    image: {
      size: 'tiny',
      ratio: '1x1'
    },
    summary: false,
    byline: {
      dateFormat: false
    }
  },
  'light': {
    image: false,
    summary: false,
    byline: {
      dateFormat: 'long'
    }
  },
  'light-boxed': {
    image: false,
    summary: false,
    byline: {
      dateFormat: 'long'
    }
  },
  'listed': {
    image: false,
    summary: false,
    byline: {
      dateFormat: 'listed'
    }
  }
};

export default function(variation) {
  return _.get(config, variation);
}
