// helpers/get-image-props.jsx
import _ from 'lodash';
import getSource from 'helpers/get-source';

export default function(asset) {
  if (asset) {

    let chosenAsset = asset.asset.content_type === 'asset_image_gallery' && asset.asset.images
      ? _.first(asset.asset.images)
      : asset;

    // get the image source based on asset type
    var mainImage = {
      asset_image: chosenAsset.image_url || (chosenAsset.asset ? chosenAsset.asset.image_url : null),
      asset_image_gallery: chosenAsset.image_url || (chosenAsset.asset ? chosenAsset.asset.image_url : null),
      asset_video: chosenAsset.asset.still_image ? (chosenAsset.asset.still_image.image_url || chosenAsset.asset.still_image.asset.image_url) : null
    };

    // if we have a cover image
    if (mainImage[asset.asset.content_type]) {
      var title = chosenAsset.title || chosenAsset.asset.title;
      var alt = chosenAsset.description || chosenAsset.asset.description;
      var source = _.has(chosenAsset, 'asset.source') ? getSource(chosenAsset.asset.source) : '';

      return {
        src: mainImage[asset.asset.content_type],
        title: alt ? title.trim() : (source ? '© ' + source : ''),
        alt: alt ? alt.trim() : title.trim()
      };
    }

    return null;
  }
}
