// helpers/get-teaser-config.jsx
import _ from 'lodash';

// Map modifiers changing based on viewport
const modifiers = {
  'basic-listed': {
    mobile: 'inlined-left',
    desktop: 'basic'
  },
};

// Map image sizes based on teaser variation
const imageSizes = {
  'basic': 'small',
  'basic-listed': {
    mobile: 'tiny',
    desktop: 'small'
  },
  'large': 'large',
  'inlined-right': 'tiny',
  'inlined-left': 'tiny'
};

// Map image ratios based on teaser variation
const imageRatios = {
  'basic': '16x9',
  'basic-listed': {
    mobile: '1x1',
    desktop: '16x9'
  },
  'large': '16x9',
  'inlined-right': '1x1',
  'inlined-left': '1x1'
};

// Map fallbacks
const imageFallbacks = {
  'basic': 'light',
  'basic-listed': 'light',
  'large': 'light'
};

// Map date formats based on teaser variation
const dateFormats = {
  'basic': 'long',
  'basic-listed': {
    mobile: 'short',
    desktop: 'long'
  },
  'light': 'long',
  'light-boxed': 'long',
  'listed': 'listed',
  'inlined-right': 'short',
  'inlined-left': 'short'
};

export default function(variation, view) {
  let mode = view || 'desktop';
  let find = variation + '.' + mode;
  return {
    modifier: _.get(modifiers, find) || variation,
    imageSize: {
      size: _.get(imageSizes, find) || imageSizes[variation],
      ratio: _.get(imageRatios, find) || imageRatios[variation]
    },
    imageFallback: _.get(imageFallbacks, find) || imageFallbacks[variation],
    dateFormat: _.get(dateFormats, find) || dateFormats[variation]
  };
}
