// helpers/get-teaser-from-article.jsx
import _ from 'lodash';

import getArticleUrl from 'helpers/get-article-url';
import getArticleTitle from 'helpers/get-article-title';
import getArticleByline from 'helpers/get-article-byline';
import getArticleSchema from 'helpers/get-article-schema';
import getArticleTheme from 'helpers/get-article-theme';
import getArticleMarker from 'helpers/get-article-marker';
import getMainAsset from 'helpers/get-main-asset';
import getImageProps from 'helpers/get-image-props';
import getTeaserConfig from 'helpers/get-teaser-config';

export default function(data, view) {

  if (!data) {
    return false;
  }

  const variation = data.variation;
  const article = data.article;

  let teaserConfig = getTeaserConfig(variation);

  // handle our image
  let mainAsset = getMainAsset(_.get(article, 'assets'));
  let hasImage = _.get(teaserConfig, 'imageSize.size');
  let testImageSrc;

  if (mainAsset) {
    testImageSrc = getImageProps(mainAsset);
  }

  // set modifier
  let teaserVariation = hasImage && !testImageSrc
    ? data.fallback || teaserConfig.imageFallback
    : teaserConfig.variation;

  return {
    variation: teaserVariation,
    theme: getArticleTheme(article, variation),
    title: getArticleTitle(article),
    byline: getArticleByline(article, variation),
    catchword: article.context_label,
    link: getArticleUrl(article),
    image: hasImage
      ? getImageProps(mainAsset)
      : null,
    imageSize: hasImage
      ? teaserConfig.imageSize
      : null,
    marker: getArticleMarker(article),
    schema: getArticleSchema(article)
  };
}
