// helpers/get-teaser-from-article.jsx
import _ from 'lodash';

import getArticleUrl from 'helpers/get-article-url';
import getMainAsset from 'helpers/get-main-asset';
import getImageProps from 'helpers/get-image-props';

// Map image sizes based on teaser variation
const imageSizes = {
  'basic': 'small',
  'basic-listed': 'small',
  'large': 'large',
  'inlined-right': 'tiny',
  'inlined-left': 'tiny'
};

// Map image ratios based on teaser variation
const imageRatios = {
  'basic': '16x9',
  'basic-listed': '16x9',
  'large': '16x9',
  'inlined-right': '1x1',
  'inlined-left': '1x1'
};

// Map date formats based on teaser variation
const dateFormats = {
  'basic': 'long',
  'basic-listed': 'long',
  'light': 'long',
  'light-boxed': 'long',
  'listed': 'listed',
  'inlined-right': 'short',
  'inlined-left': 'short'
};

// Map fallbacks
const fallbacks = {
  'basic': 'light',
  'basic-listed': 'light',
  'large': 'light'
};


let imageSizeProps = function(variation) {
  return {
    size: imageSizes[variation],
    ratio: imageRatios[variation]
  };
};

export default function(data) {

  if (!data) {
    return false;
  }

  var article = data.article;
  var mainAsset = getMainAsset(_.get(article, 'assets'));
  var hasImage = imageSizes[data.variation];

  // test for image
  var testImageSrc;
  if (mainAsset) {
    testImageSrc = getImageProps(mainAsset);
  }

  return {
    modifier: hasImage && !testImageSrc
      ? data.fallback || fallbacks[data.variation]
      : data.variation,
    title: article.title,
    catchword: article.context_label,
    link: getArticleUrl(article),
    image: hasImage
      ? _.assign(getImageProps(mainAsset), imageSizeProps(data.variation))
      : null,
    dateFormat: dateFormats[data.variation] || false,
    supplementary: dateFormats[data.variation] ? article.dc.created : false,
    schema: {
      datePublished: article.dc && article.dc.created ? article.dc.created : null,
      dateModified: article.dc && article.dc.modified ? article.dc.modified : null,
      image: !hasImage && getImageProps(mainAsset) ? getImageProps(mainAsset).src : null
    }
  };
}
