// helpers/get-teaser-from-article.jsx
import _ from 'lodash';

import getArticleUrl from 'helpers/get-article-url';
import getMainAsset from 'helpers/get-main-asset';
import getImageProps from 'helpers/get-image-props';
import getTeaserConfig from 'helpers/get-teaser-config';

export default function(data) {

  if (!data) {
    return false;
  }

  var teaserConfig = getTeaserConfig(data.variation);

  var article = data.article;
  var mainAsset = getMainAsset(_.get(article, 'assets'));

  var hasImage = teaserConfig.imageSize;

  var imageSizeProps = {
    size: teaserConfig.imageSize,
    ratio: teaserConfig.imageRatio
  };

  // test for image
  var testImageSrc;
  if (mainAsset) {
    testImageSrc = getImageProps(mainAsset);
  }

  return {
    modifier: hasImage && !testImageSrc
      ? data.fallback || teaserConfig.imageFallback
      : data.variation,
    title: article.title,
    catchword: article.context_label,
    link: getArticleUrl(article),
    image: hasImage
      ? _.assign(getImageProps(mainAsset), imageSizeProps)
      : null,
    dateFormat: teaserConfig.dateFormat || false,
    supplementary: teaserConfig.dateFormat ? article.dc.created : false,
    schema: {
      datePublished: article.dc && article.dc.created ? article.dc.created : null,
      dateModified: article.dc && article.dc.modified ? article.dc.modified : null,
      image: !hasImage && getImageProps(mainAsset) ? getImageProps(mainAsset).src : null
    }
  };
}
