// helpers/get-article-theme.jsx
import _ from 'lodash';

// Teasers that are inverted
const invertedTeasers = [
  'plain-inverted',
  'mega'
];

// Map labeltype to theme name
const names = {
  live: 'live',
  video_autoplay: 'autoplay',
  meinung: 'meinung',
  publi_reportage: 'publi',
  leserbeitrag: 'ugc',
  vereinsmeldung: 'ugc',
  sportverein: 'ugc',
  gemeindebeitrag: 'ugc',
};

// Map labeltype to theme type
const types = {
  publi_reportage: 'Publireportage',
  leserbeitrag: 'Leserbeitrag',
  vereinsmeldung: 'Vereinsmeldung',
  sportverein: 'Sportverein',
  gemeindebeitrag: 'Gemeindebeitrag'
};

// Maps UCGs

export default function(article, variation) {

  let labelId = _.get(article, 'labeltype');
  // stores the id of the theme used for styling e.g. 'ugc'
  let themeName = _.get(names, labelId) || 'article';
  // stores the type of theme, e.g. 'Leserbeitrag'
  let labelType = _.includes(types, labelId) ? _.get(types, labelId) : '';
  // make theme modifier an array for easy adding to BEM helper
  let modifiers = [];
  // add theme name
  modifiers.push('theme-' + themeName);
  // add invert if needed
  if (_.includes(invertedTeasers, variation)) modifiers.push('theme-inverted');

  return {
    name: themeName,
    type: labelType,
    modifiers: modifiers
  };

}
