// helpers/get-image-props.jsx
import _ from 'lodash';
import getSource from 'helpers/get-source';

export default function(asset) {
  if (asset) {

    let contentType = _.get(asset, 'asset.content_type');

    let chosenAsset = contentType === 'asset_image_gallery' && _.has(asset, 'asset.images')
      ? _.first(asset.asset.images)
      : asset;

    // get the image source based on asset type
    let mainImage = {
      asset_image: _.get(chosenAsset, 'image_url')
        || _.get(chosenAsset, 'asset.image_url')
        || null,
      asset_image_gallery: _.get(chosenAsset, 'image_url')
        || _.get(chosenAsset, 'asset.image_url')
        || null,
      asset_video: _.get(chosenAsset, 'asset.still_image.image_url')
        || _.get(chosenAsset, 'asset.still_image.asset.image_url')
        || null
    };

    // if we have a cover image
    if (mainImage[contentType]) {
      let title = chosenAsset.title || _.get(chosenAsset, 'asset.title') || '';
      let alt = chosenAsset.description || _.get(chosenAsset, 'asset.description') || '';
      let source = _.has(chosenAsset, 'asset.source') ? getSource(chosenAsset.asset.source) : '';

      return {
        src: mainImage[asset.asset.content_type],
        title: alt ? title.trim() : (source ? '© ' + source : ''),
        alt: alt ? alt.trim() : title.trim()
      };
    }

    return null;
  }
}
