// helpers/index.jsx
import getArticleUrl from './get-article-url';
import getAssetMeta from './get-asset-meta';
import getEnrichmentPlacement from './get-enrichment-placement';
import getImageProps from './get-image-props';
import getMainAsset from './get-main-asset';
import getScript from './get-script';
import getSource from './get-source';
import getTeaserFromArticle from './get-teaser-from-article';
import getTopicTeasers from './get-topic-teasers';
import getFileExtension from './get-file-extension';
import stripMarkup from './strip-markup';
import getViewportWidth from './get-viewport-width';
import capitalize from './capitalize';

export default {
  getArticleUrl: getArticleUrl,
  getAssetMeta: getAssetMeta,
  getEnrichmentPlacement: getEnrichmentPlacement,
  getImageProps: getImageProps,
  getMainAsset: getMainAsset,
  getScript: getScript,
  getSource: getSource,
  getTeaserFromArticle: getTeaserFromArticle,
  getTopicTeasers: getTopicTeasers,
  getFileExtension: getFileExtension,
  stripMarkup: stripMarkup,
  getViewportWidth: getViewportWidth,
  capitalize: capitalize
};
