// helpers/kaltura-api.jsx
import axios from 'axios';

export default class KalturaAPI {
  constructor(options) {
    this.setConfiguration(options.domain, options.path, options.ssl);
    this.setAction(options.service, options.action);
    this.setOrderBy(options.orderByKey, options.orderDescending);
    this.setFilter(options.filters);
    this.setPager(options.pageIndex, options.pageSize);
    this.setFormat(options.format);
  }

  setConfiguration(domain, path, useSSL) {
    this.domain = domain;
    this.path = path;
    this.protocol = (useSSL ? 'https://' : 'http://');
  }

  setAction(service, action) {
    this.service = service;
    this.action = action;
  }

  setOrderBy(key, descending) {
    if (key) {
      this.orderBy = (descending ? '-' : '+') + key;
    }
  }

  setFilter(filters) {
    this.filters = filters;

    //transform filters to match kaltura's schema
    this.filters.freeText = (this.filters.freeText ? this.filters.freeText + '*' : null);
  }

  setPager(index, size) {
    if (index && size) {
      this.pageIndex = index;
      this.pageSize = size;
    }
  }

  setFormat(format) {
    // format = 1 for JSON
    this.format = format;
  }

  getURL() {
    this.url = this.protocol
                + this.domain
                + this.path
                + '?service='+this.service
                + '&action='+this.action
                + (this.orderBy ? '&filter:orderBy='+this.orderBy : '')
                + (this.format ? '&format='+this.format : '')
                + (this.pageIndex ? '&pager:pageIndex='+this.pageIndex : '')
                + (this.pageSize ? '&pager:pageSize='+this.pageSize : '');

    // add filters
    Object.keys(this.filters).forEach((key) => {
      if (this.filters[key]) {
        this.url += '&filter:' + key + '=' + this.filters[key];
      }
    });

    return this.url;
  }

  getData() {
    var url = this.getURL();
    // Helper for debugging
    // console.log('url', 'pageIndex=', this.pageIndex, url);
    return axios.get(url);
  }

}
