// helpers/kaltura-api.jsx
/*global JSON:true*/
import axios from 'axios';

export default class KalturaAPI {
  constructor(options) {
    this.setConfiguration(options.domain, options.path, options.ssl);
    this.setAction(options.service, options.action);
    this.setOrderBy(options.orderByKey, options.orderDescending);
    this.setFilter(options.freeText, options.mediaTypeEqual, options.categoriesIdsMatchAnd);
    this.setPager(options.pageIndex, options.pageSize);
    this.setFormat(options.format);
  }

  setConfiguration(domain, path, useSSL) {
    this.domain = domain;
    this.path = path;
    this.protocol = (useSSL ? 'https://' : 'http://');
  }

  setAction(service, action) {
    this.service = service;
    this.action = action;
  }

  setOrderBy(key, descending) {
    this.orderBy = (descending ? '-' : '+') + key;
  }

  setFilter(freeText, mediaTypeEqual, categoryID) {
    this.freeText = freeText + '*';
    this.mediaTypeEqual = mediaTypeEqual;
    this.categoriesIdsMatchAnd = (categoryID ? categoryID : '');
  }

  setPager(index, size) {
    this.pageIndex = index;
    this.pageSize = size;
  }

  setFormat(format) {
    // 1 for JSON
    this.format = format;
  }

  getURL() {
    this.url = this.protocol
                + this.domain
                + this.path
                + '?action='+this.action
                + '&filter:categoriesIdsMatchAnd='+this.categoriesIdsMatchAnd
                + '&filter:freeText='+this.freeText
                + '&filter:mediaTypeEqual='+this.mediaTypeEqual
                + '&filter:orderBy='+this.orderBy
                + '&format='+this.format
                + '&pager:pageIndex='+this.pageIndex
                + '&pager:pageSize='+this.pageSize
                + '&service='+this.service;
    return this.url;
  }

  getVideos() {
    var url = this.getURL();
    // console.log('url', 'pageIndex=', this.pageIndex, url);
    return axios.get(url);
  }

}
