// components/footer-nav/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'footer-nav',
  prefix: ''
});

export default React.createClass({

  displayName: 'FooterNav',

  propTypes: {
    text: React.PropTypes.string.isRequired
  },

  render() {
    return (
      <div {...classes()}>
        <ul>
          <li><a href='http://service.aargauerzeitung.ch/kontakt_impressum/verlag.php'>Kontakt</a></li>
          <li><a href='http://service.aargauerzeitung.ch/kontakt_impressum/verlag.php'>Impressum</a></li>
        </ul>
        <ul>
          <li><a href='http://azdigital.ch/'>E-Paper</a></li>
          <li><a href='http://www.aargauerzeitung.ch/mitmachen/newsletter'>Newsletter</a></li>
          <li><a href='http://www.aargauerzeitung.ch/archiv'>Archiv</a></li>
          <li><a href='http://service.aargauerzeitung.ch/abonnement/angebot.php'>Abonnement</a></li>
        </ul>
        <ul>
          <li><a href='http://www.azwerbung.ch/'>Inserieren</a></li>
          <li><a href='http://service.aargauerzeitung.ch/azbonuscard/index.php'>az Bonus</a></li>
        </ul>
        <ul>
          <li><a href='http://www.a-z.ch/static/az/home/agb.html'>AGB</a></li>
          <li><a href='http://www.aargauerzeitung.ch/hilfe'>Hilfe</a></li>
        </ul>
        <ul>
          <li className="bold"><a href='http://www.a-z.ch/immobilien'>Immbobilien</a></li>
          <li className="bold"><a href='http://www.a-z-jobs.ch/'>Jobs</a></li>
          <li className="bold"><a href='http://www.aargauerzeitung.ch/auto/'>Auto</a></li>
          <li className="bold"><a href='http://events.aargauerzeitung.ch/'>Events</a></li>
          <li className="bold"><a href='http://www.a-z.ch/vereine'>Vereine</a></li>
        </ul>
      </div>
    );
  }

});
