// components/expander/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';

import ResponsiveVideoMixin from 'mixins/responsive-video';

const classes = new BEM({
  name: 'expander',
  prefix: ''
});

// For styleguide only
const ExpandableItem = React.createClass({
  render() {
    return (
      <div {...classes('placeholder')}>Expandable Item</div>
    );
  }
});

export default React.createClass({

  displayName: 'Expander',

  propTypes: {
    item: React.PropTypes.object,
    onSetExpandableHeight: React.PropTypes.func
  },

  getDefaultProps() {
    // set property when component rendered in style guide
    return {
      onSetExpandableHeight: function() {}
    };
  },

  componentDidMount() {
    this.setHeight();
  },

  setHeight() {
    var expander,
        height,
        playerHeight;

    expander = React.findDOMNode(this.refs.expander);

    // calculate the height of the video
    playerHeight = ResponsiveVideoMixin.getVideoHeight(16 / 9, expander.offsetWidth, 35.8);

    // calculate the height of the video library viewer
    height = expander.firstChild.offsetHeight + playerHeight;

    expander.style.height = height + 'px';

    // notify expandable to take into account the new height of the expander
    this.props.onSetExpandableHeight(height);
  },

  render() {
    var item;

    if (this.props.item) {
      item = this.props.item;
    } else {
      item = <ExpandableItem />;
    }

    return (
      <div {...classes(null, null)}>
        <div {...classes(null, 'inner')} ref={'expander'}>
          {item}
        </div>
      </div>
    );
  }

});
