// components/weather/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';
import moment from 'moment';

import Section from 'components/section';
import SectionHeader from 'components/section-header';
import WeatherDetail from 'components/weather-detail';
import WeatherForecastList from 'components/weather-forecast-list';

const classes = new BEM('Weather');

export default React.createClass({

  displayName: 'Weather',

  propTypes: {
    cities: React.PropTypes.array.isRequired,
    forecast: React.PropTypes.object.isRequired,
    days: React.PropTypes.object.isRequired
  },

  getInitialState() {
    return this.state = {
      // @TODO: make sure code works when this.props.forecast is not there
      // Hack to make it temporarily work while integrating into page
      // styleguide only: this.props.forecast
      city: 'Brugg', // change to id
      id: '1',
      forecast: this.props.forecast,
      days: this.props.forecast.days
    };
  },

  getCurrentForecast(days) {
    var getTime = moment().locale('de').hours();
    // console.log("die Zeit", getTime);
    var result = parseInt(getTime / 3);
    // console.log("das resultat", result);
    var index = Math.round(result);
    return days[index];
  },

  render() {

    var currentForecast = this.getCurrentForecast(this.props.forecast.days);

    return (
      <div {...classes()}>
        <SectionHeader title='Wetter' titleModifier='large' />
        <Section modifier='default'>
          <WeatherDetail
            {...this.props}
            city={this.state.city}
            id={this.state.id}
            forecast={this.state.forecast}
            days={currentForecast}
          />
          <WeatherForecastList
            {...this.props}
            city={this.state.city}
            id={this.state.id}
            forecast={this.state.forecast}
            days={currentForecast}
          />
        </Section>
      </div>
    );
  }
});
