// inputs/checkbox/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import InputWrapper from 'inputs/wrapper';

const classes = new bem('input');

export default React.createClass({
  displayName: 'CheckboxInput',

  propTypes: {
    label: React.PropTypes.string,
    className: React.PropTypes.string,
    name: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    hint: React.PropTypes.string,
    onChange: React.PropTypes.func,
    id: React.PropTypes.string,
    type: React.PropTypes.string,
    value: React.PropTypes.string,
  },

  getInitialState() {
    return {
      value: this.props.value
    }
  },

  componentWillReceiveProps: function(nextProps) {
    if (this.state.value !== nextProps.value) {
      this.setState({ value: nextProps.value })
    }
  },

  onValueChange: function() {
    var value = this.refs.input.getDOMNode().checked;
    this.setState({ value: value })

    if (this.props.onChange) {
      this.props.onChange(this.props.name, value);
    }
  },

  render() {
    var id = this.props.id || this.props.name;

    return (
      <InputWrapper {...this.props} label='' id={id}>
        <label htmlFor={id} {...classes('label')}>
          <input
            type="checkbox"
            {...classes({ extra: 'visuallyhidden' })}
            id={id}
            ref='input'
            onChange={this.onValueChange}
            value={this.state.value}
          />
          <span></span>
          {' ' + this.props.label}
        </label>
      </InputWrapper>
    )
  }
});
