// components/dossier-header/index.jsx
import React from 'react' ;
import bem from 'react-bem-helper';

import Tags from 'components/tags';
import ResponsiveImage from 'components/responsive-image';

const classes = new bem({
  name: 'dossier-header',
  prefix: ''
});

export default React.createClass({

  displayName: 'DossierHeader',

  getInitialState() {
    return {
      hasFilterSupport: false
    };
  },

  propTypes: {
    title: React.PropTypes.string.isRequired,
    text: React.PropTypes.string.isRequired,
    background: React.PropTypes.string,
    image: React.PropTypes.object,
    tags: React.PropTypes.arrayOf(React.PropTypes.string)
  },

  componentDidMount() {
    if ('filter' in React.findDOMNode(this).style) {
      this.setState({ hasFilterSupport: true });
    }
  },

  render() {

    const { title, text, background, image, tags } = this.props;
    const { hasFilterSupport } = this.state;

    const heroBg = {
      backgroundImage: `url(${background})`
    };

    return (
      <header {...classes()}>
        <div {...classes('hero')}>
          <div {...classes('hero-text')}>
            <p {...classes('catchword')}>Dossier</p>
            <h1 {...classes('title')}>{ title }</h1>
          </div>
          <div {...classes('hero-image')} style={hasFilterSupport && background ? heroBg : null}/>
          <div {...classes('hero-overlay')} style={{opacity: hasFilterSupport ? '0.75' : '1'}}/>
        </div>
        <div {...classes('summary')}>
          <div {...classes('summary-wrapper')}>
            {image
              ? <aside {...classes('summary-aside')}>
                  <div {...classes('thumbnail')}>
                    <ResponsiveImage {...image} ratio='1x1' size='narrow' fill='true' />
                  </div>
                </aside>
            :null}
            <div {...classes('summary-content')}>
              <h2 {...classes('summary-title')}>Zusammenfassung</h2>
              <div {...classes('summary-copy')} dangerouslySetInnerHTML={{ __html: text }} />
              <div {...classes('related')}>
                <Tags items={tags} />
              </div>
            </div>
          </div>
        </div>
      </header>
    );
  }

});
