// components/dossier-header/index.jsx
import React from 'react' ;
import bem from 'react-bem-helper';

import Container from 'components/container';
import RawContent from 'components/raw-content';
import Tags from 'components/tags';
import ResponsiveImage from 'components/responsive-image';

const classes = new bem({
  name: 'dossier-header',
  prefix: ''
});

export default React.createClass({

  displayName: 'DossierHeader',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    subtitle: React.PropTypes.string.isRequired,
    text: React.PropTypes.string.isRequired,
    background: React.PropTypes.string,
    image: React.PropTypes.object,
    tags: React.PropTypes.arrayOf(React.PropTypes.string)
  },

  render() {

    const { title, subtitle, text, background, image, tags } = this.props;

    const heroStyles = {
      backgroundImage: `url(${background})`
    };

    return (
      <header {...classes()}>
        <div {...classes('hero')} style={background ? heroStyles : null}>
          <div {...classes('hero-text')}>
            <p {...classes('hero-subtitle')}>Dossier</p>
            <h1 {...classes('hero-title')}>{ title }</h1>
          </div>
        </div>
        <div {...classes('summary')}>
          <Container>
            {image
              ? <div {...classes('left-wrapper')}>
                  <div {...classes('thumbnail')}>
                    <ResponsiveImage {...image} />
                  </div>
                </div>
            :null}
            <div {...classes('right-wrapper')}>
              <h2 {...classes('summary-title')}>{ subtitle }</h2>
              <div {...classes('summary-copy', null, 'copy')}>
                <RawContent content={text} />
              </div>
              <div {...classes('related')}>
                <h3>Verwandte Themen</h3>
                <Tags items={tags} />
              </div>
            </div>
          </Container>
        </div>
      </header>
    );
  }

});
