// components/dossier-header/slot.js
var _ = require('lodash');
var getImageProps = require('helpers/asset/get-image-props');

var lowerCaseArray = function(array) {
  if (array.length) {
    return array.map(function(str) { return str.toLowerCase(); });
  }
  return array;
};

var collateKeywords = function(dossier, keywords) {
  var all = _.merge(lowerCaseArray(keywords), lowerCaseArray(dossier.keywords));
  var uniq = _.uniq(all);
  return _.remove(uniq, function(n) {
    return n !== dossier.title;
  }).slice(0, 6);
};

var dossierText = function(header) {
  return header.text || '<p>' + header.title + '</p>';
};

var parsedDossier = function(dossier, keywords) {
  var header = dossier.header;
  var mainAsset = getImageProps(_.first(header.assets));
  return {
    title: dossier.title,
    text: dossierText(header),
    background: mainAsset ? mainAsset.src : null,
    image: mainAsset,
    tags: collateKeywords(dossier, keywords)
  };
};

var parsedData = function(state) {
  return parsedDossier(state.dossier, state.keywords);
};

module.exports = {
  iso: true,
  data: parsedData
};
