// components/rule-cta/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Icon from 'components/icon';

const classes = new bem({
  name: 'rule-cta',
  prefix: ''
});

export default React.createClass({

  displayName: 'RuleCTA',

  propTypes: {
    url: React.PropTypes.string,
    label: React.PropTypes.string,
    icon: React.PropTypes.string,
    _event: React.PropTypes.oneOfType([
      React.PropTypes.func,
      React.PropTypes.string
    ])
  },

  render() {

    const { url, label, icon, _event } = this.props;

    const labelEl = label ? <span {...classes('label')}>{ label }</span> : null;
    const iconEl = icon ? <Icon id={ icon } {...classes('icon')} /> : null;
    const linkEl = url && (label || icon) ?
      <a href={ url } {...classes('link')}>
        { labelEl }
        { iconEl }
        </a>
    : null;
    const buttonEl = _event && (label || icon) ?
      <button onClick={_event} {...classes('button')}>
        { labelEl }
        { iconEl }
      </button>
    : null;

    return (
      <div {...classes()} >
        { linkEl || buttonEl || null }
        <div {...classes('rule')} />
      </div>
    );
  }

});
