// styleguide/component/index.jsx
'use strict';

const React = require('react');
const slugify = require('slugify');
const path = require('path');

const ReadMe = require('styleguide/readme');
const Viewer = require('styleguide/viewer');
const Variations = require('styleguide/variations');

const Component = React.createClass({

  propTypes: {
    readme: React.PropTypes.string,
    children: React.PropTypes.element.isRequired,
    category: React.PropTypes.string.isRequired,
    component: React.PropTypes.object.isRequired,
    variation: React.PropTypes.object,
  },

  mapVariationItems(variations) {
    var props = this.props;
    return variations.map(function(variation) {
      var variationItem = {};
      var variationSlug = slugify(variation.title).toLowerCase();
      variationItem.title = variation.title;
      variationItem.link = path.join('/styleguide', props.category, props.component.slug, variationSlug);
      variationItem.slug = variationSlug;
      variationItem.active = variationSlug === props.variation.slug;
      return variationItem;
    });
  },

  render() {

    var category = this.props.category;
    var component = this.props.component;
    var variation = this.props.variation;
    var variations = component.variations && component.variations.length > 1 ? <Variations variations={this.mapVariationItems(component.variations)} /> : false;
    var content = component.variations
      ? <Viewer category={category} component={component} variation={variation} json={variation ? variation.data : false}>{this.props.children}</Viewer>
      : this.props.children;

    return (
      <div>
        <ReadMe {...component} />
        {content}
        {variations}
      </div>
    );

  }

});

module.exports = Component;
