// styleguide/component/index.jsx
'use strict';

const React = require('react');
const ReadMe = require('styleguide/readme');
const Viewer = require('styleguide/viewer');
const Variations = require('styleguide/variations');
const DataStructure = require('styleguide/data-structure');

const slugify = require('slugify');

var Component = React.createClass({

  propTypes: {
    readme: React.PropTypes.string,
    component: React.PropTypes.object,
  },

  constructUrl(component) {
    return '/styleguide/component/' + component.slug + '/full';
  },

  mapVariationItems(variations) {
    var props = this.props;
    return variations.map(function(variation) {
      var variationItem = {};
      var slug = slugify(variation.title).toLowerCase();
      variationItem.title = variation.title;
      variationItem.link = '/styleguide/component/' + props.component.slug + '/' + slug;
      variationItem.slug = slug;
      variationItem.active = slug === slugify(props.component.variation.title).toLowerCase();
      return variationItem;
    });
  },

  render() {
    var component = this.props.component, variations, dataStructure;

    if (component.variation && component.variation.data) {
      dataStructure = <DataStructure data={component.variation.data} />;
    }

    if (component.variations) {
      variations = <Variations variations={this.mapVariationItems(component.variations)} />;
    }

    return (
      <div>
        <ReadMe {...component} />
        <Viewer src={this.constructUrl(component)}/>
        {variations}
        {dataStructure}
      </div>
    );
  }

});

module.exports = Component;
