// components/curated-recommendations/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Teaser from 'components/teaser';

const classes = new bem({
  name: 'curated-recommendations',
  prefix: ''
});

export default React.createClass({

  propTypes: {
    zuthir: React.PropTypes.string.isRequired,
    teasers: React.PropTypes.array.isRequired
  },

  render() {
    var wrapperModifier = this.props.teasers.length > 1
        ? 'multiple'
        : null;

    var innerModifier = null;

    if (this.props.teasers.length === 2) {
      innerModifier = 'half';
    } else if (this.props.teasers.length === 3) {
      innerModifier = 'third';
    }

    return (
      <div {...classes()}>
        <h3 {...classes('title')}>{this.props.author} emfpielth</h3>
        <div {...classes('wrapper', wrapperModifier)}>
          {this.props.teasers.map((teaser, index) =>
            <div {...classes('inner', innerModifier)}>
              <Teaser {...teaser} key={index} />
            </div>
          )}
        </div>
      </div>
    );
  }

});
