// components/curated-recommendations/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';
import Teaser from 'components/teaser';
import Slider from 'mixins/slider';

const classes = new bem({
  name: 'curated-recommendations',
  prefix: ''
});

const CuratedRecommendations = React.createClass({

  displayName: 'CuratedRecommendations',

  mixins: [Slider],

  propTypes: {
    author: React.PropTypes.string,
    teasers: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  getInitialState() {
    return {
      mobile: false,
      loading: true
    };
  },

  componentDidMount() {
    var startingIndex = Math.floor(this.props.teasers.length / 2);
    var slider = this.refs.curatedRecommendationsSlider.getDOMNode();

    if (this.props.teasers.length > 1) {
      this.renderFlickity({
        startIndex: startingIndex,
        element: slider,
        delay: 250
      });
    } else {
      this.setState({ loading: false });
    }
  },

  render() {

    var { author, teasers } = this.props;

    var { loading, resizing } = this.state;

    var hasMultiple = teasers.length > 1;

    var modifiers = {
      'multiple': hasMultiple,
      ['count-' + teasers.length]: true
    };

    var extraClasses = {
      'is-slider-loading': loading,
      'is-slider-resizing': resizing
    };

    return (
      <div {...classes(null, modifiers, extraClasses)}>
        <div {...classes('inner')}>
          {author ?
            <h3 {...classes('title')}>{author + ' empfiehlt'}</h3>
          :null}
          <div {...classes('grid')} ref='curatedRecommendationsSlider'>
            {teasers.map((teaser, index) =>
              <div {...classes('item')} key={index}>
                <Teaser key={index} {...teaser} modifier='light-boxed' />
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }

});

export default Wrapper(CuratedRecommendations);
