// pages/ressort/index.jsx
import React from 'react';

import Container from 'components/container';
import RessortHeader from 'components/ressort-header';
import Grid from 'components/grid';
import Teaser from 'components/teaser';
import TeaserSwitchable from 'components/teaser-switchable';
import { getTeaserFromArticle, getTeaserSwitchableFromArticle } from 'helpers/teaser';

export default React.createClass({

  displayName: 'RessortPage',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    slug: React.PropTypes.string.isRequired,
    subnav: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    articles: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mappedTeasers(articles) {
    return articles.map((article, index) => {

      let teaserData;

      if (index > 4) {
        teaserData = getTeaserSwitchableFromArticle({
          article: article,
          variation: 'inlined-left-to-basic'
        });
        return <TeaserSwitchable key={index} {...teaserData} />;
      }

      teaserData = getTeaserFromArticle({
        article: article,
        variation: index === 0 ? 'large' : 'basic'
      });
      return <Teaser key={index} {...teaserData} />;

    });
  },

  render() {
    return (
      <Container>
        <RessortHeader {...this.props} />
        <Grid modifier='three-up-featured' items={this.mappedTeasers(this.props.articles)} />
      </Container>
    );
  }

});
