// components/ressort-teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Grid from 'components/grid';
import Container from 'components/container';

import RessortTeaserBuilder from 'higher-order/ressort-teaser-builder';

const classes = new bem({
  name: 'ressort-teaser',
  prefix: ''
});

var RessortTeaser = React.createClass({

  displayName: 'Ressort Teaser',

  propTypes: {
    title: React.PropTypes.string,
    items: React.PropTypes.array,
    modifier: React.PropTypes.string,
  },

  render() {

    const { title, items, modifier } = this.props;

    return (
      <section {...classes()}>
        <Container>
          <header {...classes('header')}>
            <h1 {...classes('heading')}>{ title }</h1>
          </header>
          <Grid modifier={ modifier } items={ items }/>
          <footer {...classes('footer')}>
            <a href='/' {...classes('cta')}>Mehr artikel zum thema Schweiz &#10142;</a>
            <div {...classes('rule')} />
          </footer>
        </Container>
      </section>
    );
  }

});

export default RessortTeaserBuilder(RessortTeaser);
