// components/ressort-teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Grid from 'components/grid';
import RuleCta from 'components/rule-cta';
import SectionHeader from 'components/section-header';

import RessortTeaserBuilder from './ressort-teaser-builder';

const classes = new bem({
  name: 'ressort-teaser',
  prefix: ''
});

var RessortTeaser = React.createClass({

  displayName: 'Ressort Teaser',

  propTypes: {
    title: React.PropTypes.string,
    ressortUrl: React.PropTypes.string,
    items: React.PropTypes.array,
    modifier: React.PropTypes.string
  },

  render() {

    const { title, ressortUrl, items, modifier } = this.props;

    // Only render component if there are items
    if (items.length === 0) return null;

    return (
      <section {...classes()}>
        {title
          ? <SectionHeader title={title} link={ressortUrl} titleModifier='large' />
        :null}
        <Grid modifier={ modifier } items={ items }/>
        <footer {...classes('footer')}>
          <RuleCta
            url={ressortUrl}
            label={'Mehr artikel in ' + title}
            icon={'arrow-right'} />
        </footer>
      </section>
    );
  }

});

export default RessortTeaserBuilder(RessortTeaser);
