// components/ressort-teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Grid from 'components/grid';
import Container from 'components/container';
import RuleCta from 'components/rule-cta';

import RessortTeaserBuilder from './ressort-teaser-builder';

const classes = new bem({
  name: 'ressort-teaser',
  prefix: ''
});

var RessortTeaser = React.createClass({

  displayName: 'Ressort Teaser',

  propTypes: {
    title: React.PropTypes.string,
    ressortUrl: React.PropTypes.string,
    items: React.PropTypes.array,
    modifier: React.PropTypes.string
  },

  render() {

    const { title, ressortUrl, items, modifier } = this.props;

    // Only render component if there are items
    if (items.length === 0) return null;

    return (
      <Container>
        <section {...classes()}>
          {title ?
            <header {...classes('header')}>
                <h1 {...classes('heading')}>
                  <a href={ ressortUrl } {...classes('heading-link')}>
                    { title }
                  </a>
                </h1>
            </header>
          : null}
          <Grid modifier={ modifier } items={ items }/>
          <footer {...classes('footer')}>
            <RuleCta
              url={ressortUrl}
              label={'Mehr artikel in ' + title}
              icon={'arrow-right'} />
          </footer>
        </section>
      </Container>
    );
  }

});

export default RessortTeaserBuilder(RessortTeaser);
