// components/ressort-teaser/ressort-teaser-builder.jsx
import React from 'react';

import Teaser from 'components/teaser';
import getTeaserFromArticle from 'helpers/get-teaser-from-article';

export default Component => React.createClass({

  displayName: 'RessortTeaserBuilder',

  propTypes: {
    title: React.PropTypes.string,
    ressortUrl: React.PropTypes.string,
    articles: React.PropTypes.array,
    slots: React.PropTypes.number.isRequired,
    modifier: React.PropTypes.oneOf(['three-up-simple', 'three-up-featured']).isRequired
  },

  render() {

    const { articles, slots, modifier } = this.props;

    // Initialise items with Teaser components
    var items = articles.map((article, index) => {
      return <Teaser key={index} {...getTeaserFromArticle({
        article: article,
        variation: modifier === 'three-up-featured' && index === 0
          ? 'large'
          : 'basic'
      })} />;
    });

    // The array of items can now be decorated with ressort-specific widgets
    // e.g. items.splice(1, 0, <FootballResultsWidget />);

    return <Component {...this.props} items={items.slice(0, slots)} modifier={ modifier }/>;
  }

});
