// components/enrichment-snippet/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getAssetMeta, getEnrichmentPlacement } from 'helpers';
import Caption from 'components/caption';

const classes = new bem({
  name: 'enrichment-snippet',
  prefix: ''
});

export default React.createClass({

  displayName: 'EnrichmentSnippet',

  propTypes: {
    asset: React.PropTypes.object,
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    placement: React.PropTypes.string
  },

  componentDidMount() {
    // DON'T re-render on the client
    // snippets might contain <script> tags which do their own thing
    return false;
  },

  filterHtmlNodes(html) {

    var filteredHtml = html;

    // Wrap Youtube or Vimeo iframes in a div
    const videoRegex = /(<iframe[^<>]*?src="[\w:/.]*(youtube.com|vimeo.com)[^<>]*?><\/iframe>)/;
    filteredHtml = filteredHtml.replace(videoRegex, '<div class="enrichment-snippet__video">$&</div>');

    // Wrap trailing text (a caption) in a span
    const captionRegex = /((?![<>]).)*$/;
    filteredHtml = filteredHtml.replace(captionRegex, '<span class="enrichment-snippet__text">$&</span>');

    return filteredHtml;
  },

  render() {
    const { asset } = this.props;

    return (
      <div {...classes(null, getEnrichmentPlacement(this.props.placement))}>
        <div {...classes('pandoras-box')} dangerouslySetInnerHTML={{ __html: this.filterHtmlNodes(asset.html) }} />
        {asset.source
         ? <Caption {...getAssetMeta(this.props)} />
        :null}
      </div>
    );
  }

});
