// components/video-library-preview/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import VideoTeaser from 'components/video-teaser';
import Grid from 'components/grid';

import Button from 'components/button';

const classes = new bem({
  name: 'video-library-preview',
  prefix: ''
});

export default React.createClass({

  displayName: 'VideoLibraryPreview',

  propTypes: {
    videos: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mapTeaserItems(items) {
    return items.map((item, index) => {
      return (
        <VideoTeaser
        name={item.name}
        thumbnailUrl={item.thumbnailUrl}
        key={index} />
      );
    });
  },

  render() {
    var videoTeasers = this.props.videos;
    var errorMessage = 'Leider konnten keine Videos zu Ihrem Suchbegriff gefunden werden';
    var items;
    var content;


    if (videoTeasers.length) {
      items = this.mapTeaserItems(videoTeasers);
      content = <Grid modifier='video-library' items={items} />;

    } else {
      content = (
        <div {...classes('fallback')}>
          <Alert text={errorMessage} modifier='centered' />
        </div>
      );
    }



    return (
      <div {...classes('container')}>
        {content}
        <Button label="Mehr anzeigen" _target="#"/>
      </div>
    );
  }

});
