// components/background-image/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem('background-image');

export default React.createClass({

  displayName: 'BackgroundImage',

  propTypes: {
    background: React.PropTypes.string,
    theme: React.PropTypes.string
  },

  getInitialState() {
    return {
      hasFilterSupport: false
    };
  },

  componentDidMount() {
    if ('filter' in React.findDOMNode(this).style) {
      this.setState({ hasFilterSupport: true });
    }
  },

  render() {

    let backgroundStyles;

    const { background, theme } = this.props;
    const { hasFilterSupport } = this.state;

    if (background) {
      backgroundStyles = {
        backgroundImage: `url(${background})`
      };
    }

    return (
      <span {...classes(null, theme)}>
        <span {...classes('image')} style={hasFilterSupport && background ? backgroundStyles : null}></span>
        <span {...classes('overlay')} style={{opacity: hasFilterSupport && background ? '0.9' : '1'}}></span>
      </span>
    );
  }

});
