// components/footer-inject/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'footer-inject',
  prefix: ''
});

export default React.createClass({

  displayName: 'FooterInject',

  propTypes: {
    text: React.PropTypes.string.isRequired
  },

  render() {
    return (
      <div {...classes()} dangerouslySetInnerHTML={{
        __html: `
        <div class="footerlinks">
          <h5><a href="http://www.azmedien.ch" target="_top" name="titleLink" class="footerTitle">AZ Medien</a></h5>
          <div class="block clearfix">
          <span class="left">Tageszeitungen:</span>
          <span class="right">
          <a target="_blank" href="http://www.nordwestschweiz.ch/">az Nordwestschweiz</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.aargauerzeitung.ch">az Aargauer Zeitung</a>&nbsp;|&nbsp;<a href="http://www.badenertagblatt.ch/">az Badener Tagblatt</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.basellandschaftlichezeitung.ch">bz Basellandschaftliche Zeitung</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.bzbasel.ch">bz Basel</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.limmattalerzeitung.ch">az Limmattaler Zeitung</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.solothurnerzeitung.ch">az Solothurner Zeitung</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.grenchnertagblatt.ch">az Grenchner Tagblatt</a>&nbsp;|&nbsp;<a href="http://www.oltnertagblatt.ch/">ot Oltner Tagblatt</a>          </span>
        </div>

        <div class="block clearfix">
          <span class="left">Sonntagszeitung:</span>
          <span class="right">
          <a target="_blank" href="http://www.schweizamsonntag.ch/">Schweiz am Sonntag</a>          </span>
        </div>

        <div class="block clearfix">
          <span class="left">Anzeiger:</span>
          <span class="right">
          <a target="_blank" href="http://www.affolteranzeiger.ch/">Anzeiger Bezirk Affoltern</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.azeiger.ch/">AZEIGER</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.berner-landbote.ch/">Berner Landbote</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.grenchnerstadtanzeiger.ch/">Grenchner Stadt-Anzeiger</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.lba.azmedien.ch/">Lenzburger Bezirksanzeiger</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.limmatwelle.ch/">Limmatwelle</a>&nbsp;|&nbsp;<a href="http://www.oberbaselbieterzeitung.ch/">Oberbaselbieter Zeitung</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.stadtanzeiger-olten.ch/">Stadtanzeiger Olten</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.wochenblatt.ch/wob/">Wochenblatt Birseck/Dorneck</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.wochenblatt.ch/wos/">Wochenblatt Schwarzbubenland/Laufental</a>          </span>
        </div>

        <div class="block clearfix">
          <span class="left">Zeitschriften/Bücher:</span>
          <span class="right">
          <a target="_blank" href="http://www.fitforlife.ch">FIT for LIFE</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.wildeisen.ch">KOCHEN</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.natuerlich-online.ch">natürlich</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.wireltern.ch">wir eltern</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.batitech.ch">Bâtitech</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.chemieplus.ch">Chemie Plus</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.elektrotechnik.ch">Elektrotechnik ET</a>&nbsp;|&nbsp;<a href="http://www.haustech-magazin.ch/">Haustech</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.hk-gebaeudetechnik.ch">HK Gebäudetechnik</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.megalink.ch">Megalink</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.technica-online.ch">Technica</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.atverlag.ch">AT Verlag</a>         </span>
        </div>

        <div class="block clearfix">
          <span class="left">TV-Senderfamilie:</span>
          <span class="right">
          <a target="_blank" href="http://www.telezueri.ch">TeleZüri</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.telem1.ch">Tele M1</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.telebaern.ch">TeleBärn</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.tv24.ch/">TV24</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.belcom.ch">Belcom</a>         </span>
        </div>

        <div class="block clearfix">
          <span class="left">Online:</span>
          <span class="right">
          <a target="_blank" href="http://www.aznetz.ch/">Nordwestschweiz Netz</a>&nbsp;|&nbsp;<a href="http://www.watson.ch">watson</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.a-z.ch/">a-z.ch</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.marktspiegel.ch/">Marktspiegel</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.heim-maert.ch/">HeimMärt</a>&nbsp;|&nbsp;<a target="_blank" href="http://gedenkzeit.ch/">Gedenkzeit.ch</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.azregiohelden.ch">AZ Regiohelden</a>         </span>
        </div>

        <div class="block clearfix">
          <span class="left">Druck/Vertrieb:</span>
          <span class="right">
          <a target="_blank" href="http://www.mittellandzeitungsdruck.ch/de/">Mittelland Zeitungsdruck</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.vsdruck.ch">Vogt-Schild Druck</a>&nbsp;|&nbsp;<a target="_blank" href="http://www.weissdruck.ch">Weiss Medien</a>&nbsp;|&nbsp;<a href="/site/dienstleistungen/vertrieb/">AZ Vertrieb</a>         </span>
        </div>
        </div>`
      }} />
    );
  }

});
