// components/article-header/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import DateFormatter from 'components/date-formatter';

const classes = new bem({
  name: 'article-header',
  prefix: ''
});

export default React.createClass({

  displayName: 'ArticleHeader',

  propTypes: {
    catchword: React.PropTypes.string.isRequired,
    title: React.PropTypes.string.isRequired,
    author: React.PropTypes.string.isRequired,
    source: React.PropTypes.string,
    published: React.PropTypes.string.isRequired,
    modified: React.PropTypes.string.isRequired
  },

  render() {
    var props = this.props;
    var source = props.source ? ' - ' + props.source : null;

    return (
      <header {...classes()}>
        <p itemProp='genre' {...classes('catchword', null, 'type-catchword')}>{props.catchword}</p>
        <h1 itemProp="headline" {...classes('title', null, 'type-h1')}>{props.title}</h1>
        <div {...classes('meta', null, 'type-detail')}>
          <p itemProp='author' {...classes('author', null)}>
            von {props.author}{source}
          </p>
          <p {...classes('date')}>
            <DateFormatter
              publicationDate={props.published}
              modificationDate={props.modified}
            />
          </p>
        </div>
      </header>
    );
  }

});
