// components/article-header/slot.js
var getSource = require('helpers/asset/get-source');
var helpers = require('helpers');
var _ = require('lodash');
var goldbachImageSource = require('helpers/image/goldbach-image-source');
var SPONSORED = 'sponsored';

function squareImage(src) {
  return goldbachImageSource({
    width: 70,
    height: 70,
    fill: true,
    src: src
  });
}

var parsedArticle = function(article) {
  return {
    catchword: article.context_label,
    title: article.title,
    author: article.author,
    source: getSource(article.source),
    published: article.dc.effective,
    modified: article.dc.modified,
    banner: getBanner(article)
  };
};

var parsedData = function(state) {
  return parsedArticle(state.article);
};

function getBannerType(article) {
  if (article.sponsor && article.sponsor.text) {
    return SPONSORED;
  } else if (article.content_type === 'ugcnewsarticle') {
    return article.labeltype || _.get(article, 'ressorts.0.urlpart');
  }

  return false;
}

function getBanner(article) {
  var bannerType = getBannerType(article);
  var icon = 'megaphone';

  if (!bannerType) {
    return null;
  }

  return {
    icon: icon,
    modifier: bannerType,
    blocks: _.compact(blocksForBannerType(bannerType, article))
  }
}

function blocksForBannerType (bannerType, article) {
  function authorBlock(article, text) {
    var author = _.get(article, 'owner') || {} ;

    if (!author.firstname) {
      return;
    }

    return {
      "text": text,
      "linkText": [author.firstname, author.lastname].join(' '),
      "image": author.image_url ? squareImage(author.image_url) : null,
      "itemProp": "author"
    }
  }

  function sponsorBlock(article) {
    var sponsor = _.get(article, 'sponsor', {});

    if (!sponsor.text) {
      return
    }

    return {
      "text": sponsor.text,
      "link": sponsor.url,
      "linkText": sponsor.title,
      "image": sponsor.image.image_url,
      "imageRounded": false
    }
  }

  function clubBlock(article) {
    return null;
  }

  function cityBlock(article, text) {
    var city = _.get(article, 'cities.0', {});

    if (!city.name) {
      return;
    }

    return {
      "text": text,
      "linkText": city.name,
      "link": '/gemeinde/' + city.urlpart,
      "image": '/static/az/img/Flaggen/gemeinde/' + city.id + '.png',
      "itemProp": "contentLocation",
      "imageRounded": false
    }
  }

  switch (bannerType) {
    case 'leserbeitrag':
      return [
        authorBlock(article, 'Leserbeitrag von'),
        cityBlock(article, 'aus')
      ]
    case 'gemeindebeitrag':
      return [
        cityBlock(article, 'Gemeindebeitrag aus'),
        authorBlock(article, 'von'),
      ]
    case 'vereinsmeldung':
    case 'sportverein':
      return [
        clubBlock(article),
        authorBlock(article)
      ]
    case SPONSORED:
      return [
        sponsorBlock(article)
      ]
  }
}

module.exports = {
  iso: false,
  data: parsedData
};
