// components/enrichment-video/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Video from 'components/video';

import getAssetMeta from 'helpers/get-asset-meta';
import getEnrichmentPlacment from 'helpers/get-enrichment-placement';
import ResponsiveVideoMixin from 'mixins/responsive-video';

const classes = new bem({
  name: 'enrichment-video',
  prefix: ''
});

const EnrichmentVideo = React.createClass({

  displayName: 'EnrichmentVideo',

  mixins: [ResponsiveVideoMixin],

  propTypes: {
    caption: React.PropTypes.object,
    meta: React.PropTypes.object.isRequired
  },

  makeMeta(props) {
    return {
      thumbnailUrl: props.still_image.image_url || props.still_image.asset.image_url
    };
  },

  render() {
    return (
      <div {...classes(null, getEnrichmentPlacment(this.props.placement))}>
        <Video
          meta={this.makeMeta(this.props.asset)}
          caption={getAssetMeta(this.props)}
          kaltura_id={this.props.asset.kaltura_id}
          accountID='1789881'
          playerID='26921101'
          playerWithoutAdsID='29381931'
          trackingURL='https://aznetz.wemfbox.ch/cgi-bin/ivw/CP/aaz?r=http%3A//www.aargauerzeitung.ch/ressort/subressort/article-title-129298028&amp;d=&amp;x='
          autoplay={false}
        />
      </div>
    );
  }

});

export default EnrichmentVideo;
