// components/enrichment-video/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Video from 'components/video';

import Wrapper from './wrapper';

import { getEnrichmentPlacement } from 'helpers/enrichment';
import ResponsiveVideoMixin from 'mixins/responsive-video';

const classes = new bem({
  name: 'enrichment-video',
  prefix: ''
});

const EnrichmentVideo = React.createClass({

  displayName: 'EnrichmentVideo',

  mixins: [ResponsiveVideoMixin],

  propTypes: {
    kaltura_id: React.PropTypes.string.isRequired,
    accountID: React.PropTypes.string.isRequired,
    meta: React.PropTypes.object.isRequired,
    caption: React.PropTypes.object,
    placement: React.PropTypes.string.isRequired,
    isMainAsset: React.PropTypes.bool
  },

  render() {
    return (
      <div {...classes(null, getEnrichmentPlacement(this.props.placement))}>
        <Video {...this.props} autoPlay={false} />
        {this.props.isMainAsset && this.props.meta.thumbnailUrl ?
          <meta itemProp='image' content={this.props.meta.thumbnailUrl} />
        :null}
      </div>
    );
  }

});

export default Wrapper(EnrichmentVideo);
