// components/enrichment-video/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Video from 'components/video';

import Wrapper from './wrapper';

import getEnrichmentPlacment from 'helpers/get-enrichment-placement';
import ResponsiveVideoMixin from 'mixins/responsive-video';

const classes = new bem({
  name: 'enrichment-video',
  prefix: ''
});

const EnrichmentVideo = React.createClass({

  displayName: 'EnrichmentVideo',

  mixins: [ResponsiveVideoMixin],

  propTypes: {
    kaltura_id: React.PropTypes.string.isRequired,
    accountID: React.PropTypes.string.isRequired,
    playerID: React.PropTypes.string.isRequired,
    playerWithoutAdsID: React.PropTypes.string.isRequired,
    trackingURL: React.PropTypes.string.isRequired,
    meta: React.PropTypes.object.isRequired,
    caption: React.PropTypes.object,
    placement: React.PropTypes.string.isRequired
  },

  render() {
    return (
      <div {...classes(null, getEnrichmentPlacment(this.props.placement))}>
        <Video {...this.props} autoPlay={false} />
      </div>
    );
  }

});

export default Wrapper(EnrichmentVideo);
