// components/enrichment-video/wrapper.jsx
import React from 'react';

import getAssetMeta from 'helpers/get-asset-meta';

export default Component =>  React.createClass({

  displayName: 'EnrichmentVideoWrapper',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    source: React.PropTypes.object.isRequired,
    asset: React.PropTypes.object.isRequired
  },

  createMeta(props) {
    return {
      thumbnailUrl: props.still_image.image_url || props.still_image.asset.image_url
    };
  },

  mappedData(props) {
    return {
      meta: this.createMeta(props.asset),
      caption: getAssetMeta(props),
      placement: props.placement,
      kaltura_id: props.asset.kaltura_id,
      accountID: '1789881',
      playerID: '26921101',
      playerWithoutAdsID: '29381931',
      trackingURL: 'https://aznetz.wemfbox.ch/cgi-bin/ivw/CP/aaz?r=http%3A//www.aargauerzeitung.ch/ressort/subressort/article-title-129298028&amp;d=&amp;x='
    };
  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
