// components/enrichment-video/wrapper.jsx
import React from 'react';

import getAssetMeta from 'helpers/get-asset-meta';

export default Component =>  React.createClass({

  displayName: 'EnrichmentVideoWrapper',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    source: React.PropTypes.object.isRequired,
    asset: React.PropTypes.object.isRequired
  },

  createMeta(props) {
    return {
      thumbnailUrl: props.still_image.image_url || props.still_image.asset.image_url
    };
  },

  mappedData(props) {
    return {
      meta: this.createMeta(props.asset),
      caption: getAssetMeta(props),
      placement: props.placement,
      kaltura_id: props.asset.kaltura_id,
      accountID: process.env.KALTURA_ACCOUNT_ID,
      playerID: process.env.KALTURA_PLAYER_ID,
      playerWithoutAdsID: process.env.KALTURA_PLAYER_NOADS_ID,
      trackingURL: process.env.KALTURA_TRACKING_URL
    };
  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
