// dashboard/authentication/api.jsx
import { pick } from 'lodash';
import { translate } from 'helpers/strings';
import axios from 'axios';

const LOGIN_URL = '/__api/v1/login';
const REGISTER_URL = '/__api/v1/registration';

function errorMessage(message) {
  return {
    errorMessage: message || translate('unknown'),
    submitting: false
  };
}

export default {
  login(formData, callback) {
    var fieldsToPost = pick(formData, ['email', 'password']);

    axios.post(LOGIN_URL, fieldsToPost).then(function(response) {
      var data = response.data;
      if (data.error) {
        callback({
          errorMessage: translate(data.error),
          submitting: false
        });
      } else if (data.status === "ok") {
        callback({
          authenticated: true,
          submitting: false
        });
      } else {
        callback(errorMessage());
      }
    }).catch( () => callback(errorMessage()) );
  },
  register(formData, callback) {
    var fieldsToPost = pick(formData, [
      'username',
      'email',
      'password'
    ]);

    if (formData.password !== formData.password_confirmation) {
      callback(errorMessage('Passwords don\'t match'));
      return;
    }

    if (!formData.accepted_agc) {
      callback(errorMessage('AGB must be accepted'));
      return;
    }

    axios.post(REGISTER_URL, fieldsToPost).then(function(response) {
      var data = response.data;
      if (data.errors) {
        callback({
          errorMessage: data.errors.map(translate),
          submitting: false
        });
      } else if (data.status === "ok") {
        // We don't want to mark the user as authenticated yet,
        // because this should be set only when there's a login cookie.
        // People still need to verify email, so we go to login+message instead
        callback({
          authenticated: false,
          type: 'login',
          errorMessage: translate('verify email address'),
          submitting: false
        });
      } else {
        callback(errorMessage());
      }
    }).catch( () => callback(errorMessage()) );
  }
};
