// dashboard/authentication/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import { capitalize } from 'lodash/string';

import SectionHeader from 'components/section-header';
import TextInput from 'inputs/text';
import TextArea from 'inputs/text-area';
import AuthenticationAPI from 'dashboard/authentication/api';
import Alert from 'components/alert';

const classes = new bem('authentication-form');

export default React.createClass({

  displayName: 'AuthenticationForm',

  propTypes: {
    type: React.PropTypes.oneOf(['login', 'register']),
    onSuccess: React.PropTypes.func
  },

  getInitialState() {
    return {
      type: this.props.type || 'login',
      submitting: false,
      errorMessage: null,
      authenticated: false,
    }
  },

  onChange(key, value) {
    this.setState({
      [key]: value
    })
  },

  onFormSubmit(e) {
    e.preventDefault();

    this.setState({
      submitting: true,
      errorMessage: null
    }, () => {
      // Wait for the state to update before sending request.
      // Otherwise race conditions might make react ignore error messages in the callback with near-instant fails.
      AuthenticationAPI[this.state.type](this.state, this.onSubmitResult)
    });
  },

  onSubmitResult(newState) {
    this.setState(newState);

    if (newState.authenticated) {
      this.props.onSuccess();
    }
  },

  isLoginType() {
    return this.state.type === 'login';
  },

  toggleAuthType(e) {
    e.preventDefault();

    this.setState({
      type: this.isLoginType() ? 'register' : 'login'
    });
  },

  renderHint() {
    return (
      <a href="https://login.azmedien.ch/omni/login/recover/">Passwort vergessen?</a>
    );
  },

  renderHelpText() {
    if (this.isLoginType()) {
      return (
        <p>
          { 'Oder haben sie noch keinen Account? '}
          <a href="/registrieren" onClick={this.toggleAuthType}>Registrieren</a>
        </p>
     )
    } else {
      return (
        <p>
          { 'Oder haben sie einen Account? '}
          <a href="/anmelden" onClick={this.toggleAuthType}>Anmelden</a>
        </p>
      )
    }
  },

  renderFieldsForAuthType(type) {
    if (this.isLoginType()) {
      return (
        <div>
          <TextInput className="form-row" name="email" value={ this.state.email } type="email" label="Email" onChange={this.onChange} required='required' />
          <TextInput className="form-row" name="password" value={ this.state.password } label="Passwort" onChange={this.onChange} type="password" hint={this.renderHint()} required='required' />
        </div>
      )
    } else {
      return (
        <div>
          <TextInput className="form-row" name="username" value={ this.state.username } label="Anzeigename" description="Erlaubt sind Buchstaben, Zahlen, Bindestrich, Leerzeichen und Punkt. Minimum 3 Buchstaben" onChange={this.onChange} required='required' />
          <TextInput className="form-row" name="email" value={ this.state.email } type="email" label="Email" onChange={this.onChange} required='required' />
          <TextInput className="form-row" name="password" value={ this.state.password } label="Passwort" onChange={this.onChange} type="password" description="Das passwort muss mindestens 6 Zeichen lang sein" required='required'/>
          <TextInput className="form-row" name="password_confirmation" value={ this.state.password_confirmation } label="Passwort bestätigen" onChange={this.onChange} type="password" required='required' />
        </div>
      )
    }
  },

  render() {
    // TODO: move error message to seperate component
    return (
      <form {...classes()} onSubmit={ this.onFormSubmit }>
        <SectionHeader title={ this.isLoginType() ? 'Anmelden' : 'Registrieren'}/>
        { this.renderHelpText() }
        { this.state.errorMessage ? <Alert text={this.state.errorMessage} /> : null }
        <div {...classes('form')}>
          { this.renderFieldsForAuthType(this.state.type) }

          <div className="form-row">
            <input type="submit" disabled={this.state.submitting} className="form__submit button button--submit" />
          </div>
        </div>
      </form>

    );
  }

});
