// pages/video-library/index.jsx
/* global KalturaConfiguration: true, KalturaBaseEntryFilter: true, KalturaClient: true */
import React from 'react';

import Container from 'components/container';
import VideoLibraryHeader from 'components/video-library-header';
import VideoLibraryContent from 'components/video-library-content';
import VideoLibraryPreview from 'components/video-library-preview';

export default React.createClass({

  displayName: 'VideoLibrary',

  /*propTypes: {
    videos: React.PropTypes.arrayOf(React.PropTypes.object)
  },*/

  getInitialState() {
    return {
      searchText: '',
      kaltura: {
        categoryId: '21939571',
        ks: 'djJ8MTcxOTIyMXxcM-06UvOvIYskV0AlrUu9pc3VszD9JqSjTnrznOElmk0tUYIS3bN_np9rWxriBbtwBtOdgP86b8mD0hSj6X-4iUDFlXIHszUTwXWLMcNujw=='
      },
      videos: null
    };
  },

  changeSearchText() {
    this.setState({searchText: event.target.value});
  },

  onChange: function(e) {
    this.setState({searchText: e.target.value});
  },

  getKalturaSession() {
    /*var _self = this;
    var setKS = function(data) {
      _self.setState({
          kaltura: {
            ks: data[0]
          }
      });
    };

    jQuery.getJSON('http://yoleidoo.com/az/kaltura-code-samples/az/player%20embedding/getKS.php', function( data ) {
      setKS(data);  //implement this on your production server, make sure you use a user with restricted rights
                    //this link is for development purposed only. The KS generated grants admin rights.
    });*/
  },

  getKalturaVideosByNameAndCategory() {
    var _self = this;
    var getKalturaVideos = function () {
      // get videos using Kaltura Client Library API
      var partnerId =1719221;
      var config = new KalturaConfiguration(partnerId);
      config.serviceUrl = "http://www.kaltura.com/";
      var client = new KalturaClient(config);
      client.ks = _self.state.kaltura.ks;
      var filter = new KalturaBaseEntryFilter();
      filter.categoriesIdsMatchOr = _self.state.kaltura.categoryId;//21939571: ZüriNews
      //filter.categoryAncestorIdIn = _self.state.kaltura.categoryId;//21939571: ZüriNews
      filter.freeText = _self.state.searchText+'*';
      filter.orderBy = '-createdAt';//KalturaBaseEntryOrderBy.CREATED_AT_ASC;
      //var filter.typeEqual = KalturaEntryType.MEDIA_CLIP;
      var pager = null;

      // load media entries
      var mediaDataLoaded = function (success, results) {
        if (!success) {
          //console.log('ERROR: '+results);
        }

        if (results.code && results.message) {
          //console.log(results.objects);
          //console.log('ERROR: '+results.message);
        }

        // set loaded data to component's state
        if (_self.isMounted()) {
          _self.setState({
            videos: results.objects
          });
        }
      };
      client.baseEntry.listAction(mediaDataLoaded, filter, pager);

    };
    getKalturaVideos();
  },

  search() {
    if (this.state.kaltura.ks) {
      this.setState({
        videos: null
      });//remove videos to render "loading spinner"
      this.getKalturaVideosByNameAndCategory();
    }
  },

  componentWillMount() {
    if (!this.state.kaltura.ks) {
      this.getKalturaSession();
    }
  },

  componentWillUpdate() {
    //console.log('componentWillUpdate: nextState: ', nextState);
  },

  render() {
    return (
      <Container>
        <script src="/client/scripts/vendor/KalturaClient.min.js"></script>
        <VideoLibraryHeader searchText={this.state.searchText} changeSearchText={this.changeSearchText}/>
          <a onClick={this.search}>suchen</a>
        <VideoLibraryContent/>
        <VideoLibraryPreview videos={this.state.videos}/>
      </Container>
    );
  }

});
