// components/button/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Icon from 'components/icon';

const classes = new bem({
  name: 'button',
  prefix: ''
});

export default React.createClass({

  displayName: 'Button',

  propTypes: {
    label: React.PropTypes.string.isRequired,
    modifier: React.PropTypes.string,
    attributes: React.PropTypes.object,
    icon: React.PropTypes.string,
    // target is either a string (for a link)
    // or a function (for a button)
    target: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.func
    ]).isRequired
  },

  render() {

    var content;

    if (typeof this.props.target === 'string') {
      content =
        <a href={this.props.target} {...this.props.attributes} {...classes(null, this.props.modifier)}>
          {this.props.icon ? <Icon id={this.props.icon} /> : ''}
          {this.props.label}
        </a>;
    } else {
      content =
        <button onClick={this.props.target} {...this.props.attributes} {...classes(null, this.props.modifier)}>
          {this.props.icon ? <Icon id={this.props.icon} /> : ''}
          {this.props.label}
        </button>;
    }

    return content;

  }

});
