// components/article-intro/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Intro from 'components/intro';
import MainAsset from 'components/main-asset';
import NewsletterForRegion from 'components/newsletter-for-region';

const classes = new bem({
  name: 'article-intro',
  prefix: ''
});

export default React.createClass({

  displayName: 'ArticleIntro',

  propTypes: {
    intro: React.PropTypes.string.isRequired,
    mainAsset: React.PropTypes.object,
    path: React.PropTypes.string
  },

  getRegionFromPath() {
    // We want to extract aargau/baden from here.
    // for example '/__component__/aargau/baden/artikel-titel'.
    if (this.props.path) {
      return this.props.path.split('/').slice(2, 4);
    }

    return null;
  },

  render() {
    const { intro, mainAsset } = this.props;

    return (
      <div {...classes()}>
        {mainAsset
          ? <MainAsset asset={mainAsset} />
        :null}
        <div {...classes('lower')}>
          <Intro text={intro} />
        </div>
        <NewsletterForRegion regionPath={this.getRegionFromPath()}/>
      </div>
    );
  }

});
