// pages/wireframe/index.jsx
import React from 'react';
import _ from 'lodash';

// import ArticleHeader from 'components/article-header';
// import ArticleDetail from 'components/article-detail';
// import RecommendationsBottom from 'components/recommendations-bottom';
// import RecommendationsTop from 'components/recommendations-top';
import Container from 'components/container';

export default React.createClass({

  displayName: 'WireframePage',

  breakpoint: 1150,

  getInitialState() {
    return {
      // No JS / server.js
      support: false,
      // Initiale state is not mobile
      mobile: false
    };
  },

  // Return the viewport width
  getViewportWidth() {
    return Math.max(document.documentElement.clientWidth, window.innerWidth);
  },

  // Method called when the window gets resized; redefines the `mobile` state
  handleResize() {
    this.setState({
      mobile: this.getViewportWidth() <= this.breakpoint
    });
  },

  componentDidMount() {
    window.addEventListener('resize', _.debounce(this.handleResize, 250));

    this.setState({ support: true });
    this.handleResize();

    // Reperform the check as `setState(..)` is asynchronous
    var isMobile = (this.getViewportWidth() <= this.breakpoint);

    if (isMobile) {
      this.injectSidebar();
    } else {
      this.offsetSidebar();
    }
  },

  render() {

    var imgStyle = {
      width: '100%'
    };

    return (
      <Container>
        <div className='ad'>
          <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=WideboardAd&w=994&h=250&txttrack=0" style={imgStyle} />
        </div>
        <div className='article-header' style={imgStyle}>
          <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=ArticleHeader&w=1108&h=180&txttrack=0" style={imgStyle} />
        </div>
        <div className='wrapper'>
          <div className='main main--flex'>

            <div className='article-intro'>
              <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=ArticleIntro&w=758&h=690&txttrack=0" style={imgStyle} />
            </div>

            <div className='article-body'>
              <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=ArticleBody&w=758&h=1000&txttrack=0" style={imgStyle} />
            </div>

            <div className='recommendations-top'>
              <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=RecommendationsTop&w=332&h=709&txttrack=0" style={imgStyle} />
            </div>

            <div className='ad'>
              <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=WideboardAd&w=994&h=250&txttrack=0" style={imgStyle} />
            </div>

            <div className='recommendations-bottom'>
              <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=RecommendationsBottom&w=758&h=300&txttrack=0" style={imgStyle} />
            </div>

          </div>
          <div className='sidebar'>
            <div className='ad ad--skyscraper'>
              <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=SkyscraperAd&w=160&h=600&txttrack=0" style={imgStyle} />
            </div>
          </div>
        </div>
        <div className='ad'>
          <img src="https://placeholdit.imgix.net/~text?txtsize=26&txt=WideboardAd&w=994&h=250&txttrack=0" style={imgStyle} />
        </div>
      </Container>
    );
  }

});
