// components/user-content-header/block.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Icon from 'components/icon';
import { capitalize } from 'lodash/string';
import { get } from 'lodash/object';

const classes = new bem('ugc-header');

export default React.createClass({

  displayName: 'UGCHeaderBlock',

  propTypes: {
    text : React.PropTypes.string.isRequired,
    linkText : React.PropTypes.string.isRequired,
    image : React.PropTypes.string.isRequired,
    itemProp : React.PropTypes.string,
    imageRounded: React.PropTypes.bool,
    modifier: React.PropTypes.string
  },

  getDefaultProps() {
    return {
      imageRounded: true
    };
  },

  render() {
    var { text, linkText, link, image, itemProp } = this.props;

    var avatarModifiers = {
      [this.props.modifier]: !!this.props.modifier,
      rounded: this.props.imageRounded
    };

    return (
      <span {...classes('block')} itemProp={ itemProp }>
        <span {...classes('text')} >
          { text ? text + "\u00a0" : null }
          <a href={link} {...classes('link')} itemProp="name">
             { linkText }
           </a>
           <img itemProp="image" src={ image } alt="" {...classes('avatar', avatarModifiers)} />
        </span>
      </span>
    )
  }
});
