// components/user-content-header/block.jsx
import React from 'react';
import bem from 'react-bem-helper';

const nbsp = "\u00a0";
const classes = new bem('ugc-header');

export default React.createClass({

  displayName: 'UGCHeaderBlock',

  propTypes: {
    text: React.PropTypes.string.isRequired,
    linkText: React.PropTypes.string.isRequired,
    image: React.PropTypes.string.isRequired,
    itemProp: React.PropTypes.string,
    imageRounded: React.PropTypes.bool,
    modifier: React.PropTypes.string
  },

  getDefaultProps() {
    return {
      imageRounded: true
    };
  },

  renderImage(props, modifiers) {
    if (!props.image) {
      return null;
    }

    let image = <img itemProp="image" src={ props.image } alt={props.linkText} {...classes('avatar', modifiers)} />;

    if (props.link) {
      return (
        <a {...classes('image')} href={props.link}>
          { image }
        </a>
      );
    }

    return (
      <span {...classes('image')}>
        { image }
      </span>
    );
  },

  renderLink(linkText, link) {
    if (link) {
      return (
        <a href={link} {...classes('link')} itemProp="name">
          { linkText }
        </a>
      );
    }

    return (
      <span {...classes('link')} itemProp="name">
       { linkText }
      </span>
    );
  },

  render() {
    var { text, linkText, link, itemProp } = this.props;

    var avatarModifiers = {
      [this.props.modifier]: !!this.props.modifier,
      rounded: this.props.imageRounded
    };

    return (
      <span {...classes('block')} itemProp={ itemProp }>
        <span {...classes('text')} >
          { nbsp }
          { text ? text + nbsp : null }
          { this.renderLink(linkText, link) }
          { this.renderImage(this.props, avatarModifiers) }
        </span>
      </span>
    );
  }
});
