// components/user-content-header/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Container from 'components/container';
import Icon from 'components/icon';
import Block from './block';

const classes = new bem('ugc-header');

export default React.createClass({

  displayName: 'UGCHeader',

  propTypes: {
    icon: React.PropTypes.string,
    modifier: React.PropTypes.string,
    blocks: React.PropTypes.arrayOf(React.PropTypes.object)
  },

  getDefaultProps() {
    return {
      icon: 'megaphone',
      blocks: []
    };
  },

  renderBlocks() {
    return this.props.blocks.map( (block, index) =>
      <Block key={index} {...block} modifier={this.props.modifier} />
    );
  },

  render() {
    if (!this.props.blocks.length) {
      return null;
    }

    return (
      <div {...classes(null, this.props.modifier, 'type-detail')}>
        <Container>
           <div {...classes('container') }>
          { this.props.icon ? <Icon id={this.props.icon} {...classes('icon')} /> : null }
          { this.renderBlocks() }
          </div>
        </Container>
      </div>
    );
  }
});
