// components/user-content-header/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Icon from 'components/icon';
import { capitalize } from 'lodash/string';

const classes = new bem('ugc-header');

export default React.createClass({

  displayName: 'UGCHeader',

  propTypes: {
    type: React.PropTypes.oneOf(['leserbeitrag', 'vereinsmeldung', 'sportverein', 'gemeindebeitrag']).isRequired,
    city: React.PropTypes.shape({
      name: React.PropTypes.string.isRequired,
      url: React.PropTypes.string.isRequired,
      shield_url: React.PropTypes.string
    }),
    author: React.PropTypes.shape({
      name: React.PropTypes.string.isRequired,
      url: React.PropTypes.string.isRequired,
      image: React.PropTypes.string
    })
  },

  renderCity() {
    var { name, image, url } = this.props.city;
    return (
      <span {...classes('block')} itemProp="contentLocation">
        { 'aus ' }
        <a href={url}>
          <span itemProp="name">{ name }</span>
          <img itemProp="logo" src={ image } alt="" {...classes('avatar', 'city')} />
        </a>
      </span>
    )
  },

  renderAuthor() {
    var { name, image, url } = this.props.author;

    return (
      <span {...classes('block')} itemProp="author">
        { 'von ' }
        <a href={url}>
         <span itemProp="name"> { name }</span>
        <img itemProp="image" src={ image } alt="" {...classes('avatar', 'author')} />
        </a>
      </span>
    )
  },

  render() {
    var { type } = this.props;
    var isAuthorFirst = type === 'leserbeitrag';

    if (!type) {
      return null;
    }

    return (
      <div {...classes({ extra: 'type-detail'})}>
        <Icon {...classes('icon')} id='megaphone' />
        { capitalize(type) + ' ' }
        { this.renderCity() }
        { this.renderAuthor() }
      </div>
    );
  }

});
