// components/cover/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import ResponsiveImage from 'components/responsive-image';
import BackgroundImage from 'components/background-image';
import Icon from 'components/icon';

const classes = new bem('cover');

export default React.createClass({

  displayName: 'Cover',

  propTypes: {
    variation: React.PropTypes.string.isRequired,
    title: React.PropTypes.object.isRequired,
    catchword: React.PropTypes.string,
    link: React.PropTypes.string.isRequired,
    linkText: React.PropTypes.string,
    image: React.PropTypes.object,
    background: React.PropTypes.string
  },

  render() {

    let { variation, title, catchword, link, linkText, image, background } = this.props;

    return (
      <article {...classes(null, variation)}>
        <a {...classes('link')} href={link}>
          <div {...classes('inner')}>
            {catchword
              ? <p {...classes('catchword')}>{catchword}</p>
            :null}
            <h1 {...classes('title')}>{title}</h1>
            {image
              ? (
                <div {...classes('image')}>
                  <ResponsiveImage {...image} fill={true} rounded={true} />
                </div>
              )
            :null}
            {linkText
              ? <p {...classes('link-text')}>
                  <span>{linkText}</span>
                  <Icon id='arrow-right' {...classes('icon')} />
                </p>
            :null}
          </div>
          <BackgroundImage background={background} theme={variation} />
        </a>
      </article>

    );
  }

});
