// components/category-overview/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Container from 'components/container';
import Teaser from 'components/teaser';
import Cover from 'components/cover';
import Grid from 'components/grid';

const classes = new bem({
  name: 'category-overview',
  prefix: ''
});


export default React.createClass({

  displayName: 'CategoryOverview',

  propTypes: {
    teasers: React.PropTypes.array.isRequired,
    theme: React.PropTypes.string,
    component: React.PropTypes.string
  },

  getComponentType(type) {
    const types = {
      'Teaser': Teaser,
      'Cover': Cover
    };

    return types[type];
  },

  getDefaultProps() {
    return {
      theme: 'meinung',
      component: 'Teaser'
    };
  },

  mapGridComponents(teasers, type) {
    return teasers.map((teaser) => {
      let Component = this.getComponentType(type);
      return (
        <Component {...teaser} />
      );
    });
  },

  render() {

    const { teasers, theme, component } = this.props;

    return (
      <section {...classes('', theme)}>
        <Container>
          <Grid items={this.mapGridComponents(teasers, component)} modifier={'three-up-simple'} />
        </Container>
      </section>
    );
  }

});
