// components/comments/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getScript } from 'helpers';
import Button from 'components/button';

const classes = new bem({
  name: 'comments',
  prefix: ''
});

const DISQUS_SHORTNAME = process.env.DISQUS_SHORTNAME || 'azmedien-staging';

export default React.createClass({

  displayName: 'Comments',

  getInitialState() {
    return {
      loaded: false
    };
  },

  loadDisqus() {
    var disqusURL = '//' + DISQUS_SHORTNAME + '.disqus.com/embed.js';
    getScript(disqusURL);
    // immediatly set state to loaded as Disqus needs DOM node present at load
    this.setState({ loaded: true });
  },

  render() {
    var contentJSX;

    if (this.state.loaded) {
      contentJSX = <div id='disqus_thread'></div>;
    } else {
      contentJSX = (
        <Button {...classes('btn')}
          _event={this.loadDisqus}
          modifier='huge'
          label='Kommentare anzeigen'
        />
      );
    }

    return (
      <section ref='comments' {...classes()}>
        {contentJSX}
      </section>
    );

  }

});
