// components/typography/index.jsx
'use strict';

const React = require('react');
const TypographyItem = require('styleguide/typography-item');
const bem = require('react-bem-helper');

const classes = new bem({
  name: 'typography',
  prefix: 'styleguide-'
});

var data = [
  {
    props: {
      title: 'Primary headline',
      description: 'This is intended to be used for the main headline, such as article headline.',
      meta: '48px / 1.125 $headline-font-stack'
    },
    content: <h1 {...classes('example', null, 'type-h1')}>Staatsanwaltschaft ermittelt nach Brückeneinsturz gegen Unbekannt</h1>
  },
  {
    props: {
      title: 'Secondary headline',
      description: 'This is intended to be used when <code>type-h1</code> is a bit too <em>in yo face</em>.',
      meta: '32px / 1.25 $headline-font-stack'
    },
    content: <h2 {...classes('example', null, 'type-h2')}>Staatsanwaltschaft ermittelt nach Brückeneinsturz gegen Unbekannt</h2>
  },
  {
    props: {
      title: 'Tertiary headline',
      description: 'This is intended to be used when <code>type-h2</code> is a bit too <em>in yo face</em>.',
      meta: '24px / 1.25 $headline-font-stack'
    },
    content: <h3 {...classes('example', null, 'type-h3')}>Staatsanwaltschaft ermittelt nach Brückeneinsturz gegen Unbekannt</h3>
  },
  {
    props: {
      title: 'Mini headline',
      description: 'This is intended to style titles that are part of secondary content, for instance related content within an article.',
      meta: '18px / 1.3333 $headline-font-stack'
    },
    content: <h4 {...classes('example', null, 'type-h4')}>Staatsanwaltschaft ermittelt nach Brückeneinsturz gegen Unbekannt</h4>
  },
  {
    props: {
      title: 'Introduction paragraph',
      description: 'This is intended to be an introduction paragraph to an article. It should not be used for more than a single paragraph in a row.',
      meta: '24px / 1.41 $supporting-font-stack'
    },
    content: <p {...classes('example', null, 'type-intro')}>Vier Tage lang haben drei Schneeschuhwanderer aus Tschechien im Salzburger Pinzgau im Winterraum einer Hütte in 2100 Metern Seehöhe ausgeharrt, weil sie nach einem massiven Wetterumschwung nicht mehr absteigen konnten.</p>
  },
  {
    props: {
      title: 'Copy text',
      description: 'This is intended to be the default typographic styles for paragraphs, lists and basically the core content of an article.',
      meta: '22px / 1.41 $text-font-stack'
    },
    content: <p {...classes('example', null, 'type-copy')}>Die Männer seien am vergangen Freitagmorgen zu der Tour aufgebrochen, die sie zum Ingolstädterhaus auf 2119 Meter führen sollte, teilte die Polizei mit. Gegen 23.00 Uhr erreichten sie die im Winter nicht bewirtschaftete Hütte und quartierten sich im Winterraum ein, der frei zugänglich ist, über zwölf Schlafplätze, aber keine Verpflegung verfügt.</p>
  },
  {
    props: {
      title: 'Alternate copy text',
      description: 'To be determined.',
      meta: '22px / 1.41 $supporting-font-stack'
    },
    content: <p {...classes('example', null, 'type-copy-alternate')}>Die Redaktion sucht nach Augenzeugen, die am vergangenen Dienstag den Unfall am Marktplatz in Aarau beobachet haben. Ein schwarzer Mercedes war an einen Poller geprallt und hatte sich überschlagen, 2 Passanten wurden leicht verletzt.</p>
  },
  {
    props: {
      title: 'Blockquote',
      description: 'This is intended to style blockquotes. It should not be used on inline quotes.',
      meta: '32px / 38px $headline-font-stack'
    },
    content: <blockquote {...classes('example', null, 'type-blockquote')}>«Am Sonntag wurden die Tschechen im Alter von 32, 35 und 38 Jahren dann vom Wetterumschwung überrascht»</blockquote>
  },
  {
    props: {
      title: 'Tiny header',
      description: 'This is intended to be a word, phrase or name to give context to an article or a quote.',
      meta: '13px / 1.23 $supporting-font-stack'
    },
    content: <h2 {...classes('example', null, 'type-tiny-header')}>Deutschland</h2>
  },
  {
    props: {
      title: 'Caption',
      description: 'This is intended to be used for captions such as under an image or video.',
      meta: '16px / 1.375 $supporting-font-stack'
    },
    content: <p {...classes('example', null, 'type-caption')}>Zunzger Kinder im Fasnachtsfieber</p>
  },
  {
    props: {
      title: 'Detail information',
      description: 'This is intended to be used for tiny details, such as secondary labels or meta data on articles like the author name or the date.',
      meta: '12px / 1.3333 $supporting-font-stack'
    },
    content: <p {...classes('example', null, 'type-detail')}>Abonnieren</p>
  }
];

const Typography = React.createClass({

  render: function() {
    return (
      <div>
        {data.map((section, index) =>
          <TypographyItem key={index} {...section.props}>
            {section.content}
          </TypographyItem>
        )}
      </div>
    );
  }

});

module.exports = Typography;
