// components/tabbed-navigation/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Icon from 'components/icon';

const classes = new bem({
  name: 'tabbed-navigation',
  prefix: ''
});

export default React.createClass({

  displayName: 'TabbedNavigation',

  propTypes: {
    showing: React.PropTypes.string.isRequired,
    onShowLatest: React.PropTypes.func.isRequired,
    onShowPopular: React.PropTypes.func.isRequired
  },

  render() {

    let { showing, onShowLatest, onShowPopular } = this.props;

    let latestModifiers  = { 'is-selected': showing === 'latest' };
    let popularModifiers = { 'is-selected': showing === 'popular' };

    return (
      <ul {...classes()}>
        <li {...classes('item', 'latest', latestModifiers)}>
          <button {...classes('button', 'latest', latestModifiers)} onClick={onShowLatest}>
            <Icon {...classes('icon')} id='bullet' />
            Neueste <span {...classes('detailed-label')}>Meldungen</span>
          </button>
        </li>
        <li {...classes('item', 'popular', popularModifiers)}>
          <button {...classes('button', 'popular', popularModifiers)} onClick={onShowPopular}>
            <Icon {...classes('icon')} id='views' />
            Beliebste <span {...classes('detailed-label')}>Meldungen</span>
          </button>
        </li>
      </ul>
    );
  }

});
