// components/tabbed-navigation/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Icon from 'components/icon';

const classes = new bem({
  name: 'tabbed-navigation',
  prefix: ''
});

export default React.createClass({

  displayName: 'TabbedNavigation',

  propTypes: {
    tabs: React.PropTypes.oneOfType([
      React.PropTypes.array,
      React.PropTypes.object
    ]).isRequired,
    showingIndex: React.PropTypes.number,
    handleSelect: React.PropTypes.func.isRequired
  },

  getInitialState() {
    return { showing: this.props.showing };
  },

  handleClick(index) {
    this.props.handleSelect(index);
  },

  render() {

    const { tabs, showingIndex } = this.props;

    if (!tabs || !tabs.length) {
      return null;
    }

    return (
      <ul {...classes()}>
        {tabs.map((tab, index) =>
          <li {...classes('item', null, [showingIndex === index ? 'is-selected' : null])} key={index}>
            <button {...classes('button', 'latest')} onClick={this.handleClick.bind(this, index)}>
              {tab.icon
                ? <Icon {...classes('icon')} id={tab.icon} />
              :null}

              {tab.title}

              {tab.detailedLabel
                ? <span {...classes('detailed-label')}> {tab.detailedLabel}</span>
              :null}
            </button>
          </li>
        )}
      </ul>
    );
  }

});
