// components/enrichment-image/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import DataMap from 'higher-order/data/enrichment-image';
import ResponsiveImage from 'components/responsive-image';
import Caption from 'components/caption';

const classes = new bem({
  name: 'enrichment-image',
  prefix: ''
});

const EnrichmentImage = React.createClass({

  propTypes: {
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    url: React.PropTypes.string.isRequired,
    size: React.PropTypes.string.isRequired,
    source: React.PropTypes.string
  },

  render() {
    return (
      <figure itemScope itemType='http://schema.org/ImageObject' {...classes()}>
        <ResponsiveImage
          ratio='16x9'
          size={this.props.size}
          alt={this.props.description}
          title={this.props.title}
          src={this.props.url} />
        <Caption
          title={this.props.title}
          description={this.props.description}
          source={this.props.source} />
      </figure>
    );
  }

});

export default DataMap(EnrichmentImage);
