// components/enrichment-image/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import ResponsiveImage from 'components/responsive-image';
import Caption from 'components/caption';

const classes = new bem({
  name: 'enrichment-image',
  prefix: ''
});

export default React.createClass({

  displayName: 'EnrichmentImage',

  propTypes: {
    asset: React.PropTypes.object,
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    plaement: React.PropTypes.string
  },

  mappedData() {

    var props = this.props;
    var asset = props.asset;
    var size = this.props.placement && ['left', 'right'].indexOf(this.props.placement) > -1
      ? 'small'
      : 'max';

    return {
      title: props.title || asset.title,
      description: props.description || asset.description,
      url: asset.image_url,
      size: size,
      source: asset.source.title || (asset.source.source ? asset.source.source.title : null)
    };

  },

  render() {
    return (
      <figure itemScope itemType='http://schema.org/ImageObject' {...classes()}>
        <ResponsiveImage
          ratio='16x9'
          size={this.mappedData().size}
          alt={this.mappedData().description}
          title={this.mappedData().title}
          src={this.mappedData().url} />
        <Caption
          title={this.mappedData().title}
          description={this.mappedData().description}
          source={this.mappedData().source} />
      </figure>
    );
  }

});
