// components/enrichment-image/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getEnrichmentPlacement } from 'helpers';
import ImageAsset from 'components/image-asset';

const classes = new bem({
  name: 'enrichment-image',
  prefix: ''
});

export default React.createClass({

  displayName: 'EnrichmentImage',

  propTypes: {
    asset: React.PropTypes.object,
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    image_url: React.PropTypes.string,
    placement: React.PropTypes.string,
    stoerer_type: React.PropTypes.string
  },

  render() {
    var placement = getEnrichmentPlacement(this.props.placement);

    var itemProp = this.props.stoerer_type === 'image_stoerer'
      ? 'associatedMedia'
      : this.props.asset.content_type === 'asset_image'
        ? 'image'
        :null;

    return (
      <div {...classes(null, placement)}>
        <ImageAsset asset={this.props} size={placement === 'full' ? 'full' : 'small'} ratio="16x9" itemProp={itemProp}/>
      </div>
    );
  }

});
