// components/latest-articles/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import _ from 'lodash';

import Container from 'components/container';
import SectionHeader from 'components/section-header';
import TabbedNavigation from 'components/tabbed-navigation';
import LatestArticlesSection from 'components/latest-articles-section';

const classes = new bem({
  name: 'latest-articles',
  prefix: ''
});


const sections = [
  {
    id: 'latest',
    icon: 'bullet',
    title: 'Neueste',
    detailedLabel: 'Meldungen'
  },
  {
    id: 'popular',
    icon: 'views',
    title: 'Neueste',
    detailedLabel: 'Meldungen'
  }
];

export default React.createClass({

  displayName: 'LatestArticles',

  propTypes: {
    latest: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    popular: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  getInitialState() {
    return { showing: _.first(sections).id };
  },

  showSection(id) {
    this.setState({ showing: id });
  },

  render() {

    let { showing } = this.state;

    return (
      <section {...classes()}>
        <Container>
          <SectionHeader title='Halten Sie sich auf dem Laufenden' />
          <TabbedNavigation
            tabs={sections}
            showing={showing}
            handleSelect={this.showSection}
          />
          <LatestArticlesSection items={this.props[showing]} />
        </Container>
      </section>
    );

  }

});
