// components/latest-articles/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Container from 'components/container';
import SectionHeader from 'components/section-header';
import TabbedNavigation from 'components/tabbed-navigation';
import LatestArticlesSection from 'components/latest-articles-section';

const classes = new bem({
  name: 'latest-articles',
  prefix: ''
});

export default React.createClass({

  displayName: 'LatestArticles',

  propTypes: {
    latest: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    popular: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  getInitialState() {
    return {
      showing: 'latest'
    };
  },

  showLatest() {
    this.setState({ showing: 'latest' });
  },

  showPopular() {
    this.setState({ showing: 'popular' });
  },

  render() {

    let { showing } = this.state;

    return (
      <section {...classes()}>
        <Container>
          <SectionHeader title='Halten Sie sich auf dem Laufenden' />
          <TabbedNavigation
            showing={showing}
            onShowLatest={this.showLatest}
            onShowPopular={this.showPopular}
          />
          <LatestArticlesSection items={this.props[showing]} />
        </Container>
      </section>
    );

  }

});
