// components/latest-articles/wrapper.jsx
import React from 'react';
import _ from 'lodash';

import { getTeaserFromArticle, getTeaserSwitchableFromArticle } from 'helpers/teaser';

const title = 'Halten Sie sich auf dem Laufenden';

const tabs = [
  {
    icon: 'bullet',
    title: 'Neueste',
    detailedLabel: 'Meldungen'
  },
  {
    icon: 'views',
    title: 'Popular',
    detailedLabel: 'Meldungen'
  }
];

export default Component => React.createClass({

  displayName: 'LatestArticlesWrapper',

  propTypes: {
    articles: React.PropTypes.oneOfType([
      React.PropTypes.array,
      React.PropTypes.object
    ]).isRequired
  },

  mappedTeasers(teasers) {
    return teasers.map((teaser, index) => {
      let teaserData;
      if (index === 0) {
        teaserData = getTeaserSwitchableFromArticle({
          article: teaser,
          variation: 'listed-to-summary'
        });
      } else {
        teaserData = getTeaserFromArticle({
          article: teaser,
          variation: 'listed'
        });
      }
      return teaserData;
    });
  },

  mapArticlesToSections(articles, sections) {
    return articles.map((articlesArray, index) => {
      let section = sections[index];
      let teasers = this.mappedTeasers(articlesArray);
      let merged = _.assign({}, section, { teasers: teasers });
      return merged;
    });
  },

  mappedData(props) {
    return {
      title: title,
      sections: this.mapArticlesToSections(props.articles.slice(0, 9), tabs)
    };
  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
