// components/latest-articles-section/index.jsx
import _ from 'lodash';

import React from 'react';
import bem from 'react-bem-helper';

import Teaser from 'components/teaser';
import TeaserSwitchable from 'components/teaser-switchable';

const classes = new bem({
  name: 'latest-articles-section',
  prefix: ''
});

export default React.createClass({

  displayName: 'LatestArticlesSection',

  propTypes: {
    items: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
  },

  renderFeaturedTeaser(teasers) {
    let teaser = _.first(teasers);
    return (
      <div {...classes('teaser-wrap', 'featured')}>
        <TeaserSwitchable {...teaser} />
      </div>
    );
  },

  renderTeaserList(teasers) {
    return teasers.slice(1).map((teaser, index) =>
      <div {...classes('list-item')} key={index}>
        <div {...classes('teaser-wrap', 'list')}>
          <Teaser {...teaser} />
        </div>
      </div>
    );
  },

  render() {

    const { items } = this.props;

    return (
      <div {...classes()}>
        <div {...classes('grid')}>
          <div {...classes('grid-item', 'featured')}>
            {this.renderFeaturedTeaser(items)}
          </div>
          <div {...classes('grid-item', 'list')}>
            <div {...classes('grid', 'list')}>
              {this.renderTeaserList(items)}
            </div>
          </div>
        </div>
      </div>
    );
  }

});
