// legacy/helpers/assets.jsx
// contains the Relation and Asset classes stored in the registry
import { registerContentType, getImplementation } from "./registry.jsx";

const KALTURA_CONFIG = {
  cropType: 3,                                // should remain at 3
  partnerId: process.env.KALTURA_ACCOUNT_ID   // from existing config, e.g. 1719221
};

/**
 * base class from which all Assets derive
 */
export class BaseAsset {
  constructor(context, contentType, canBeTeaserImage=true) {
    this.context = context;
    this.contentType = contentType;
    this._canBeTeaserImage = canBeTeaserImage;
  }

  get canBeTeaserImage() {
    return this._canBeTeaserImage;
  }

  set canBeTeaserImage(val) {
    this._canBeTeaserImage = val;
  }

  get imageUrl() {
    return '';
  }
}

/**
 * base class from which all Relations derive
 */
export class BaseRelation {
  constructor(context) {
    this.context = context;
    this._canBeTeaserImage = true;
  }

  // right now, only imageGallery calls the base class implementation
  get imageUrl() {
    return this._getAsset().imageUrl;
  }

  // all Relations so far can be a teaserImage, except for htmlsnippet, document and infographic
  get canBeTeaserImage() {
    return true;
  }

  set canBeTeaserImage(val) {
    this._canBeTeaserImage = val;
  }

  get contentType() {
    return this.context.asset.content_type;
  }

  get asset() {
    return this._getAsset();
  }

  _getAsset() {
    if (!this._assetImpl) { // lazy impl ...
      // call the actual implementation: in case of imageGallery: images[0]
      this._assetImpl = getImplementation(this.context.asset);
    }
    return this._assetImpl;
  }
}

export class RelationImage extends BaseRelation {
  get imageUrl() {
    return this.context.image_url;
  }
}
registerContentType(RelationImage, 'asset_image_relation');

export class RelationImageGallery extends BaseRelation {
  // the only Relation that asks the ImageGallery object how to get the image_url (from images[0])
}
registerContentType(RelationImageGallery, 'asset_image_gallery_relation');

/*
def kalturaImageUrl(kalturaId, variant, request):
       variant = VARIANTS['kalturaTeaser'].get(variant)
       if not variant:
           return None
       settings = kalturaSettings(request)
       cropType = 3  # crop according to the given dimensions while
                     # maintaining the original aspect ratio. The
                     # resulting image may be cover only part of the
                     # original image.
                     # http://knowledge.kaltura.com/kaltura-thumbnail-api
       return 'http://cdnbakmi.kaltura.com/p/%(partnerId)s/' \
              'sp/%(partnerId)s00/thumbnail/entry_id/%(entryId)s/' \
              'width/%(width)s/height/%(height)s/type/%(type)s' % dict(
                  partnerId=settings['partnerId'],
                  entryId=kalturaId,
                  width=variant['w'],
                  height=variant['h'],
                  type=cropType
              )
              */

function getKalturaImageUrl(kalturaId, size) {
  // kalturaId e.g.: '1_7zywfjdn'

  // crop according to the given dimensions while maintaining the original aspect ratio.
  // The resulting image may be only covers part of the original image.
  // http://knowledge.kaltura.com/kaltura-thumbnail-api
  // working sample: 'http://cdnbakmi.kaltura.com/p/1789881/sp/178988100/thumbnail/entry_id/1_7zywfjdn/width/220/height/130/type/3'
  let url = `http://cdnbakmi.kaltura.com/p/${KALTURA_CONFIG.partnerId}/sp/${KALTURA_CONFIG.partnerId }00/thumbnail/entry_id/${kalturaId}/width/${size.width}/height/${size.height}/type/${KALTURA_CONFIG.cropType}`;
  return url;
}

export class RelationVideo extends BaseRelation {
  getImageUrl(size) {
    // current logic is: pick image_url from still_image.asset if exists, otherwise try kaltury_id ...
    if (this.context.asset && this.context.asset.still_image && this.context.asset.still_image.image_url && this.context.asset.still_image.image_url.length > 0) {
      return this.context.asset.still_image.image_url;
    }
    if (this.context.asset && this.context.asset.kaltura_id && this.context.asset.kaltura_id.length > 0) {
      return getKalturaImageUrl(this.context.asset.kaltura_id, size);
    }
    return "";
  }
}
registerContentType(RelationVideo, 'asset_video_relation');

export class RelationAudio extends BaseRelation {
  get imageUrl() {
    if (this.context.asset && this.context.asset.still_image && this.context.asset.still_image.length > 0) {
      return this.context.asset.still_image;
    }
    return this.context.image_url;
  }
}
registerContentType(RelationAudio, 'asset_audio_relation');

export class RelationHtmlSnippet extends BaseRelation {
  get canBeTeaserImage() {
    return false;
  }
  get imageUrl() {
    return "";
  }
}
registerContentType(RelationHtmlSnippet, 'asset_htmlsnippet_relation');

export class RelationSurvey extends BaseRelation {
  get imageUrl() {
    if (this.context.asset && this.context.asset.teaser) {
      return this.context.asset.teaser.image_url;
    }
  }
}
registerContentType(RelationSurvey, 'asset_survey_relation');

export class RelationQuiz extends BaseRelation {
  get imageUrl() {
    if (this.context.asset && this.context.asset.teaser) {
      return this.context.asset.teaser.image_url;
    }
  }
}
registerContentType(RelationQuiz, 'asset_quiz_relation');

// and now for the the asset classes ...

// dummy skeleton, RelationVideo does the logic since the image_url?process is picked up from relation ...
export class Image extends BaseAsset {
  constructor(context) {
    super(context, 'asset_image');
  }

  get imageUrl() {
    return this.context.image_url;
  }
}
registerContentType(Image, 'asset_image');

// in use
export class ImageGallery extends BaseAsset {
  constructor(context) {
    super(context, 'asset_image_gallery');
  }
  get canBeTeaserImage() {
    // only galleries with more than one image are eligible to serve as teaser
    if (this.context.images.length > 1) {
      return true;
    }
  }
  get imageUrl() {
    if (this.context.images.length > 1 && this.context.images[0]) {
      return this.context.images[0].image_url;
    }
    return null;
  }
}
registerContentType(ImageGallery, 'asset_image_gallery');

// dummy skeleton, RelationVideo does the logic since image_url?process is picked up from relation ...
export class Video extends BaseAsset {
  constructor(context) {
    super(context, 'asset_video');
  }
  get imageUrl() {
    return "";
  }
}
registerContentType(Video, 'asset_video');

// dummy skeleton, RelationAudio does the logic since image_url?process is picked up from relation ...
export class Audio extends BaseAsset {
  constructor(context) {
    super(context, 'asset_audio');
  }
  get imageUrl() {
    return "";
  }
}
registerContentType(Audio, 'asset_audio');

// dummy skeleton, RelationHtmlSnippet does the logic since image_url?process is picked up from relation ...
export class HtmlSnippet extends BaseAsset {
  constructor(context) {
    super(context, 'asset_htmlsnippet', false);
  }

  get imageUrl() {
    return "";
  }
}
registerContentType(HtmlSnippet, 'asset_htmlsnippet');

// dummy skeleton, RelationSurvey does the logic since image_url?process is picked up from relation ...
export class Survey extends BaseAsset {
  constructor(context) {
    super(context, 'asset_survey');
  }
  get imageUrl() {
    return this.context.asset.teaser.image_url;
  }
}
registerContentType(Survey, 'asset_survey');

// dummy skeleton, RelationQuiz does the logic since image_url?process is picked up from relation ...

export class Quiz extends BaseAsset {
  constructor(context) {
    super(context, 'asset_quiz');
  }
}
registerContentType(Quiz, 'asset_quiz');
