// legacy/helpers/relations.jsx
// contains the Relation and Asset classes stored in the registry
import { registerContentType, getImplementation } from './registry.jsx';
import {createImageSrc} from '../../../../common/image';


/**
 * base class from which all Relations derive
 */
export class BaseRelation {
  constructor(context) {
    this.context = context;
    this._canBeTeaserImage = true;
    this._assetImpl = null;
  }

  get canBeTeaserImage() {
    return this.asset ? this.asset.canBeTeaserImage : false;
  }

  getImageUrl(size) {
    return this.asset ? this.asset.getImageUrl(size) : false;
  }

  get contentType() {
    return this.asset ? this.asset.contentType : false;
  }

  get asset() {
    if (!this._assetImpl) { // lazy impl ...
      // call the actual implementation for the context
      this._assetImpl = getImplementation(this.context.asset);
    }
    return this._assetImpl;
  }
}


export class RelationImage extends BaseRelation {

  get canBeTeaserImage() {
    return true;
  }

  getImageUrl(size) {
    return createImageSrc(this.context.image_url,
                          size.width,
                          size.height,
                          false);
  }
}

registerContentType(RelationImage, 'asset_image_relation');
registerContentType(BaseRelation, 'asset_image_gallery_relation');
registerContentType(BaseRelation, 'asset_video_relation');
registerContentType(BaseRelation, 'asset_audio_relation');
registerContentType(BaseRelation, 'asset_htmlsnippet_relation');
registerContentType(BaseRelation, 'asset_survey_relation');
registerContentType(BaseRelation, 'asset_quiz_relation');
