// legacy/helpers/relations.jsx
// contains the Relation and Asset classes stored in the registry
import { registerContentType, getImplementation, getRelationFor } from './registry.jsx';
import KalturaImageProxy from 'helpers/kaltura-image-proxy';

const KALTURA_CONFIG = {
  cropType: 3,                                // should remain at 3
  partnerId: process.env.KALTURA_ACCOUNT_ID   // from existing config, e.g. 1719221
};


/**
 * base class from which all Relations derive
 */
export class BaseRelation {
  constructor(context) {
    this.context = context;
    this._canBeTeaserImage = true;
    this._assetImpl = null;
  }

  getImageUrl(size) {
    return this.asset.getImageUrl(size);
  }

  // all Relations so far can be a teaserImage, except for htmlsnippet, document and infographic
  get canBeTeaserImage() {
    return true;
  }

  set canBeTeaserImage(val) {
    this._canBeTeaserImage = val;
  }

  get contentType() {
    return this.asset.contentType;
  }

  get asset() {
    if (!this._assetImpl) { // lazy impl ...
      // call the actual implementation for the context
      this._assetImpl = getImplementation(this.context.asset);
    }
    return this._assetImpl;
  }
}


export class RelationImage extends BaseRelation {
}
registerContentType(RelationImage, 'asset_image_relation');


export class RelationImageGallery extends BaseRelation {
}
registerContentType(RelationImageGallery, 'asset_image_gallery_relation');

/*
def kalturaImageUrl(kalturaId, variant, request):
       variant = VARIANTS['kalturaTeaser'].get(variant)
       if not variant:
           return None
       settings = kalturaSettings(request)
       cropType = 3  # crop according to the given dimensions while
                     # maintaining the original aspect ratio. The
                     # resulting image may be cover only part of the
                     # original image.
                     # http://knowledge.kaltura.com/kaltura-thumbnail-api
       return 'http://cdnbakmi.kaltura.com/p/%(partnerId)s/' \
              'sp/%(partnerId)s00/thumbnail/entry_id/%(entryId)s/' \
              'width/%(width)s/height/%(height)s/type/%(type)s' % dict(
                  partnerId=settings['partnerId'],
                  entryId=kalturaId,
                  width=variant['w'],
                  height=variant['h'],
                  type=cropType
              )
              */

/**
 * if still_images holds an Image, get it from there. if kaltura_id exists, return the kaltura url.
 */
export class RelationVideo extends BaseRelation {
  getImageUrl(size) {
    let relObj = null;
    if (this.context.asset.still_image && this.context.asset.still_image.asset) {
      relObj = getRelationFor(this.context.asset.still_image.asset, this.context.asset.still_image);
    }
    if (relObj && relObj.canBeTeaserImage) {
      return relObj.getImageUrl(size);
    }
    if (this.context.asset && this.context.asset.kaltura_id && this.context.asset.kaltura_id.length > 0) {
      var proxy = new KalturaImageProxy(KALTURA_CONFIG.partnerId, true, this.context.asset.kaltura_id);
      return proxy.getImageURL(size.width, size.height, true);
    }
    return "";
  }
}
registerContentType(RelationVideo, 'asset_video_relation');

/**
 * only return an image url if still_image exists
 */
export class RelationAudio extends BaseRelation {
  getImageUrl(size) {
    let relObj = null;
    if (this.context.asset.still_image && this.context.asset.still_image.asset) {
      relObj = getRelationFor(this.context.asset.still_image.asset, this.context.asset.still_image);
    }
    if (relObj && relObj.canBeTeaserImage) {
      return relObj.getImageUrl(size);
    }
    return "";
  }
}
registerContentType(RelationAudio, 'asset_audio_relation');


export class RelationHtmlSnippet extends BaseRelation {
  get canBeTeaserImage() {
    return false;
  }

  getImageUrl() {
    return "";
  }
}
registerContentType(RelationHtmlSnippet, 'asset_htmlsnippet_relation');


export class RelationSurvey extends BaseRelation {
  getImageUrl(size) {
    let relObj = null;
    if (this.context.asset && this.context.asset.teaser && this.context.asset.teaser.asset) {
      relObj = getRelationFor(this.context.asset.teaser.asset, this.context.asset.teaser);
    }
    if (relObj && relObj.canBeTeaserImage) {
      return relObj.getImageUrl(size);
    }
    return "";
  }
}
registerContentType(RelationSurvey, 'asset_survey_relation');


export class RelationQuiz extends BaseRelation {
  getImageUrl(size) {
    let relObj = null;
    if (this.context.asset && this.context.asset.teaser && this.context.asset.teaser.asset) {
      relObj = getRelationFor(this.context.asset.teaser.asset, this.context.asset.teaser);
    }
    if (relObj && relObj.canBeTeaserImage) {
      return relObj.getImageUrl(size);
    }
    return "";
  }
}
registerContentType(RelationQuiz, 'asset_quiz_relation');
