// components/recommendations-top/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Teaser from 'components/teaser';

const classes = new bem({
  name: 'recommendations-top',
  prefix: ''
});

export default React.createClass({

  displayName: 'RecommendationsTop',

  propTypes: {
    teasers: React.PropTypes.array.isRequired
  },

  getInitialState() {
    return {
      mode: 'collapsed'
    };
  },

  expand() {
    this.setState({
      mode: 'expanded'
    });
  },

  render() {

    var { teasers } = this.props;
    var items = [], i = 0, len = 6;
    while (++i <= len) items.push(i);

    return (
      <div {...classes(null, this.state.mode)}>
        <h3 {...classes('title')}>Artikel zum Thema</h3>
        <div {...classes('grid')}>
          {items.map(function (n) {
            return (
              <div {...classes('cell', 'inlined-right')} key={n-1}>
                <Teaser key={n-1} {...teasers[n-1]} />
              </div>
            );
          })}
        </div>
        <button onClick={this.expand} {...classes('button')}>
          <span>Mehr Artikel</span>
        </button>
      </div>
    );

  }

});
