// components/weather-forecast-list/index.jsx
import React from 'react';
import BEM from 'react-bem-helper';
import _ from 'lodash';
import { getAssetPath } from 'helpers';

import Icon from 'components/icon';
import Container from 'components/container';

const classes = new BEM('weather-forecast-list');

export default React.createClass({

  displayName: 'WeatherForecastList',

  propTypes: {
    forecast: React.PropTypes.object.isRequired
  },

  getWeatherIcon(symbol) {
    return (
      <img {...classes('img')} src={getAssetPath('images/weather/color/' + symbol + '.svg')} />
    );
  },

  getDirectionModifier(direction) {
    var modifiers = {
      n: "north",
      nw: "north-west",
      sw: "south-west",
      s: "south",
      so: "south-east",
      o: "east",
      no: "north-east"
    };
    return _.get(modifiers, direction.toLowerCase());
  },

  render() {

    let allForecastData = this.props.forecast.days || [];

    let forecastList = allForecastData.map((item, index) => {

      let classy = {};

      classy[this.getDirectionModifier(item.Dir)] = true;
      classy.icon = true;

      return (
        <tr key={index}>
          <td {...classes('time')}>00:00</td>
          <td>{this.getWeatherIcon(item.Symb)}</td>
          <td {...classes('group')}>
            <span {...classes('desc')}>{item.Txt}</span>
            <span {...classes('temp')}>{item.Temp}°</span>
            <div>222</div>
          </td>
          <td {...classes('wind')}><Icon id="arrow-circle"  {...classes(null, null, classy)} />{item.Force}km/h</td>
        </tr>
      );
    });

    return (
      <Container>
        <table {...classes('table')}>
          <tbody>
            {forecastList}
          </tbody>
        </table>
      </Container>
    );
  }
});
