// styleguide/contextual-navigation/index.jsx
const React = require('react');
const _ = require('lodash');
const bem = require('react-bem-helper');
const path = require('path');
const slugToTitle = require('slug-to-title');

const classes = new bem({
  name: 'contextual-navigation',
  prefix: 'styleguide-'
});

const ContextualNavigation = React.createClass({

  propTypes: {
    components: React.PropTypes.array.isRequired,
    active: React.PropTypes.string
  },

  findComponents() {
    var component = _.find(this.props.components, function(obj) {
      return obj.slug === this.props.active;
    }.bind(this));

    var index = this.props.components.indexOf(component);
    var prevIndex = index - 1 < 0 ? null : index - 1;
    var nextIndex = index + 1 > this.props.components.length - 1 ? null : index + 1;

    return {
      prev: prevIndex !== null ? this.props.components[prevIndex] : null,
      next: nextIndex !== null ? this.props.components[nextIndex] : null
    };
  },

  render() {
    var components = this.findComponents();

    return (
      <div {...classes()}>
        {components.prev
          ? <a {...classes('link', 'prev')} href={path.join('/styleguide', components.prev.category, components.prev.slug)}><strong>Previous</strong>: {slugToTitle(components.prev.slug)}</a>
          : <span {...classes('link', ['prev', 'disabled'])}>No previous page.</span>}
        {components.next
          ? <a {...classes('link', 'next')} href={path.join('/styleguide', components.next.category, components.next.slug)}><strong>Next</strong>: {slugToTitle(components.next.slug)}</a>
          : <span {...classes('link', ['next', 'disabled'])}>No next page.</span>}
      </div>
    );
  }

});

module.exports = ContextualNavigation;
