// components/image-asset/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';

import ResponsiveImage from 'components/responsive-image';
import Caption from 'components/caption';

const classes = new bem({
  name: 'image-asset',
  prefix: ''
});

const ImageAsset = React.createClass({

  displayName: 'ImageAsset',

  propTypes: {
    size: React.PropTypes.string.isRequired,
    alt: React.PropTypes.string.isRequired,
    title: React.PropTypes.string,
    src: React.PropTypes.string.isRequired,
    caption: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    source: React.PropTypes.string
  },

  render() {
    return (
      <figure itemScope itemType='http://schema.org/ImageObject' {...classes()}>
        <ResponsiveImage
          ratio='16x9'
          size={this.props.size}
          alt={this.props.alt}
          title={this.props.title}
          src={this.props.src} />
        <Caption
          title={this.props.caption}
          description={this.props.description}
          source={this.props.source} />
      </figure>
    );
  }

});

export default Wrapper(ImageAsset);
