// components/image-asset/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';

import ResponsiveImage from 'components/responsive-image';
import Caption from 'components/caption';

const classes = new bem({
  name: 'image-asset',
  prefix: ''
});

const ImageAsset = React.createClass({

  displayName: 'ImageAsset',

  propTypes: {
    src: React.PropTypes.string.isRequired,
    title: React.PropTypes.string,
    alt: React.PropTypes.string.isRequired,
    size: React.PropTypes.string.isRequired,
    caption: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    source: React.PropTypes.string
  },

  render() {
    var { size, alt, title, src, caption, description, source } = this.props;

    return (
      <figure itemScope itemType='http://schema.org/ImageObject' {...classes()}>
        <ResponsiveImage
          src={src}
          title={title}
          alt={alt}
          size={size}
          ratio='16x9' />
        <Caption
          title={caption}
          description={description}
          source={source} />
      </figure>
    );
  }

});

export default Wrapper(ImageAsset);
