// components/image-asset/wrapper.jsx
import React from 'react';

export default Component =>  React.createClass({

  displayName: 'ImageAssetWrapper',

  propTypes: {
    asset: React.PropTypes.object,
    title: React.PropTypes.string.isRequired,
    description: React.PropTypes.string,
    placement: React.PropTypes.string
  },

  mappedData(props) {

    var asset = props.asset;

    var size = this.props.placement && ['left', 'right'].indexOf(this.props.placement) > -1
      ? 'small'
      : 'max';

    var strippedDescription = props.description
      ? props.description.replace(/(<([^>]+)>)/ig, '')
      : false;

    return {
      size: size,
      src: asset.image_url,
      alt: props.title || asset.title,
      title: strippedDescription || asset.description,
      caption: props.title || asset.title,
      description: strippedDescription || asset.description,
      source: asset.source.title || (asset.source.source ? asset.source.source.title : null)
    };

  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
