// components/image-asset/wrapper.jsx
import React from 'react';

export default Component =>  React.createClass({

  displayName: 'ImageAssetWrapper',

  propTypes: {
    asset: React.PropTypes.object,
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    plaement: React.PropTypes.string
  },

  mappedData() {

    var props = this.props;
    var asset = props.asset;
    var size = this.props.placement && ['left', 'right'].indexOf(this.props.placement) > -1
      ? 'small'
      : 'max';

    var strippedDescription = props.description
      ? props.description.replace('/(<([^>]+)>)/ig', '')
      : false;

    return {
      size: size,
      alt: strippedDescription || asset.description,
      title: props.title || asset.title,
      src: asset.image_url,
      description: strippedDescription || asset.description,
      source: asset.source.title || (asset.source.source ? asset.source.source.title : null)
    };

  },

  render() {
    return (
      <Component {...this.mappedData()} />
    );
  }

});
