// components/search-input/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'search-input',
  prefix: ''
});

export default React.createClass({

  displayName: 'SearchInput',

  getInitialState() {
    return {
      focused: false
    };
  },

  comingSoon() {
    /* eslint-disable no-alert */
    window.alert('Leider wurde dieses Feature noch nicht implementiert.');
  },

  focusInput() {
    this.setState({ focused: true });
  },

  blurInput() {
    this.setState({ focused: false });
  },

  render() {
    const { focused } = this.state;
    return (
      <div {...classes(null, null, [focused ? 'is-focused' : null])}>
        <form onClick={this.comingSoon}>
          <input type="text" name="input"
            {...classes('input')}
            placeholder="Suchen"
            onFocus={this.focusInput}
            onBlur={this.blurInput}
          />
          <input type="submit" name="submit" {...classes('submit')} value="Suchen" />
        </form>
      </div>
    );
  }

});
