// components/teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import ResponsiveImage from 'components/responsive-image';

const classes = new bem({
  name: 'teaser',
  prefix: ''
});

export default React.createClass({

  displayName: 'Teaser',

  propTypes: {
    primary: React.PropTypes.string.isRequired,
    secondary: React.PropTypes.string.isRequired,
    tertiary: React.PropTypes.string,
    asset: React.PropTypes.object
  },

  getDefaultProps() {
    return {
      modifier: 'basic',
      theme: 'default'
    };
  },

  render() {
    var props = this.props;
    var modifiers = [this.props.modifier];
    var ratio = '16x9';

    if (this.props.modifier === 'promoted') {
      modifiers.push(props.theme);
      ratio = '9x10';
    }

    return (
      <article {...classes(null, props.modifier)} itemScope itemType='http://schema.org/Article'>
        <div {...classes('inner')}>
          {props.asset
            ? <div {...classes('image')}>
                <ResponsiveImage
                  ratio={ratio}
                  size='max'
                  alt={props.asset.description ? props.asset.description : ''}
                  title={props.asset.title ? props.asset.title : ''}
                  src={props.asset.image_url} />
              </div>
            : ''}
          <footer {...classes('footer')}>
            <p {...classes('secondary')}>{props.secondary}</p>
            <h1 {...classes('primary')} itemProp='name'>{props.primary}</h1>
            {props.tertiary
              ? <p {...classes('tertiary')}>{props.tertiary}</p>
              : ''}
          </footer>
        </div>
      </article>
    );

  }

});
