// components/teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import ResponsiveImage from 'components/responsive-image';

const classes = new bem({
  name: 'teaser',
  prefix: ''
});

export default React.createClass({

  propTypes: {
    primary: React.PropTypes.string.isRequired,
    secondary: React.PropTypes.string.isRequired,
    tertiary: React.PropTypes.string.isRequired,
    image: React.PropTypes.object
  },

  getDefaultProps() {
    return {
      modifier: 'basic',
      theme: 'default'
    };
  },

  render() {
    var modifiers = [this.props.modifier];
    var ratio = '16x9';

    if (this.props.modifier === 'promoted') {
      modifiers.push(this.props.theme);
      ratio = '9x10';
    }

    return (
      <div {...classes(null, this.props.modifier)}>
        <div {...classes('inner')}>
          <div {...classes('image')}>
          {this.props.asset
            ? <ResponsiveImage
              ratio={ratio}
              size='max'
              alt={this.props.asset.description ? this.props.asset.description : ''}
              title={this.props.asset.title ? this.props.asset.title : ''}
              src={this.props.asset.image_url} />
            : ''}
          </div>
          <footer {...classes('footer')}>
            <p {...classes('tertiary')}>{this.props.tertiary}</p>
            <p {...classes('primary')}>{this.props.primary}</p>
            <p {...classes('secondary')}>{this.props.secondary}</p>
          </footer>
        </div>
      </div>
    );

  }

});
