// components/teaser/slot.js
var _ = require('lodash');
var getArticleUrl = require('helpers/get-article-url');
var getMainAsset = require('helpers/get-main-asset');
var getImageProps = require('helpers/get-image-props');

module.exports = function(state) {

  // Map image sizes based on teaser variation
  var imageSizes = {
    'large': 'max',
    'basic': 'narrow',
    'promoted': '9x10',
    'inlined-right': 'tiny',
    'inlined-left': 'tiny'
  };

  // Map image ratios based on teaser variation
  var imageRatios = {
    'large': '16x9',
    'basic': '16x9',
    'promoted': '9x10',
    'inlined-right': '1x1',
    'inlined-left': '1x1'
  };

  // Map date formats based on teaser variation
  var dateFormats = {
    'large': 'long',
    'basic': 'long',
    'promoted': 'long',
    'light': 'long',
    'light-boxed': 'long',
    'inlined-right': 'short',
    'inlined-left': 'short'
  };

  var imageSizeProps = function(variation) {
    return {
      size: imageSizes[variation],
      ratio: imageRatios[variation]
    };
  };

  var parsedData = function(props) {

    var article = props.article;
    var mainAsset = getMainAsset(article.assets);
    var hasImage = imageSizes[props.variation];

    return {
      modifier: props.variation,
      title: article.title,
      catchword: article.context_label,
      link: getArticleUrl(article),
      image: hasImage
        ? _.assign(getImageProps(mainAsset), imageSizeProps(props.variation))
        : null,
      dateFormat: dateFormats[props.variation] || false,
      supplementary: dateFormats[props.variation] ? article.dc.created : false
    };

  };

  return parsedData(state);

};
