// components/mega-teaser/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Teaser from 'components/teaser';

const classes = new bem({
  name: 'mega-teaser',
  prefix: ''
});

export default React.createClass({

  displayName: 'MegaTeaser',

  propTypes: {
    teasers: React.PropTypes.array.isRequired
  },

  mapGridTeasers(teasers) {
    return teasers.map((teaser, index) => {
      return (
        <div key={index} {...classes('grid-item')}>
          <Teaser {...teaser} />
        </div>
      );
    });
  },

  render() {

    const { teasers } = this.props;

    let firstTeaser = teasers[0];

    let gridTeasers =  teasers.slice(1, teasers.length);

    return (
      <section {...classes()}>
        <Teaser {...firstTeaser} />
        {gridTeasers
          ? (
              <div {...classes('grid-wrap')}>
                <div {...classes('grid')}>
                  {this.mapGridTeasers(gridTeasers)}
                </div>
              </div>
            )
        :null}
      </section>
    );
  }

});
