// config/images.jsx
const screenTypes = {
  smallPhone: [0, 330],
  regularPhone: [330, 450],
  landscapePhone: [450, 650],
  smallTablet: [650, 800],
  regularTablet: [800, 900],
  landscapeTablet: [900, 1050],
  regularDesktop: [1050, 1150],
  wideDesktop: [1150]
};

const standardImageSizes = {
  tiny: 100,
  xsmall: 200,
  small: 300,
  regular: 350,
  medium: 450,
  large: 580,
  xlarge: 720,
  full: 900
};

const imageVariations = {
  tiny: {
    smallPhone: 80,
    regularPhone: 100,
    landscapePhone: 100,
    smallTablet: 80,
    regularTablet: 100,
    landscapeTablet: 100,
    regularDesktop: 100,
    wideDesktop: 100
  },
  small: {
    smallPhone: 280,
    regularPhone: 400,
    landscapePhone: 450,
    smallTablet: 325,
    regularTablet: 280,
    landscapeTablet: 325,
    regularDesktop: 325,
    wideDesktop: 325
  },
  large: {
    smallPhone: 280,
    regularPhone: 400,
    landscapePhone: 550,
    smallTablet: 670,
    regularTablet: 550,
    landscapeTablet: 670,
    regularDesktop: 670,
    wideDesktop: 670
  },
  full: {
    smallPhone: 330,
    regularPhone: 450,
    landscapePhone: 650,
    smallTablet: 800,
    regularTablet: 900,
    landscapeTablet: 850,
    regularDesktop: 880,
    wideDesktop: 680
  }
};

export default {
  screenTypes: screenTypes,
  standardImageSizes: standardImageSizes,
  imageVariations: imageVariations
};
