// config/images.jsx
/*eslint-disable key-spacing */

// Viewport type given by screen width range
const screenTypes = {
  smallPhone:       [0,    330],
  regularPhone:     [330,  450],
  landscapePhone:   [450,  650],
  smallTablet:      [650,  800],
  regularTablet:    [800,  900],
  landscapeTablet:  [900,  1050],
  regularDesktop:   [1050, 1150],
  wideDesktop:      [1150]
};

// Widths requested from server by variation
const standardImageSizes = {
  micro:   80,
  avatar:  100,
  narrow:  280,
  mobile:  330,
  small:   400,
  regular: 450,
  medium:  550,
  large:   680,
  xlarge:  800,
  wide:    850,
  full:    900
};

// Image sizes used as <ResponsiveImage size="small" />
// Maps viewport type to image size
const imageVariations = {
  tiny: {
    smallPhone: 'micro',
    regularPhone: 'avatar',
    landscapePhone: 'avatar',
    smallTablet: 'micro',
    regularTablet: 'avatar',
    landscapeTablet: 'avatar',
    regularDesktop: 'avatar',
    wideDesktop: 'avatar'
  },
  small: {
    smallPhone: 'narrow',
    regularPhone: 'small',
    landscapePhone: 'regular',
    smallTablet: 'mobile',
    regularTablet: 'narrow',
    landscapeTablet: 'mobile',
    regularDesktop: 'mobile',
    wideDesktop: 'mobile'
  },
  large: {
    smallPhone: 'narrow',
    regularPhone: 'small',
    landscapePhone: 'medium',
    smallTablet: 'large',
    regularTablet: 'medium',
    landscapeTablet: 'large',
    regularDesktop: 'large',
    wideDesktop: 'large'
  },
  full: {
    smallPhone: 'mobile',
    regularPhone: 'regular',
    landscapePhone: 'large',
    smallTablet: 'xlarge',
    regularTablet: 'full',
    landscapeTablet: 'wide',
    regularDesktop: 'full',
    wideDesktop: 'large'
  }
};

/*eslint-enable key-spacing */

export default {
  screenTypes: screenTypes,
  standardImageSizes: standardImageSizes,
  imageVariations: imageVariations
};
