// pages/sub-ressort/index.jsx
import React from 'react';

import Container from 'components/container';
import RessortHeader from 'components/ressort-header';
import Grid from 'components/grid';
import Teaser from 'components/teaser';
import { getTeaserFromArticle } from 'helpers';

export default React.createClass({

  displayName: 'TopicPage',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    slug: React.PropTypes.string.isRequired,
    subnav: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    articles: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mapTeaserItems(articles) {
    return articles.map((article, index) => {
      var articleTeaser = getTeaserFromArticle({
        article: article,
        variation: index === 0 || index === 3 ? 'large' : 'basic',
        fallback: 'light'
      });
      return <Teaser key={index} {...articleTeaser} />;
    });
  },

  render() {
    return (
      <Container>
        <RessortHeader {...this.props} />
        <Grid modifier='ressort' items={this.mapTeaserItems(this.props.articles)} />
      </Container>
    );
  }

});
