// helpers/article/get-title.jsx
import _ from 'lodash';

import getName from 'helpers/user/get-name';

// Map title icons based on teaser labeltype
const titleIcons = {
  video: 'video',
  quiz: 'quiz',
  survey: 'survey',
  map: 'map',
  live: 'bullet',
  leserbeitrag: 'megaphone',
  vereinsmeldung: 'megaphone',
  sportverein: 'megaphone',
  gemeindebeitrag: 'megaphone'
};

// Map prefixes to theme type
const prefixes = {
  leserbeitrag: 'Leserbeitrag:',
  vereinsmeldung: 'Vereinsmeldung:',
  sportverein: 'Sportverein:',
  gemeindebeitrag: 'Gemeindebeitrag:'
};

// Map prefixes to teaser variation
const prefixedVariations = [
  'large'
];

// Map author suffixes based on teaser labeltype
const suffixes = [
  'meinung'
];

// Ignored suffixes by teaser variation
const ignoredSuffixes = [
  'quote',
  'author'
];

export default function(article, variation) {

  let labelType = _.get(article, 'labeltype');
  let user = _.get(article, 'author_detail.user');

  let needsPrefix = _.includes(prefixedVariations, variation);
  let prefix = needsPrefix ? _.get(prefixes, labelType) : false;

  let needsSuffix = user && _.includes(suffixes, labelType) && !_.includes(ignoredSuffixes, variation);
  let suffix = needsSuffix ? getName(user) : false;

  return {
    text: _.get(article, 'title'),
    icon: _.get(titleIcons, labelType),
    prefix: prefix,
    suffix: suffix
  };
}
