// helpers/article/get-banner.jsx
import _ from 'lodash';

import { goldbachImageSource } from 'helpers/image';

const SPONSORED = 'sponsored';

function squareImage(src) {
  return goldbachImageSource({
    width: 70,
    height: 70,
    fill: true,
    src: src
  });
}

function getBannerType(article) {
  if (_.get(article, 'sponsor.text')) {
    return SPONSORED;
  } else if (article.content_type === 'ugcnewsarticle') {
    return article.labeltype || _.get(article, 'ressorts.0.urlpart');
  }

  return false;
}

function blocksForBannerType (bannerType, article) {
  function authorBlock (data, text) {
    var author = _.get(data, 'owner') || {};

    if (!author.firstname) {
      return null;
    }

    return {
      "text": text,
      "linkText": [author.firstname, author.lastname].join(' '),
      "image": author.image_url ? squareImage(author.image_url) : null,
      "itemProp": "author"
    };
  }

  function sponsorBlock(data) {
    var sponsor = _.get(data, 'sponsor', {});

    if (!sponsor.text) {
      return null;
    }

    return {
      "text": sponsor.text,
      "link": sponsor.url,
      "linkText": sponsor.title,
      "image": sponsor.image.image_url,
      "imageRounded": false
    };
  }

  function clubBlock(data, type) {
    var club = _.get(data, 'for_content') || {};
    var text = type === 'sportverein' ? 'Sportvereinsmeldung' : _.capitalize(type);

    if (!club.title) {
      return null;
    }

    return {
      "text": text + ' zu',
      "linkText": club.title,
      "image": _.get(club, 'logo.image_url'),
      "itemProp": 'about',
      "imageRounded": false
    };
  }

  function cityBlock(data, text) {
    var city = _.get(article, 'cities.0', {});

    if (!city.name) {
      return null;
    }

    return {
      "text": text,
      "linkText": city.name,
      "link": '/gemeinde/' + city.urlpart,
      "image": '/static/az/img/Flaggen/gemeinde/' + city.id + '.png',
      "itemProp": "contentLocation",
      "imageRounded": false
    };
  }

  switch (bannerType) {
    case 'leserbeitrag':
      return [
        authorBlock(article, 'Leserbeitrag von'),
        cityBlock(article, 'aus')
      ];
    case 'gemeindebeitrag':
      return [
        cityBlock(article, 'Gemeindebeitrag aus'),
        authorBlock(article, 'von'),
      ];
    case 'vereinsmeldung':
    case 'sportverein':
      return [
        clubBlock(article, bannerType),
        authorBlock(article, 'von')
      ];
    case SPONSORED:
      return [
        sponsorBlock(article)
      ];
  }
}

export default function(article) {
  var bannerType = getBannerType(article);

  if (!bannerType) {
    return null;
  }

  return {
    icon: bannerType === SPONSORED ? false : 'megaphone',
    modifier: bannerType,
    blocks: _.compact(blocksForBannerType(bannerType, article))
  };
}
