// components/topic-body/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';

import Teaser from 'components/teaser';
import Button from 'components/button';

const classes = new bem({
  name: 'topic-body',
  prefix: ''
});

const TopicBody = React.createClass({

  displayName: 'TopicBody',

  propTypes: {
    articles: React.PropTypes.arrayOf(React.PropTypes.object),
    page: React.PropTypes.number
  },

  defaultShowing: 12,

  getInitialState() {
    return {
      showing: this.props.page * this.defaultShowing || this.defaultShowing
    };
  },

  sliceResults(articles, count) {
    return articles.slice(0, count);
  },

  paginationCalc(offset) {
    var page = offset || 1;
    return this.state.showing / this.defaultShowing + page;
  },

  loadMore() {
    this.setState({
      showing: this.state.showing + this.defaultShowing
    });
    this.updateHistory(this.paginationCalc(1));
  },

  updateHistory(index) {
    history.replaceState(null, document.title, '?page=' + index);
  },

  render() {

    var { articles } = this.props;
    var { showing } = this.state;
    var needsPagination = articles.length > showing;

    return (
      <div {...classes()}>
        <div {...classes('grid')}>
          {this.sliceResults(articles, showing).map((article, index) => {
            return (
              <div {...classes('cell')} key={index}>
                <Teaser key={index} {...article} modifier='basic' />
              </div>
            );
          })}
        </div>
        {needsPagination
         ? <nav {...classes('pagination')}>
             <Button href={'?page=' + this.paginationCalc(2)} _event={this.loadMore} {...classes('btn')} label='Mehr' />
           </nav>
        :null}
      </div>
    );

  }

});

export default Wrapper(TopicBody);
