// components/top-articles/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import { getTeaserFromArticle } from 'helpers';

import Teaser from 'components/teaser';
import Slider from 'mixins/slider';

const classes = new bem({
  name: 'top-articles',
  prefix: ''
});

export default React.createClass({

  displayName: 'TopArticles',

  mixins: [Slider],

  propTypes: {
    articles: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mapTeaserItems(articles) {
    return articles.map((article, index) => {
      var articleTeaser = getTeaserFromArticle({
        article: article,
        variation: 'title'
      });
      return <Teaser key={index} {...articleTeaser} />;
    });
  },

  render() {

    return (
      <section {...classes()}>
        <header {...classes('header')}>
          <p {...classes('catchword')}>Und Jetzt</p>
          <h1 {...classes('title')}>Auf Einen Blick</h1>
        </header>
        <div {...classes('teasers')}>
          {this.mapTeaserItems(this.props.articles)}
        </div>
      </section>
    );
  }

});
