// components/teaser-title/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Icon from 'components/icon';

const classes = new bem({
  name: 'teaser-title',
  prefix: ''
});

export default React.createClass({

  displayName: 'TeaserTitle',

  propTypes: {
    title: React.PropTypes.string.isRequired,
    modifier: React.PropTypes.string,
    theme: React.PropTypes.string,
    prefix: React.PropTypes.string,
    suffix: React.PropTypes.string,
    icon: React.PropTypes.string,
    label: React.PropTypes.string
  },

  render() {
    const { title, modifier, theme, prefix, suffix, icon, label } = this.props;

    let modifiers = {
      [modifier]: modifier,
      [theme]: theme,
      [label]: label,
      'theme-article': !theme
    };

    return (
      <h1 {...classes(null, modifiers)}>
        {icon
          ? <Icon className='teaser-title__icon' id={icon} />
        :null}
        {prefix
          ? <span {...classes('prefix')}>{prefix}</span>
        :null}
        <span {...classes('text-wrap')} itemProp='headline'>{title}</span>
        {suffix
          ? <span {...classes('suffix')}>{suffix}</span>
        :null}
      </h1>
    );
  }

});
