// components/teaser-title/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Icon from 'components/icon';

const classes = new bem({
  name: 'teaser-title',
  prefix: ''
});

export default React.createClass({

  displayName: 'TeaserTitle',

  propTypes: {
    text: React.PropTypes.string.isRequired,
    modifiers: React.PropTypes.array,
    prefix: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.bool
    ]),
    suffix: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.bool
    ]),
    icon: React.PropTypes.string
  },

  render() {
    const { text, modifiers, prefix, suffix, icon } = this.props;

    let iconEl;

    if (icon) {
      iconEl = icon === 'bullet'
        ? <i {...classes('bullet')} aira-hidden={true} />
        : <Icon {...classes('icon')} id={icon} />;
    }

    return (
      <h1 {...classes(null, modifiers)}>
        {iconEl}
        {prefix
          ? <span {...classes('prefix')}>{prefix}</span>
        :null}
        <span {...classes('text-wrap')} itemProp='headline'>{text}</span>
        {suffix
          ? <span {...classes('suffix')}>{suffix}</span>
        :null}
      </h1>
    );
  }

});
