// components/topic-pagination/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Button from 'components/button';

const classes = new bem({
  name: 'topic-pagination',
  prefix: ''
});

export default React.createClass({

  displayName: 'TopicPagination',

  propTypes: {
    topic: React.PropTypes.string.isRequired,
    perPage: React.PropTypes.number.isRequired,
    current: React.PropTypes.number.isRequired,
    total: React.PropTypes.number.isRequired
  },

  paginationUrl(topic, modifier) {
    var isFirstPageLink = this.props.current === 2 && modifier < 0;

    var path = '/thema/';
    path += topic;

    if (!isFirstPageLink) {
      path += '/seite/';
      path += this.props.current + modifier;
    }

    return path;
  },

  render() {

    var { topic, perPage, current, total } = this.props;

    var last = Math.ceil(total / perPage);

    var isFirst = current === 1;
    var isLast = current === last;
    var firstPageAttrs = isFirst ? { disabled: true } : {};
    var lastPageAttrs = isLast ? { disabled: true } : {};

    return (
      <nav {...classes()}>
        <ul {...classes('list')}>
          <li {...classes('list-item', (isFirst ? 'is-disabled' : null))}>
            <Button {...classes('btn')}
              url={!isFirst ? this.paginationUrl(topic, -1) : null}
              label='Zurück'
              attributes={firstPageAttrs}
            />
          </li>
          <li {...classes('list-item', (isLast ? 'is-disabled' : null))}>
            <Button {...classes('btn')}
              url={!isLast ? this.paginationUrl(topic, +1) : null}
              label='Weiter'
              attributes={lastPageAttrs}
            />
          </li>
        </ul>
        <em {...classes('info')}>Seite<br /> {current} / {last}</em>
     </nav>
    );
  }

});
