// components/newsletter-form/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import Checkbox from 'inputs/checkbox';
import TextInput from 'inputs/text';
import Button from 'components/button';
import Wrapper from './wrapper';

const classes = new bem('newsletter-form');

const NewsletterForm = React.createClass({

  propTypes: {
    host: React.PropTypes.string.isRequired,
    u: React.PropTypes.string.isRequired,
    id: React.PropTypes.string.isRequired,
    regions: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    className: React.PropTypes.string
  },

  getDefaultProps: () => ({
    host: 'azmedien.us8.list-manage.com',
    u: '16deca519793f52a2dbd22829',
    id: 'bf9fdc38c3'
  }),

  getPostPath() {
    const { host, u, id } = this.props;
    return 'https://' + host + '/subscribe/post?u=' + u + '&amp;id=' + id;
  },

  renderFormField(field) {
    let { hidden, group, id, label, checked, region } = field;
    let fieldName = `group[${group}][${id}]`;
    let labelComponent = (
      <span>
        <strong>Newsletter {region}</strong>
        {' ' + label}
      </span>
    );

    if (!hidden) {
      return (
        <Checkbox
          className="form-row"
          key= { fieldName }
          name={ fieldName }
          label={ labelComponent }
          value={id}
        />
      );
    } else if (checked) {
      return <input name={fieldName} type="hidden" value={id} />;
    }

    return null;
  },

  renderWerbungCheckbox() {
    // we only want to render the checkbox when we're showing the list of items
    if (this.props.regions.length > 1) {
      return (
        <Checkbox
          className="form-row"
          name="MERGE6"
          value="Ja"
          label='Ich bin damit einverstanden, dass mich die AZ Medien AG künftig per Email über weitere interessante Angebote informiert.'
          hint={ <span>Diese Einwilligungserklärung kann jederzeit via <a href="mailto:unsubscribe@azmedien.ch">unsubscribe@azmedien.ch</a> widerrufen werden. </span>}
        />
      );
    }
  },

  renderAntiSpamField() {
    return (
      <div style={{
        position: 'absolute',
        left: '-5000px'
      }}>
        <input type="text" name="b_16deca519793f52a2dbd22829_bf9fdc38c3" tabIndex="-1" value=""/>
      </div>
    );
  },

  render() {
    return (
      <form {...classes({ extra: this.props.className })} action={ this.getPostPath() } method="post" target="_blank" noValidate="novalidate">
          { this.props.regions.map(this.renderFormField) }

          <TextInput
            className="form-row"
            label="E-Mail"
            name="EMAIL"
            type="email"
          />

          { this.renderWerbungCheckbox() }

          { this.renderAntiSpamField() }

          <div className="form-row">
            <Button attributes={{ type: 'submit' }} label="Anmelden" modifier="primary"/>
          </div>
      </form>
    );
  }
});

export default Wrapper(NewsletterForm);
