// components/newsletter-form/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem('newsletter-form');

const NewsletterForm = React.createClass({

  propTypes: {
    host: React.PropTypes.string.isRequired,
    u: React.PropTypes.string.isRequired,
    id: React.PropTypes.string.isRequired,
    className: React.PropTypes.string
  },

  getDefaultProps: () => ({
    host: 'azmedien.us8.list-manage.com',
    u: '16deca519793f52a2dbd22829',
    id: 'bf9fdc38c3'
  }),

  getPostPath() {
    const { host, u, id } = this.props;
    return 'https://' + host + '/subscribe/post?u=' + u + '&amp;id=' + id;
  },

  renderAntiSpamField() {
    return (
      <div style={{
        position: 'absolute',
        left: '-5000px'
      }}>
        <input type="text" name={['b', this.props.u, this.props.id].join('_')} tabIndex="-1" value=""/>
      </div>
    );
  },

  render() {
    return (
      <form {...classes({ extra: this.props.className })} action={ this.getPostPath() } method="post" target="_blank" noValidate="novalidate">
        { this.renderAntiSpamField() }
        { this.props.children }
      </form>
    );
  }
});

export default NewsletterForm;
