// components/newsletter-form/wrapper.jsx
import { get, defaults, each, capitalize } from 'lodash';
import React from 'react';
import regionData from 'components/newsletter-form/.config.json';

function flattenRegions(regions) {
  let output = [];

  each(regions, function(towns, regionName) {
    each(towns, town => output.push(defaults({
      region: capitalize(regionName)
    }, town)));
  });

  return output;
}

export default Compoonent => React.createClass({

  displayName: 'NewsletterFormWrapper',

  propTypes: {
    regionPath: React.PropTypes.string
  },

  getInitialState() {
    let regionByFilter = this.regionByFilter(this.props.regionPath);

    return {
      regions: regionByFilter ? [regionByFilter] : flattenRegions(regionData)
    };
  },

  regionByFilter(filter) {
    var region = get(regionData, filter && filter.split('/'));

    if (region) {
      return defaults({
        hidden: true,
        checked: true
      }, region);
    }

    return null;
  },

  render() {
    return (
      <Compoonent {...this.props} regions={this.state.regions} />
    );
  }
});
