// components/video-library-viewer/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Video from 'components/video';

const classes = new bem({
  name: 'video-library-viewer',
  prefix: ''
});

export default React.createClass({

  displayName: 'VideoLibraryViewer',

  propTypes: {
    accountID: React.PropTypes.string.isRequired,
    playerID: React.PropTypes.string.isRequired,
    playerWithoutAdsID: React.PropTypes.string,
    autoPlay: React.PropTypes.bool.isRequired,
    videoID: React.PropTypes.string.isRequired,
    meta: React.PropTypes.object.isRequired,
    caption: React.PropTypes.object,
    trackingURL: React.PropTypes.string.isRequired
  },

  // @TODO: put into helpers function (used in article and video library)
  // @QUESTION: best practice to construct URL (any lib?)?
  getNetMetrixTrackingURL(account, domain, location) {
    var url = 'https://'
                + account
                + '.wemfbox.ch/cgi-bin/ivw/CP/'
                + domain
                + '?r='
                //@TODO: encode URI Component
                + location
                + '&d=&x=';

    return url;
  },

  render() {
    var trackingURL = this.getNetMetrixTrackingURL('aznetz', 'aaz', 'http://www.aargauerzeitung.ch/ressort/subressort/article-title-129298028');
    var meta = {
      thumbnailUrl: this.props.thumbnailUrl
    };
    var caption = {
      title: this.props.name,
      description: this.props.description,
      source: "@TODO: Source needed from extended meta data or from source:-tag of Kaltura"
    };

    return (
      <div {...classes('container')}>
        <Video
          {...this.props}
          autoPlay={false}
          accountID={this.props.accountID}
          playerID={this.props.playerID}
          playerWithoutAdsID={this.props.playerWithoutAdsID}
          kaltura_id={this.props.videoID}
          meta={meta}
          caption={caption}
          trackingURL={trackingURL}
          placement=""
        />
      </div>
    );
  }

});
