// components/footer-sections/index.jsx
import React from 'react';
import bem from 'react-bem-helper';
import _ from 'lodash';
import Button from 'components/button';
import path from 'path';

const classes = new bem({
  name: 'footer-links',
  prefix: ''
});

export default React.createClass({

  displayName: 'FooterSections',

  propTypes: {
    sections: React.PropTypes.object.isRequired,
    id: React.PropTypes.string.isRequired
  },

  getInitialState() {
    return {
      mode: 'collapsed'
    };
  },

  toggleFooter() {
    this.setState({
      mode: this.state.mode === 'collapsed'
        ? 'expanded'
        : 'collapsed'
    });
  },

  renderSection(section, links) {
    let linksMarkup = [];

    _.forEach(links, function(link) {
      linksMarkup.push(
        <li {...classes('links-item')}>
          <a {...classes('link')} href={link.url}>{link.name}</a>
        </li>
      );
    });

    return (
      <section {...classes('section')}>
        <h2 {...classes('section-title')}>{section.replace(':', '')}</h2>
        <ul {...classes('links')}>{linksMarkup}</ul>
      </section>
    );
  },

  mapFooterSections(sections) {
    let markup = [];

    for (let section in sections) {
      let links = sections[section];
      markup.push(this.renderSection(section, links));
    }

    return markup;
  },

  render() {

    const { sections } = this.props;

    const { mode } = this.state;

    let label = mode === 'collapsed'
      ? 'Verlagsangebote einblenden'
      : 'Verlagsangebote ausblenden';

    const prefix = process.env.STATIC_ASSETS || '/__node__/__static__/client';
    let logoSrc = path.join(prefix, 'images', 'logos', 'az_medien.svg');

    return (
      <div {...classes()}>
        <Button label={label} _event={this.toggleFooter} />
        <div {...classes(null, mode)}>
          <h2 className="footer-headline">Alle Angebote von <img className="footer-azm-logo" src={logoSrc} /></h2>
          {this.mapFooterSections(sections)}
        </div>
      </div>
    );
  }

});
