// styleguide/iframe/index.jsx
'use strict';

const React = require('react');
const bem = require('react-bem-helper');
const path = require('path');

const classes = new bem({
  name: 'iframe',
  prefix: 'styleguide-'
});

const Menu = React.createClass({

  propTypes: {
    component: React.PropTypes.object.isRequired,
    variation: React.PropTypes.object
  },

  constructUrl(component, variation) {
    return path.join('/styleguide', component.category, component.slug, (variation ? variation.slug : ''), 'preview');
  },

  resizeHeight() {
    var iframe = React.findDOMNode(this.refs.iframePreview);
    var html = iframe.contentWindow.document.documentElement;
    var body = iframe.contentWindow.document.body;
    var height = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight);
    iframe.height = height + 'px';
  },

  render() {
    var props = this.props;
    var modifier = props.variation && props.variation.width ? props.variation.width : props.component.width || 'default';

    return (
      <iframe {...classes('', modifier)} width='100%' ref='iframePreview' onLoad={this.resizeHeight} src={this.constructUrl(props.component, props.variation)}>
        {'Unfortunately, the iframe could not be displayed, sorry! :('}
      </iframe>
    );
  }

});

module.exports = Menu;
