// styleguide/layout/index.jsx
'use strict';

const React = require('react');
const _ = require('lodash');

const Main = require('styleguide/main');
const Sidebar = require('styleguide/sidebar');
const Intro = require('styleguide/intro');
const Component = require('styleguide/component');

var Styleguide = React.createClass({

  propTypes: {
    components: React.PropTypes.array,
    route: React.PropTypes.string,
    children: React.PropTypes.element
  },

  getInitialState() {

    var props = this.props;

    return {
      component: _.find(props.components, function(chr) {
        return chr.slug === props.route;
      })
    };

  },

  render() {

    return (
      <div className='styleguide'>
        <Sidebar
          components={this.props.components}
          active={this.state.component}
        />
        <Main>
          {this.props.children ? <Component component={this.state.component}>{this.props.children}</Component> : <Intro/>}
        </Main>
      </div>
    );

  }

});

module.exports = Styleguide;
