// styleguide/layout/index.jsx
'use strict';

const React = require('react');
const slugify = require('slugify');

const Main = require('styleguide/main');
const Sidebar = require('styleguide/sidebar');
const Intro = require('styleguide/intro');
const Component = require('styleguide/component');

const Styleguide = React.createClass({

  // These are passed from server.js/browser.js
  propTypes: {
    components: React.PropTypes.object.isRequired,
    category: React.PropTypes.string.isRequired,
    component: React.PropTypes.object,
    children: React.PropTypes.element
  },

  // findComponentBySlug(slug) {
  //   var props = this.props;
  //   return _.find(props.components, function(chr) {
  //     return chr.slug === slug;
  //   });
  // },

  getInitialState() {
    return {
      component: this.props.component,
      category: this.props.category
    };
  },

  render() {

    var variation, component = this.state.component;
    if (component && component.variations) {
      variation = component.variations[component.variationIndex];
      variation.slug = slugify(variation.title).toLowerCase();
    }

    var mainContent = <Intro/>;
    if (component) {
      mainContent = (
        <Component component={component} category={this.props.category} variation={variation}>
          {this.props.children}
        </Component>
      );
    }

    return (
      <div className='styleguide'>
        <Sidebar
          category={this.props.category}
          components={this.props.components}
          active={component || false}
        />
        <Main>
          {mainContent}
        </Main>
      </div>
    );
  }

});

module.exports = Styleguide;
