// components/subressort-featured/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import { getTeaserFromArticle } from 'helpers';

import Teaser from 'components/teaser';
import TopArticles from 'components/top-articles';

const classes = new bem({
  name: 'subressort-featured',
  prefix: ''
});

export default React.createClass({

  displayName: 'SubressortFeatured',

  propTypes: {
    featured: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
    articles: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mapTeaserItem(article) {
    var articleTeaser = getTeaserFromArticle({
      article: article,
      variation: 'large'
    });
    return <Teaser {...articleTeaser} />;
  },

  render() {
    return (
      <div {...classes()}>
        <div {...classes('articles')}>
          <div {...classes('articles-item')}>
            {this.mapTeaserItem(this.props.articles[0])}
          </div>
          <div {...classes('articles-item')}>
            {this.mapTeaserItem(this.props.articles[1])}
          </div>
        </div>
        <div {...classes('top-articles')}>
          <TopArticles articles={this.props.featured} />
        </div>
      </div>
    );
  }

});
