// components/caption/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'caption',
  prefix: ''
});

export default React.createClass({

  displayName: 'Caption',

  propTypes: {
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    source: React.PropTypes.string,
    element: React.PropTypes.string
  },

  getDefaultProps() {
    return {
      element: 'figcaption'
    };
  },

  render() {

    var title, description, source, captionContents;
    var props = this.props;

    if (props.title) {
      title = <p {...classes('title')} itemProp='name'>{props.title}</p>;
    }

    if (props.source) {
      source = <span {...classes('source')} itemProp='author'>{props.source}</span>;
    }

    if (props.description) {
      description = <span {...classes('description')} itemProp='description'>{props.description}</span>;
    }

    captionContents = (
      <div>
        {title}
        {description || source ? <p>{description} {source}</p> : ''}
      </div>
    );

    return React.createElement(props.element, { className: classes().className }, captionContents);

  }

});
