// components/caption/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'caption',
  prefix: ''
});

export default React.createClass({

  displayName: 'Caption',

  propTypes: {
    title: React.PropTypes.string,
    description: React.PropTypes.string,
    source: React.PropTypes.string,
    element: React.PropTypes.string
  },

  getDefaultProps() {
    return {
      element: 'figcaption'
    };
  },

  render() {

    var { title, source, description, element } = this.props;

    var captionContents = (
      <div>
        {title && source !== title
          ? <p {...classes('title', [title && source ? 'bold' : null])} itemProp='name'>{title}</p>
        :null}
        {source || description
          ? <p>
              {description
                ? <span {...classes('description')} itemProp='description'>{description}</span>
              :null}
              {source
                ? <span {...classes('source')} itemProp='author'>{source}</span>
              :null}
            </p>
        :null}
      </div>
    );

    return React.createElement(element, { className: classes().className }, captionContents);

  }

});
