// components/recommendations-bottom/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

const classes = new bem({
  name: 'recommendations-bottom',
  prefix: ''
});

export default React.createClass({

  variations: ['large', 'light', 'light', 'basic', 'basic', 'basic'],

  propTypes: {
    teasers: React.PropTypes.array.isRequired
  },

  addModifiers(teasers) {
    return teasers.map(function(teaser, index) {
      teaser.modifier = this.variations[index];
      return teaser;
    }, this);
  },

  render() {
    return (
      <div {...classes()}>
        <p {...classes('catchword')}>Und Jetzt</p>
        <h3 {...classes('title')}>Lesen Sie Ausserdem</h3>
        <div {...classes('grid')}>
          <div {...classes('cell', 'large')}></div>
          <div {...classes('cell', 'light')}></div>
          <div {...classes('cell', 'light')}></div>
          <div {...classes('cell', 'basic')}></div>
          <div {...classes('cell', 'basic')}></div>
          <div {...classes('cell', 'basic')}></div>
        </div>
      </div>
    );
  }

});
