// components/recommendations-bottom/index.jsx
import React from 'react';
import bem from 'react-bem-helper';

import Wrapper from './wrapper';
import Section from 'components/section';
import SectionHeader from 'components/section-header';
import Container from 'components/container';
import Teaser from 'components/teaser';
import TeaserSwitchable from 'components/teaser-switchable';
import Grid from 'components/grid';

const classes = new bem({
  name: 'recommendations-bottom',
  prefix: ''
});

const RecommendationsBottom = React.createClass({

  displayName: 'RecommendationsBottom',

  propTypes: {
    teasers: React.PropTypes.array.isRequired
  },

  render() {

    var items = [];
    var stash = [];

    this.props.teasers.forEach((teaser, index) => {
      if (index === 0) {
        items.push(<Teaser key={index} {...teaser} />);
      } else if (index === 1 || index === 2) {
        stash.push(<div {...classes('teaser-wrapper')} key={index}><Teaser key={index} {...teaser} /></div>);
        if (index === 2) items.push(<div key={index}>{stash}</div>);
      } else {
        items.push(<TeaserSwitchable key={index} {...teaser} />);
      }
    });

    return (
      <div {...classes()}>
        <Section modifier="default">
          <Container>
            <SectionHeader title='Lesen Sie Ausserdem' catchword='Und Jetzt' />
            <Grid items={items} modifier={'three-up-featured'} />
          </Container>
        </Section>
      </div>
    );
  }

});

export default Wrapper(RecommendationsBottom);
