// components/recommendations-bottom/wrapper.jsx
import React from 'react';
import _ from 'lodash';

import { getTeaserFromArticle, getTeaserSwitchableFromArticle } from 'helpers/teaser';

export default Component => React.createClass({

  displayName: 'LatestArticlesWrapper',

  propTypes: {
    articles: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
  },

  mappedTeasers(teasers) {
    return teasers.map((teaser, index) => {
      let teaserData;
      if (index === 0) {
        teaserData = getTeaserFromArticle({
          article: teaser,
          variation: 'large'
        });
      } else if (index === 1 || index === 2) {
        teaserData = getTeaserFromArticle({
          article: teaser,
          variation: 'light'
        });
      } else {
        teaserData = getTeaserSwitchableFromArticle({
          article: teaser,
          variation: 'inlined-left-to-basic'
        });
      }
      return teaserData;
    });
  },


  mappedData(props) {
    return _.assign({}, props, { teasers: this.mappedTeasers(props.articles.slice(0, 6)) });
  },

  render() {
    return (
      <Component {...this.mappedData(this.props)} />
    );
  }

});
